{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "from collections import Counter\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get Explanation Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g1_e2v\\expl_5\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\expl_all\n"
     ]
    }
   ],
   "source": [
    "dataset_nrs = [1, 2]\n",
    "embedding_method = 'e2v'\n",
    "\n",
    "expl_folders = {1: 'expl_5', 2: 'expl_all'}\n",
    "explanations_per_dataset = {}\n",
    "explanation_pairs_per_dataset = {}\n",
    "\n",
    "curr_working_dir = os.getcwd()\n",
    "curr_output_dir = os.path.join(curr_working_dir, 'output')\n",
    "\n",
    "chosen_explanations_per_dataset = {}\n",
    "\n",
    "for dataset_nr in dataset_nrs:\n",
    "    nodes = pd.read_csv(f'output/indexed_nodes_{dataset_nr}.csv')\n",
    "    \n",
    "    dataset_output_dir = os.path.join(curr_output_dir, f'g{dataset_nr}_{embedding_method}', expl_folders[dataset_nr])\n",
    "    print(dataset_output_dir)\n",
    "    \n",
    "    all_explanations = []\n",
    "    all_graphs = []\n",
    "    all_pairs = []\n",
    "    for item in os.listdir(dataset_output_dir):\n",
    "        if 'incomplete' not in item and '.gpickle' in item:\n",
    "            \n",
    "            with open(os.path.join(dataset_output_dir, item), 'rb') as f:\n",
    "                G = pickle.load(f)\n",
    "                all_graphs.append(G)\n",
    "                \n",
    "            file_name_explanation = item.split('_graph.gpickle')[0]\n",
    "            all_explanations.append(file_name_explanation)\n",
    "            pair_file_name = f'{file_name_explanation}_pair.pkl'\n",
    "            \n",
    "            with open(os.path.join(dataset_output_dir, pair_file_name), 'rb') as f:\n",
    "                loaded_info = pickle.load(f)\n",
    "                \n",
    "                symptom_index = nodes.loc[nodes['id'] == loaded_info['symptom_id']].index[0]\n",
    "                symptom_label = nodes.loc[nodes['id'] == loaded_info['symptom_id']]['label'].values[0]\n",
    "                \n",
    "                drug_index = nodes.loc[nodes['id'] == loaded_info['drug_id']].index[0]\n",
    "                drug_label = nodes.loc[nodes['id'] == loaded_info['drug_id']]['label'].values[0]\n",
    "                \n",
    "                all_pairs.append([f'{symptom_label} {symptom_index}', f'{drug_label} {drug_index}'])\n",
    "            \n",
    "    explanations_per_dataset[dataset_nr] = all_graphs\n",
    "    explanation_pairs_per_dataset[dataset_nr] = all_pairs\n",
    "    \n",
    "    random.seed(111)\n",
    "    chosen_explanations = []\n",
    "    for expl_name in random.choices(all_explanations, k=3):\n",
    "        with open(os.path.join(dataset_output_dir, f'{expl_name}_graph.gpickle'), 'rb') as f:\n",
    "            G = pickle.load(f)\n",
    "            \n",
    "        with open(os.path.join(dataset_output_dir, f'{expl_name}_pair.pkl'), 'rb') as f:\n",
    "            loaded_info = pickle.load(f)\n",
    "        \n",
    "        chosen_explanations.append({'filename': expl_name, 'G': G, 'pair': loaded_info})\n",
    "    chosen_explanations_per_dataset[dataset_nr] = chosen_explanations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: [{'filename': 'explanation_2_run_007',\n",
       "   'G': <networkx.classes.graph.Graph at 0x1cbf599a020>,\n",
       "   'pair': {'drug_id': '5252',\n",
       "    'drug_name': 'neratinib',\n",
       "    'symptom_id': 'HP:0000750',\n",
       "    'symptom_name': 'Delayed speech and language development',\n",
       "    'complete': True}},\n",
       "  {'filename': 'explanation_1_run_005',\n",
       "   'G': <networkx.classes.graph.Graph at 0x1cbf59985e0>,\n",
       "   'pair': {'drug_id': '5252',\n",
       "    'drug_name': 'neratinib',\n",
       "    'symptom_id': 'HP:0001263',\n",
       "    'symptom_name': 'Global developmental delay',\n",
       "    'complete': True}},\n",
       "  {'filename': 'explanation_2_run_010',\n",
       "   'G': <networkx.classes.graph.Graph at 0x1cbf599bf70>,\n",
       "   'pair': {'drug_id': '5252',\n",
       "    'drug_name': 'neratinib',\n",
       "    'symptom_id': 'HP:0000750',\n",
       "    'symptom_name': 'Delayed speech and language development',\n",
       "    'complete': True}}],\n",
       " 2: [{'filename': 'explanation_2_run_006',\n",
       "   'G': <networkx.classes.graph.Graph at 0x1cbf59984f0>,\n",
       "   'pair': {'drug_id': '1576',\n",
       "    'drug_name': 'levosimendan',\n",
       "    'symptom_id': 'HP:0011675',\n",
       "    'symptom_name': 'Arrhythmia',\n",
       "    'complete': True}},\n",
       "  {'filename': 'explanation_1_run_004',\n",
       "   'G': <networkx.classes.graph.Graph at 0x1cbf5999300>,\n",
       "   'pair': {'drug_id': '1576',\n",
       "    'drug_name': 'levosimendan',\n",
       "    'symptom_id': 'HP:0001638',\n",
       "    'symptom_name': 'Cardiomyopathy',\n",
       "    'complete': True}},\n",
       "  {'filename': 'explanation_2_run_009',\n",
       "   'G': <networkx.classes.graph.Graph at 0x1cbf59989d0>,\n",
       "   'pair': {'drug_id': '1576',\n",
       "    'drug_name': 'levosimendan',\n",
       "    'symptom_id': 'HP:0011675',\n",
       "    'symptom_name': 'Arrhythmia',\n",
       "    'complete': True}}]}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen_explanations_per_dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Information for Questionnaires"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For dataset 1\n",
      "For explanation explanation_2_run_007\n",
      "ARHGEF6 1320 HGNC:685 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:685\n",
      "neratinib 1772 5252 \n",
      "XYLT1 2112 HGNC:15516 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:15516\n",
      "SPEF2 2661 HGNC:26293 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:26293\n",
      "PRNP 4016 HGNC:9449 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:9449\n",
      "non-syndromic X-linked intellectual disability 4911 MONDO:0019181 http://purl.obolibrary.org/obo/MONDO_0019181\n",
      "LTBP4 5487 HGNC:6717 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6717\n",
      "DAG1 5616 HGNC:2666 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:2666\n",
      "Duchenne muscular dystrophy 6315 MONDO:0010679 http://purl.obolibrary.org/obo/MONDO_0010679\n",
      "LAMA1 6406 HGNC:6481 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6481\n",
      "PAK1 6478 HGNC:8590 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:8590\n",
      "Delayed speech and language development 6575 HP:0000750 http://purl.obolibrary.org/obo/HP_0000750\n",
      "\n",
      "\n",
      "[('ARHGEF6 1320', 'LAMA1 6406', {'label': 'interacts with'}), ('ARHGEF6 1320', 'DAG1 5616', {'label': 'interacts with'}), ('ARHGEF6 1320', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('ARHGEF6 1320', 'non-syndromic X-linked intellectual disability 4911', {'label': 'is causal germline mutation in'}), ('neratinib 1772', 'PAK1 6478', {'label': 'targets'}), ('XYLT1 2112', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('XYLT1 2112', 'Duchenne muscular dystrophy 6315', {'label': 'contributes to condition'}), ('SPEF2 2661', 'Duchenne muscular dystrophy 6315', {'label': 'contributes to condition'}), ('SPEF2 2661', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('PRNP 4016', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('LTBP4 5487', 'Duchenne muscular dystrophy 6315', {'label': 'is causal germline mutation partially giving rise to'}), ('LTBP4 5487', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('DAG1 5616', 'PAK1 6478', {'label': 'interacts with'}), ('DAG1 5616', 'LAMA1 6406', {'label': 'interacts with'}), ('LAMA1 6406', 'PAK1 6478', {'label': 'interacts with'}), ('LAMA1 6406', 'Delayed speech and language development 6575', {'label': 'has phenotype'})]\n",
      "For explanation explanation_1_run_005\n",
      "CDC42 827 HGNC:1736 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:1736\n",
      "ARHGEF6 1320 HGNC:685 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:685\n",
      "neratinib 1772 5252 \n",
      "LAMB2 2722 HGNC:6487 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6487\n",
      "ACTB 5137 HGNC:132 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:132\n",
      "DAG1 5616 HGNC:2666 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:2666\n",
      "HNF4A 6089 HGNC:5024 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:5024\n",
      "Duchenne muscular dystrophy 6315 MONDO:0010679 http://purl.obolibrary.org/obo/MONDO_0010679\n",
      "LAMC3 6381 HGNC:6494 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6494\n",
      "LAMA1 6406 HGNC:6481 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6481\n",
      "PAK1 6478 HGNC:8590 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:8590\n",
      "DMD 6612 HGNC:2928 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:2928\n",
      "PAK3 7165 HGNC:8592 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:8592\n",
      "Global developmental delay 7687 HP:0001263 http://purl.obolibrary.org/obo/HP_0001263\n",
      "\n",
      "\n",
      "[('CDC42 827', 'DMD 6612', {'label': 'interacts with'}), ('CDC42 827', 'ACTB 5137', {'label': 'interacts with'}), ('CDC42 827', 'DAG1 5616', {'label': 'interacts with'}), ('ARHGEF6 1320', 'LAMB2 2722', {'label': 'interacts with'}), ('ARHGEF6 1320', 'DAG1 5616', {'label': 'interacts with'}), ('ARHGEF6 1320', 'PAK3 7165', {'label': 'interacts with'}), ('neratinib 1772', 'PAK1 6478', {'label': 'targets'}), ('DAG1 5616', 'PAK1 6478', {'label': 'interacts with'}), ('DAG1 5616', 'PAK3 7165', {'label': 'interacts with'}), ('HNF4A 6089', 'Global developmental delay 7687', {'label': 'has phenotype'}), ('HNF4A 6089', 'DMD 6612', {'label': 'interacts with'}), ('Duchenne muscular dystrophy 6315', 'DMD 6612', {'label': 'causes condition'}), ('LAMC3 6381', 'DMD 6612', {'label': 'interacts with'}), ('LAMA1 6406', 'Global developmental delay 7687', {'label': 'has phenotype'}), ('PAK1 6478', 'Global developmental delay 7687', {'label': 'has phenotype'}), ('DMD 6612', 'PAK3 7165', {'label': 'interacts with'})]\n",
      "For explanation explanation_2_run_010\n",
      "ARHGEF6 1320 HGNC:685 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:685\n",
      "neratinib 1772 5252 \n",
      "XYLT1 2112 HGNC:15516 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:15516\n",
      "SPEF2 2661 HGNC:26293 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:26293\n",
      "PRNP 4016 HGNC:9449 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:9449\n",
      "non-syndromic X-linked intellectual disability 4911 MONDO:0019181 http://purl.obolibrary.org/obo/MONDO_0019181\n",
      "LTBP4 5487 HGNC:6717 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6717\n",
      "DAG1 5616 HGNC:2666 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:2666\n",
      "Duchenne muscular dystrophy 6315 MONDO:0010679 http://purl.obolibrary.org/obo/MONDO_0010679\n",
      "LAMA1 6406 HGNC:6481 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6481\n",
      "PAK1 6478 HGNC:8590 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:8590\n",
      "Delayed speech and language development 6575 HP:0000750 http://purl.obolibrary.org/obo/HP_0000750\n",
      "\n",
      "\n",
      "[('ARHGEF6 1320', 'LAMA1 6406', {'label': 'interacts with'}), ('ARHGEF6 1320', 'DAG1 5616', {'label': 'interacts with'}), ('ARHGEF6 1320', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('ARHGEF6 1320', 'non-syndromic X-linked intellectual disability 4911', {'label': 'is causal germline mutation in'}), ('neratinib 1772', 'PAK1 6478', {'label': 'targets'}), ('XYLT1 2112', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('XYLT1 2112', 'Duchenne muscular dystrophy 6315', {'label': 'contributes to condition'}), ('SPEF2 2661', 'Duchenne muscular dystrophy 6315', {'label': 'contributes to condition'}), ('SPEF2 2661', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('PRNP 4016', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('LTBP4 5487', 'Duchenne muscular dystrophy 6315', {'label': 'is causal germline mutation partially giving rise to'}), ('LTBP4 5487', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('DAG1 5616', 'PAK1 6478', {'label': 'interacts with'}), ('DAG1 5616', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('Duchenne muscular dystrophy 6315', 'Delayed speech and language development 6575', {'label': 'has phenotype'}), ('LAMA1 6406', 'PAK1 6478', {'label': 'interacts with'})]\n",
      "\n",
      "\n",
      "For dataset 2\n",
      "For explanation explanation_2_run_006\n",
      "cacna1c 290 ZFIN:ZDB-GENE-020129-1 \n",
      "Congestive heart failure 1660 HP:0001635 http://purl.obolibrary.org/obo/HP_0001635\n",
      "MYBPC3 1850 HGNC:7551 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:7551\n",
      "Duchenne muscular dystrophy 1913 MONDO:0010679 http://purl.obolibrary.org/obo/MONDO_0010679\n",
      "MYH6 2835 HGNC:7576 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:7576\n",
      "LAMA2 2961 HGNC:6482 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6482\n",
      "Myh6 3091 MGI:97255 \n",
      "DMD 3310 HGNC:2928 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:2928\n",
      "Tmod1 3672 MGI:98775 \n",
      "aprindine 3697 231 \n",
      "LTBP4 4082 HGNC:6717 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6717\n",
      "Arrhythmia 4336 HP:0011675 http://purl.obolibrary.org/obo/HP_0011675\n",
      "levosimendan 5262 1576 \n",
      "\n",
      "\n",
      "[('cacna1c 290', 'Arrhythmia 4336', {'label': 'causes condition'}), ('Congestive heart failure 1660', 'MYH6 2835', {'label': 'causes condition'}), ('Congestive heart failure 1660', 'Tmod1 3672', {'label': 'causes condition'}), ('Congestive heart failure 1660', 'MYBPC3 1850', {'label': 'causes condition'}), ('Congestive heart failure 1660', 'DMD 3310', {'label': 'causes condition'}), ('Congestive heart failure 1660', 'Duchenne muscular dystrophy 1913', {'label': 'associated with phenotype'}), ('Congestive heart failure 1660', 'levosimendan 5262', {'label': 'is substance that treats'}), ('MYBPC3 1850', 'MYH6 2835', {'label': 'interacts with'}), ('MYBPC3 1850', 'DMD 3310', {'label': 'interacts with'}), ('Duchenne muscular dystrophy 1913', 'LTBP4 4082', {'label': 'contributes to condition'}), ('MYH6 2835', 'Arrhythmia 4336', {'label': 'causes condition'}), ('LAMA2 2961', 'DMD 3310', {'label': 'interacts with'}), ('Myh6 3091', 'Tmod1 3672', {'label': 'interacts with'}), ('Tmod1 3672', 'Arrhythmia 4336', {'label': 'causes condition'}), ('aprindine 3697', 'Arrhythmia 4336', {'label': 'is substance that treats'}), ('LTBP4 4082', 'Arrhythmia 4336', {'label': 'causes condition'})]\n",
      "For explanation explanation_1_run_004\n",
      "TPM3 109 HGNC:12012 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:12012\n",
      "Dmd<mdx>; Dtna<tm1Jrs>/Dtna<tm1Jrs>; Utrn<tm1Jrs>/Utrn<tm1Jrs> [involves: 129S1/Sv * 129X1/SvJ * C57BL/10ScSn] 236 MGI:2176891 \n",
      "Dtna 599 MGI:106039 \n",
      "Utrn<tm1Jrs> 677 MGI:2148539 \n",
      "Dmd<mdx>; Dtna<tm1Jrs>/Dtna<tm1Jrs> [involves: 129X1/SvJ * C57BL/10ScSn] 798 MGI:2176892 \n",
      "Congestive heart failure 1660 HP:0001635 http://purl.obolibrary.org/obo/HP_0001635\n",
      "ACTA1 1769 HGNC:129 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:129\n",
      "MYBPC3 1850 HGNC:7551 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:7551\n",
      "Dmd<mdx>/Dmd<mdx> [D2.B10-Dmd<mdx>/J] 1911 MGI:5697652 \n",
      "Duchenne muscular dystrophy 1913 MONDO:0010679 http://purl.obolibrary.org/obo/MONDO_0010679\n",
      "Dmd 2323 MGI:94909 \n",
      "LAMA2 2961 HGNC:6482 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6482\n",
      "Sgca 3038 MGI:894698 \n",
      "Cardiomyopathy 3222 HP:0001638 http://purl.obolibrary.org/obo/HP_0001638\n",
      "levosimendan 5262 1576 \n",
      "\n",
      "\n",
      "[('TPM3 109', 'Cardiomyopathy 3222', {'label': 'causes condition'}), ('Dmd<mdx>; Dtna<tm1Jrs>/Dtna<tm1Jrs>; Utrn<tm1Jrs>/Utrn<tm1Jrs> [involves: 129S1/Sv * 129X1/SvJ * C57BL/10ScSn] 236', 'Utrn<tm1Jrs> 677', {'label': 'is variant in'}), ('Dmd<mdx>; Dtna<tm1Jrs>/Dtna<tm1Jrs>; Utrn<tm1Jrs>/Utrn<tm1Jrs> [involves: 129S1/Sv * 129X1/SvJ * C57BL/10ScSn] 236', 'Dtna 599', {'label': 'expresses gene'}), ('Dmd<mdx>; Dtna<tm1Jrs>/Dtna<tm1Jrs>; Utrn<tm1Jrs>/Utrn<tm1Jrs> [involves: 129S1/Sv * 129X1/SvJ * C57BL/10ScSn] 236', 'Cardiomyopathy 3222', {'label': 'causes condition'}), ('Dtna 599', 'Sgca 3038', {'label': 'interacts with'}), ('Dtna 599', 'Dmd 2323', {'label': 'interacts with'}), ('Utrn<tm1Jrs> 677', 'Cardiomyopathy 3222', {'label': 'causes condition'}), ('Utrn<tm1Jrs> 677', 'Duchenne muscular dystrophy 1913', {'label': 'has role in modeling'}), ('Dmd<mdx>; Dtna<tm1Jrs>/Dtna<tm1Jrs> [involves: 129X1/SvJ * C57BL/10ScSn] 798', 'Dmd 2323', {'label': 'expresses gene'}), ('Congestive heart failure 1660', 'MYBPC3 1850', {'label': 'causes condition'}), ('Congestive heart failure 1660', 'Duchenne muscular dystrophy 1913', {'label': 'associated with phenotype'}), ('Congestive heart failure 1660', 'levosimendan 5262', {'label': 'is substance that treats'}), ('ACTA1 1769', 'LAMA2 2961', {'label': 'interacts with'}), ('Dmd<mdx>/Dmd<mdx> [D2.B10-Dmd<mdx>/J] 1911', 'Dmd 2323', {'label': 'expresses gene'}), ('LAMA2 2961', 'Cardiomyopathy 3222', {'label': 'causes condition'}), ('Sgca 3038', 'Cardiomyopathy 3222', {'label': 'causes condition'})]\n",
      "For explanation explanation_2_run_009\n",
      "cacna1c 290 ZFIN:ZDB-GENE-020129-1 \n",
      "Congestive heart failure 1660 HP:0001635 http://purl.obolibrary.org/obo/HP_0001635\n",
      "MYBPC3 1850 HGNC:7551 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:7551\n",
      "Duchenne muscular dystrophy 1913 MONDO:0010679 http://purl.obolibrary.org/obo/MONDO_0010679\n",
      "MYH6 2835 HGNC:7576 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:7576\n",
      "LAMA2 2961 HGNC:6482 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6482\n",
      "Myh6 3091 MGI:97255 \n",
      "DMD 3310 HGNC:2928 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:2928\n",
      "aprindine 3697 231 \n",
      "LTBP4 4082 HGNC:6717 https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:6717\n",
      "Arrhythmia 4336 HP:0011675 http://purl.obolibrary.org/obo/HP_0011675\n",
      "levosimendan 5262 1576 \n",
      "\n",
      "\n",
      "[('cacna1c 290', 'Arrhythmia 4336', {'label': 'causes condition'}), ('Congestive heart failure 1660', 'LTBP4 4082', {'label': 'causes condition'}), ('Congestive heart failure 1660', 'levosimendan 5262', {'label': 'is substance that treats'}), ('MYBPC3 1850', 'MYH6 2835', {'label': 'interacts with'}), ('MYBPC3 1850', 'Arrhythmia 4336', {'label': 'causes condition'}), ('MYBPC3 1850', 'DMD 3310', {'label': 'interacts with'}), ('Duchenne muscular dystrophy 1913', 'LTBP4 4082', {'label': 'contributes to condition'}), ('Duchenne muscular dystrophy 1913', 'Arrhythmia 4336', {'label': 'associated with phenotype'}), ('MYH6 2835', 'Myh6 3091', {'label': 'in orthology relationship with'}), ('MYH6 2835', 'Arrhythmia 4336', {'label': 'causes condition'}), ('LAMA2 2961', 'DMD 3310', {'label': 'interacts with'}), ('LAMA2 2961', 'Arrhythmia 4336', {'label': 'causes condition'}), ('Myh6 3091', 'Arrhythmia 4336', {'label': 'causes condition'}), ('DMD 3310', 'Arrhythmia 4336', {'label': 'causes condition'}), ('aprindine 3697', 'Arrhythmia 4336', {'label': 'is substance that treats'}), ('LTBP4 4082', 'Arrhythmia 4336', {'label': 'causes condition'})]\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for dataset_nr in dataset_nrs:\n",
    "    print(f'For dataset {dataset_nr}')\n",
    "    \n",
    "    nodes = pd.read_csv(f'output/indexed_nodes_{dataset_nr}.csv')\n",
    "    \n",
    "    explanations = chosen_explanations_per_dataset[dataset_nr]\n",
    "    for expl in explanations:\n",
    "        print(f\"For explanation {expl['filename']}\")\n",
    "        \n",
    "        for node in expl['G'].nodes:\n",
    "            node_label = node\n",
    "            node_index = node_label.split(' ')[-1]\n",
    "            node_id = nodes.iloc[[node_index]].id.values[0]\n",
    "            \n",
    "            if any(prefix in node_id for prefix in ['HP', 'MONDO']):\n",
    "                url = f\"http://purl.obolibrary.org/obo/{node_id.replace(':', '_')}\"\n",
    "            elif 'HGNC' in node_id:\n",
    "                url = f\"https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/{node_id}\"\n",
    "            else:\n",
    "                url = ''\n",
    "            \n",
    "            print(node_label, node_id, url)\n",
    "            \n",
    "        print('\\n')\n",
    "        \n",
    "        print(expl['G'].edges(data=True))\n",
    "            \n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>pat-2</td>\n",
       "      <td>5</td>\n",
       "      <td>1542</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00006789</td>\n",
       "      <td>unc-54</td>\n",
       "      <td>5</td>\n",
       "      <td>6544</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000015555</td>\n",
       "      <td>LAMC1</td>\n",
       "      <td>5</td>\n",
       "      <td>9268</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ZFIN:ZDB-GENE-021226-3</td>\n",
       "      <td>lamc1</td>\n",
       "      <td>5</td>\n",
       "      <td>5387</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "      <td>5</td>\n",
       "      <td>2204</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85987</th>\n",
       "      <td>458</td>\n",
       "      <td>scopolamine butylbromide</td>\n",
       "      <td>4</td>\n",
       "      <td>5945</td>\n",
       "      <td>targets</td>\n",
       "      <td>P11229</td>\n",
       "      <td>Muscarinic acetylcholine receptor M1</td>\n",
       "      <td>6</td>\n",
       "      <td>5919</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85988</th>\n",
       "      <td>OMIM:300377.0080</td>\n",
       "      <td>DMD, IVS62, A-G, -285</td>\n",
       "      <td>11</td>\n",
       "      <td>1578</td>\n",
       "      <td>is allele of</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>5</td>\n",
       "      <td>3310</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85989</th>\n",
       "      <td>5297</td>\n",
       "      <td>dacomitinib</td>\n",
       "      <td>4</td>\n",
       "      <td>8798</td>\n",
       "      <td>targets</td>\n",
       "      <td>P12931</td>\n",
       "      <td>Proto-oncogene tyrosine-protein kinase Src</td>\n",
       "      <td>6</td>\n",
       "      <td>2379</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85990</th>\n",
       "      <td>ClinVarVariant:981988</td>\n",
       "      <td>NC_000023.11:g.(31875374_31929595)_(31968515_3...</td>\n",
       "      <td>11</td>\n",
       "      <td>8189</td>\n",
       "      <td>has affected feature</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>5</td>\n",
       "      <td>3310</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85991</th>\n",
       "      <td>ClinVarVariant:584105</td>\n",
       "      <td>NC_000023.11:g.(?_32595737)_(32699313_?)del</td>\n",
       "      <td>11</td>\n",
       "      <td>9866</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85992 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          head   \n",
       "0      WormBase:WBGene00006787  \\\n",
       "1      WormBase:WBGene00006787   \n",
       "2      WormBase:WBGene00006787   \n",
       "3      WormBase:WBGene00006787   \n",
       "4      WormBase:WBGene00006787   \n",
       "...                        ...   \n",
       "85987                      458   \n",
       "85988         OMIM:300377.0080   \n",
       "85989                     5297   \n",
       "85990    ClinVarVariant:981988   \n",
       "85991    ClinVarVariant:584105   \n",
       "\n",
       "                                              label_head  class_head   \n",
       "0                                                 unc-52           5  \\\n",
       "1                                                 unc-52           5   \n",
       "2                                                 unc-52           5   \n",
       "3                                                 unc-52           5   \n",
       "4                                                 unc-52           5   \n",
       "...                                                  ...         ...   \n",
       "85987                           scopolamine butylbromide           4   \n",
       "85988                              DMD, IVS62, A-G, -285          11   \n",
       "85989                                        dacomitinib           4   \n",
       "85990  NC_000023.11:g.(31875374_31929595)_(31968515_3...          11   \n",
       "85991        NC_000023.11:g.(?_32595737)_(32699313_?)del          11   \n",
       "\n",
       "       index_head                        relation                        tail   \n",
       "0             304                  interacts with     WormBase:WBGene00003929  \\\n",
       "1             304                  interacts with     WormBase:WBGene00006789   \n",
       "2             304  in orthology relationship with  ENSEMBL:ENSSSCG00000015555   \n",
       "3             304  in orthology relationship with      ZFIN:ZDB-GENE-021226-3   \n",
       "4             304  in orthology relationship with  ENSEMBL:ENSOANG00000001050   \n",
       "...           ...                             ...                         ...   \n",
       "85987        5945                         targets                      P11229   \n",
       "85988        1578                    is allele of                   HGNC:2928   \n",
       "85989        8798                         targets                      P12931   \n",
       "85990        8189            has affected feature                   HGNC:2928   \n",
       "85991        9866                causes condition               MONDO:0010679   \n",
       "\n",
       "                                       label_tail  class_tail  index_tail   \n",
       "0                                           pat-2           5        1542  \\\n",
       "1                                          unc-54           5        6544   \n",
       "2                                           LAMC1           5        9268   \n",
       "3                                           lamc1           5        5387   \n",
       "4                      ENSEMBL:ENSOANG00000001050           5        2204   \n",
       "...                                           ...         ...         ...   \n",
       "85987        Muscarinic acetylcholine receptor M1           6        5919   \n",
       "85988                                         DMD           5        3310   \n",
       "85989  Proto-oncogene tyrosine-protein kinase Src           6        2379   \n",
       "85990                                         DMD           5        3310   \n",
       "85991                 Duchenne muscular dystrophy           3        1913   \n",
       "\n",
       "       type  \n",
       "0         0  \n",
       "1         0  \n",
       "2         1  \n",
       "3         1  \n",
       "4         1  \n",
       "...     ...  \n",
       "85987    17  \n",
       "85988    15  \n",
       "85989    17  \n",
       "85990    11  \n",
       "85991     2  \n",
       "\n",
       "[85992 rows x 10 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = pd.read_csv(f'output/indexed_edges_{dataset_nr}.csv')\n",
    "edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2805</th>\n",
       "      <td>HGNC:11427</td>\n",
       "      <td>STUB1</td>\n",
       "      <td>5</td>\n",
       "      <td>2304</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5708</th>\n",
       "      <td>HGNC:17997</td>\n",
       "      <td>FKRP</td>\n",
       "      <td>5</td>\n",
       "      <td>6985</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12988</th>\n",
       "      <td>HGNC:11138</td>\n",
       "      <td>SNCA</td>\n",
       "      <td>5</td>\n",
       "      <td>9864</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14534</th>\n",
       "      <td>HGNC:7573</td>\n",
       "      <td>MYH3</td>\n",
       "      <td>5</td>\n",
       "      <td>3388</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28110</th>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>5</td>\n",
       "      <td>3310</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32448</th>\n",
       "      <td>HGNC:18028</td>\n",
       "      <td>OSGEP</td>\n",
       "      <td>5</td>\n",
       "      <td>3981</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36920</th>\n",
       "      <td>HGNC:6717</td>\n",
       "      <td>LTBP4</td>\n",
       "      <td>5</td>\n",
       "      <td>4082</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41398</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43273</th>\n",
       "      <td>HGNC:9449</td>\n",
       "      <td>PRNP</td>\n",
       "      <td>5</td>\n",
       "      <td>8674</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44932</th>\n",
       "      <td>HGNC:6481</td>\n",
       "      <td>LAMA1</td>\n",
       "      <td>5</td>\n",
       "      <td>9080</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45284</th>\n",
       "      <td>HGNC:6482</td>\n",
       "      <td>LAMA2</td>\n",
       "      <td>5</td>\n",
       "      <td>2961</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                head                   label_head  class_head  index_head   \n",
       "2805      HGNC:11427                        STUB1           5        2304  \\\n",
       "5708      HGNC:17997                         FKRP           5        6985   \n",
       "12988     HGNC:11138                         SNCA           5        9864   \n",
       "14534      HGNC:7573                         MYH3           5        3388   \n",
       "28110      HGNC:2928                          DMD           5        3310   \n",
       "32448     HGNC:18028                        OSGEP           5        3981   \n",
       "36920      HGNC:6717                        LTBP4           5        4082   \n",
       "41398  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "43273      HGNC:9449                         PRNP           5        8674   \n",
       "44932      HGNC:6481                        LAMA1           5        9080   \n",
       "45284      HGNC:6482                        LAMA2           5        2961   \n",
       "\n",
       "                        relation        tail            label_tail   \n",
       "2805            causes condition  HP:0100543  Cognitive impairment  \\\n",
       "5708            causes condition  HP:0100543  Cognitive impairment   \n",
       "12988           causes condition  HP:0100543  Cognitive impairment   \n",
       "14534           causes condition  HP:0100543  Cognitive impairment   \n",
       "28110           causes condition  HP:0100543  Cognitive impairment   \n",
       "32448           causes condition  HP:0100543  Cognitive impairment   \n",
       "36920           causes condition  HP:0100543  Cognitive impairment   \n",
       "41398  associated with phenotype  HP:0100543  Cognitive impairment   \n",
       "43273           causes condition  HP:0100543  Cognitive impairment   \n",
       "44932           causes condition  HP:0100543  Cognitive impairment   \n",
       "45284           causes condition  HP:0100543  Cognitive impairment   \n",
       "\n",
       "       class_tail  index_tail  type  \n",
       "2805            9        9574     2  \n",
       "5708            9        9574     2  \n",
       "12988           9        9574     2  \n",
       "14534           9        9574     2  \n",
       "28110           9        9574     2  \n",
       "32448           9        9574     2  \n",
       "36920           9        9574     2  \n",
       "41398           9        9574    18  \n",
       "43273           9        9574     2  \n",
       "44932           9        9574     2  \n",
       "45284           9        9574     2  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges.loc[edges['label_tail'] == 'Cognitive impairment']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Number of Types of Edges, Nodes, Metapaths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_occurrence_df(c, label_name):\n",
    "    c_perc = [{label_name: i, 'Percentage': c[i] / c.total() * 100.0, 'Appearances': c[i]} for i in c]\n",
    "    c_df = pd.DataFrame.from_dict(c_perc)\n",
    "    c_df['Percentage'] = c_df['Percentage'].transform(lambda x: '{:,.2f}%'.format(x))\n",
    "    c_df = c_df.sort_values(by = ['Appearances'], ascending = False)\n",
    "    return c_df\n",
    "\n",
    "def count_occurrences(G):    \n",
    "    node_types = []\n",
    "    for n, attr in G.nodes(data=True):\n",
    "        node_types.append(attr['type'])\n",
    "        \n",
    "    edge_types = []\n",
    "    triplets = []\n",
    "    for n1, n2, attr in G.edges(data=True):\n",
    "        edge_type = attr['label']\n",
    "        edge_types.append(edge_type)\n",
    "        \n",
    "        n1_type = G.nodes[n1]['type']\n",
    "        n2_type = G.nodes[n2]['type']\n",
    "        \n",
    "        triplet = tuple([n1_type, edge_type, n2_type])\n",
    "        triplets.append(triplet)\n",
    "    \n",
    "    node_types_df = get_occurrence_df(Counter(node_types), 'Node Type')\n",
    "    \n",
    "    edge_types_df = get_occurrence_df(Counter(edge_types), 'Edge Type')\n",
    "    \n",
    "    metapaths_df = get_occurrence_df(Counter(triplets), 'Metapath')\n",
    "    \n",
    "    return node_types_df, edge_types_df, metapaths_df\n",
    "\n",
    "def get_shortest_path_len_drug_symptom(n1, n2, G):\n",
    "    path_len = nx.shortest_path_length(G, n1, n2)\n",
    "    return path_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the 10 explanations generated from dataset 1\n",
      "Average number of node types: 3.0\n",
      "Average number of edge types: 4.5\n",
      "Average number of triplets: 5.0\n",
      "Average shortest path length between drug and symptom pair: 2.6\n",
      "For the 17 explanations generated from dataset 2\n",
      "Average number of node types: 4.647058823529412\n",
      "Average number of edge types: 5.529411764705882\n",
      "Average number of triplets: 8.058823529411764\n",
      "Average shortest path length between drug and symptom pair: 3.2941176470588234\n"
     ]
    }
   ],
   "source": [
    "for dataset_nr in dataset_nrs:\n",
    "    node_types_per_expl = []\n",
    "    edge_types_per_expl = []\n",
    "    triplet_types_per_expl = []\n",
    "    path_len_per_expl = []\n",
    "    \n",
    "    nr_expl = 0\n",
    "    for expl_g, pair in zip(explanations_per_dataset[dataset_nr], explanation_pairs_per_dataset[dataset_nr]):\n",
    "        node_types_df, edge_types_df, metapaths_df = count_occurrences(expl_g)\n",
    "        \n",
    "        nr_node_types = node_types_df['Node Type'].nunique()\n",
    "        node_types_per_expl.append(nr_node_types)\n",
    "        \n",
    "        nr_edge_types = edge_types_df['Edge Type'].nunique()\n",
    "        edge_types_per_expl.append(nr_edge_types)\n",
    "        \n",
    "        nr_triplet_types = metapaths_df['Metapath'].nunique()\n",
    "        triplet_types_per_expl.append(nr_triplet_types)\n",
    "        \n",
    "        node_1, node_2 = pair\n",
    "        shortest_path_len = get_shortest_path_len_drug_symptom(node_1, node_2, expl_g)\n",
    "        path_len_per_expl.append(shortest_path_len)\n",
    "        \n",
    "        nr_expl += 1\n",
    "    \n",
    "    print(f'For the {nr_expl} explanations generated from dataset {dataset_nr}')\n",
    "    print(f'Average number of node types: {(sum(node_types_per_expl)/len(node_types_per_expl))}')\n",
    "    print(f'Average number of edge types: {(sum(edge_types_per_expl)/len(edge_types_per_expl))}')\n",
    "    print(f'Average number of triplets: {(sum(triplet_types_per_expl)/len(triplet_types_per_expl))}')\n",
    "    print(f'Average shortest path length between drug and symptom pair: {(sum(path_len_per_expl)/len(path_len_per_expl))}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xaifognn",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
