{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This Jupyter notebook contains the GNNExplainer method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import torch\n",
    "\n",
    "import pickle\n",
    "import copy\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from gensim.models import KeyedVectors\n",
    "\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_nr = 1\n",
    "assert dataset_nr == 1 or 2\n",
    "\n",
    "embedding_method = 'e2v'\n",
    "assert embedding_method == 'e2v' or 'm2v'\n",
    "\n",
    "seeded_emb = False\n",
    "\n",
    "if seeded_emb:\n",
    "    fixed_emb = '_seeded'\n",
    "else:\n",
    "    fixed_emb = ''\n",
    "    \n",
    "if dataset_nr == 1 and not seeded_emb:\n",
    "    threshold_nr_runs = '7'\n",
    "else:\n",
    "    threshold_nr_runs = 'all'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder for dataset 1 already exists and will be used: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g1_e2v\n",
      "Output folder for predictions from dataset 1 using method e2v already exists and will be used: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g1_e2v\\expl\n"
     ]
    }
   ],
   "source": [
    "curr_working_dir = os.getcwd()\n",
    "curr_output_dir = os.path.join(curr_working_dir, 'output')\n",
    "dataset_output_dir = os.path.join(curr_output_dir, f'g{dataset_nr}_{embedding_method}{fixed_emb}')\n",
    "\n",
    "if not os.path.exists(dataset_output_dir):\n",
    "    print('First, run the edge2vec embedding and predictor script. Then, run this script.')\n",
    "else:\n",
    "    print(f'Output folder for dataset {dataset_nr} already exists and will be used: {dataset_output_dir}')\n",
    "    \n",
    "run_folders_list = []\n",
    "for item in os.listdir(dataset_output_dir):\n",
    "    curr_path = os.path.join(dataset_output_dir, item)\n",
    "    if os.path.isdir(curr_path) and 'run' in item:\n",
    "        run_folders_list.append(item)\n",
    "        \n",
    "expl_output_dir = os.path.join(dataset_output_dir, 'expl')\n",
    "if not os.path.exists(expl_output_dir):\n",
    "    os.mkdir(expl_output_dir)\n",
    "    print(f'Output folder for predictions from dataset {dataset_nr} using method {embedding_method} is created: {expl_output_dir}')\n",
    "else:\n",
    "    print(f'Output folder for predictions from dataset {dataset_nr} using method {embedding_method} already exists and will be used: {expl_output_dir}')\n",
    "\n",
    "run_folders_paths = []\n",
    "if len(run_folders_list) > 0:\n",
    "    for run_folder in run_folders_list:\n",
    "        latest_run = run_folder\n",
    "        \n",
    "        run_dir = os.path.join(dataset_output_dir, latest_run)\n",
    "        run_folders_paths.append(run_dir)\n",
    "    \n",
    "else:\n",
    "    print('First, run the edge2vec embedding and predictor script. Then, run this script.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load all data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>WormBase:WBGene00000389</td>\n",
       "      <td>ORTH</td>\n",
       "      <td>cdc-25.4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ZP:0018675</td>\n",
       "      <td>DISO</td>\n",
       "      <td>right side lateral plate mesoderm mislocalised...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ZFIN:ZDB-GENE-040426-1197</td>\n",
       "      <td>ORTH</td>\n",
       "      <td>tbc1d5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>DRUG</td>\n",
       "      <td>(S)-nicardipine</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>RGD:3443</td>\n",
       "      <td>ORTH</td>\n",
       "      <td>Ptk2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10029</th>\n",
       "      <td>10029</td>\n",
       "      <td>MP:0009763</td>\n",
       "      <td>DISO</td>\n",
       "      <td>increased sensitivity to induced morbidity/mor...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10030</th>\n",
       "      <td>10030</td>\n",
       "      <td>MP:0011057</td>\n",
       "      <td>DISO</td>\n",
       "      <td>absent brain ependyma motile cilia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10031</th>\n",
       "      <td>10031</td>\n",
       "      <td>MP:0001412</td>\n",
       "      <td>DISO</td>\n",
       "      <td>excessive scratching</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10032</th>\n",
       "      <td>10032</td>\n",
       "      <td>WBPhenotype:0004023</td>\n",
       "      <td>DISO</td>\n",
       "      <td>frequency of body bend variant</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10033</th>\n",
       "      <td>10033</td>\n",
       "      <td>ClinVarVariant:858516</td>\n",
       "      <td>VARI</td>\n",
       "      <td>NM_004006.3(DMD):c.9852G&gt;A (p.Trp3284Ter)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10034 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index_id                         id semantic  \\\n",
       "0             0    WormBase:WBGene00000389     ORTH   \n",
       "1             1                 ZP:0018675     DISO   \n",
       "2             2  ZFIN:ZDB-GENE-040426-1197     ORTH   \n",
       "3             3                          5     DRUG   \n",
       "4             4                   RGD:3443     ORTH   \n",
       "...         ...                        ...      ...   \n",
       "10029     10029                 MP:0009763     DISO   \n",
       "10030     10030                 MP:0011057     DISO   \n",
       "10031     10031                 MP:0001412     DISO   \n",
       "10032     10032        WBPhenotype:0004023     DISO   \n",
       "10033     10033      ClinVarVariant:858516     VARI   \n",
       "\n",
       "                                                   label  semantic_id  \n",
       "0                                               cdc-25.4            5  \n",
       "1      right side lateral plate mesoderm mislocalised...            1  \n",
       "2                                                 tbc1d5            5  \n",
       "3                                        (S)-nicardipine            2  \n",
       "4                                                   Ptk2            5  \n",
       "...                                                  ...          ...  \n",
       "10029  increased sensitivity to induced morbidity/mor...            1  \n",
       "10030                 absent brain ependyma motile cilia            1  \n",
       "10031                               excessive scratching            1  \n",
       "10032                     frequency of body bend variant            1  \n",
       "10033          NM_004006.3(DMD):c.9852G>A (p.Trp3284Ter)            7  \n",
       "\n",
       "[10034 rows x 5 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = pd.read_csv(f'output/indexed_nodes_{dataset_nr}.csv')\n",
    "nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "index_id            5487\n",
       "id             HGNC:6717\n",
       "semantic            GENE\n",
       "label              LTBP4\n",
       "semantic_id            3\n",
       "Name: 5487, dtype: object"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes.iloc[5487]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{5: 'ORTH',\n",
       " 1: 'DISO',\n",
       " 2: 'DRUG',\n",
       " 4: 'GENO',\n",
       " 7: 'VARI',\n",
       " 3: 'GENE',\n",
       " 0: 'ANAT',\n",
       " 6: 'PHYS'}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_labels_dict = nodes[['semantic_id', 'semantic']].drop_duplicates().set_index('semantic_id').to_dict()['semantic']\n",
    "node_labels_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZFIN:ZDB-GENE-050626-112</td>\n",
       "      <td>myl4</td>\n",
       "      <td>5</td>\n",
       "      <td>5279</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:FBgn0085464</td>\n",
       "      <td>CG34435</td>\n",
       "      <td>5</td>\n",
       "      <td>6825</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZFIN:ZDB-GENE-050626-112</td>\n",
       "      <td>myl4</td>\n",
       "      <td>5</td>\n",
       "      <td>5279</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>HGNC:7585</td>\n",
       "      <td>MYL4</td>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZFIN:ZDB-GENE-050626-112</td>\n",
       "      <td>myl4</td>\n",
       "      <td>5</td>\n",
       "      <td>5279</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:FBgn0002772</td>\n",
       "      <td>Mlc1</td>\n",
       "      <td>5</td>\n",
       "      <td>8901</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZFIN:ZDB-GENE-050626-112</td>\n",
       "      <td>myl4</td>\n",
       "      <td>5</td>\n",
       "      <td>5279</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>NCBIGene:396472</td>\n",
       "      <td>MYL4</td>\n",
       "      <td>3</td>\n",
       "      <td>9508</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZFIN:ZDB-GENE-050626-112</td>\n",
       "      <td>myl4</td>\n",
       "      <td>5</td>\n",
       "      <td>5279</td>\n",
       "      <td>in 1 to 1 orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSECAG00000020967</td>\n",
       "      <td>ENSEMBL:ENSECAG00000020967</td>\n",
       "      <td>5</td>\n",
       "      <td>8807</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82908</th>\n",
       "      <td>4810</td>\n",
       "      <td>ibrutinib</td>\n",
       "      <td>2</td>\n",
       "      <td>1618</td>\n",
       "      <td>targets</td>\n",
       "      <td>HGNC:11283</td>\n",
       "      <td>SRC</td>\n",
       "      <td>3</td>\n",
       "      <td>3279</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82909</th>\n",
       "      <td>522</td>\n",
       "      <td>carvedilol</td>\n",
       "      <td>2</td>\n",
       "      <td>184</td>\n",
       "      <td>targets</td>\n",
       "      <td>HGNC:620</td>\n",
       "      <td>APP</td>\n",
       "      <td>3</td>\n",
       "      <td>547</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82910</th>\n",
       "      <td>OMIM:300377.0013</td>\n",
       "      <td>DMD, EX18DEL</td>\n",
       "      <td>1</td>\n",
       "      <td>2822</td>\n",
       "      <td>is allele of</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>3</td>\n",
       "      <td>6612</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82911</th>\n",
       "      <td>Coriell:GM05113</td>\n",
       "      <td>NIGMS-GM05113</td>\n",
       "      <td>4</td>\n",
       "      <td>8105</td>\n",
       "      <td>has role in modeling</td>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>1</td>\n",
       "      <td>6315</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82912</th>\n",
       "      <td>OMIM:300377.0039</td>\n",
       "      <td>DMD, 1-BP INS, NT1554</td>\n",
       "      <td>1</td>\n",
       "      <td>7364</td>\n",
       "      <td>is allele of</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>3</td>\n",
       "      <td>6612</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82913 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           head             label_head  class_head  \\\n",
       "0      ZFIN:ZDB-GENE-050626-112                   myl4           5   \n",
       "1      ZFIN:ZDB-GENE-050626-112                   myl4           5   \n",
       "2      ZFIN:ZDB-GENE-050626-112                   myl4           5   \n",
       "3      ZFIN:ZDB-GENE-050626-112                   myl4           5   \n",
       "4      ZFIN:ZDB-GENE-050626-112                   myl4           5   \n",
       "...                         ...                    ...         ...   \n",
       "82908                      4810              ibrutinib           2   \n",
       "82909                       522             carvedilol           2   \n",
       "82910          OMIM:300377.0013           DMD, EX18DEL           1   \n",
       "82911           Coriell:GM05113          NIGMS-GM05113           4   \n",
       "82912          OMIM:300377.0039  DMD, 1-BP INS, NT1554           1   \n",
       "\n",
       "       index_head                               relation  \\\n",
       "0            5279         in orthology relationship with   \n",
       "1            5279         in orthology relationship with   \n",
       "2            5279         in orthology relationship with   \n",
       "3            5279         in orthology relationship with   \n",
       "4            5279  in 1 to 1 orthology relationship with   \n",
       "...           ...                                    ...   \n",
       "82908        1618                                targets   \n",
       "82909         184                                targets   \n",
       "82910        2822                           is allele of   \n",
       "82911        8105                   has role in modeling   \n",
       "82912        7364                           is allele of   \n",
       "\n",
       "                             tail                   label_tail  class_tail  \\\n",
       "0             FlyBase:FBgn0085464                      CG34435           5   \n",
       "1                       HGNC:7585                         MYL4           3   \n",
       "2             FlyBase:FBgn0002772                         Mlc1           5   \n",
       "3                 NCBIGene:396472                         MYL4           3   \n",
       "4      ENSEMBL:ENSECAG00000020967   ENSEMBL:ENSECAG00000020967           5   \n",
       "...                           ...                          ...         ...   \n",
       "82908                  HGNC:11283                          SRC           3   \n",
       "82909                    HGNC:620                          APP           3   \n",
       "82910                   HGNC:2928                          DMD           3   \n",
       "82911               MONDO:0010679  Duchenne muscular dystrophy           1   \n",
       "82912                   HGNC:2928                          DMD           3   \n",
       "\n",
       "       index_tail  type  \n",
       "0            6825     0  \n",
       "1              27     0  \n",
       "2            8901     0  \n",
       "3            9508     0  \n",
       "4            8807     1  \n",
       "...           ...   ...  \n",
       "82908        3279    14  \n",
       "82909         547    14  \n",
       "82910        6612    16  \n",
       "82911        6315    15  \n",
       "82912        6612    16  \n",
       "\n",
       "[82913 rows x 10 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = pd.read_csv(f'output/indexed_edges_{dataset_nr}.csv')\n",
    "edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def loadEdge2VecEmbedding(current_run_dir): \n",
    "    node_feat = KeyedVectors.load(f'{current_run_dir}/w2v_{dataset_nr}.dvectors', mmap='r')\n",
    "    e2v_embedding = pd.DataFrame(columns = ['Node', 'Embedding'])\n",
    "    for idx, key in enumerate(node_feat.index_to_key):\n",
    "        e2v_embedding.loc[int(key)] = pd.Series({'Node':int(key), 'Embedding':list(node_feat[key])})\n",
    "        \n",
    "    return e2v_embedding.sort_values('Node')\n",
    "\n",
    "def loadMetapath2VecEmbedding(current_run_dir):\n",
    "    metapath2vec_embedding = pd.read_csv(f'{current_run_dir}/metapath2vec_embedding_{dataset_nr}.csv')\n",
    "    metapath2vec_embedding['Embedding'] = metapath2vec_embedding.values.tolist()\n",
    "    metapath2vec_embedding = metapath2vec_embedding[['Embedding']]\n",
    "    metapath2vec_embedding['Node'] = metapath2vec_embedding.index\n",
    "    return metapath2vec_embedding[['Node', 'Embedding']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_graph(embedding):\n",
    "    G = nx.Graph() # TODO: DiGraph?\n",
    "    for ind, node in embedding.iterrows(): \n",
    "        G.add_node(int(node['Node']), node_feature=torch.Tensor(node['Embedding']))\n",
    "    for ind, edge in edges.iterrows(): \n",
    "        G.add_edge(int(edge['index_head']), int(edge['index_tail']), edge_label = edge['relation'])\n",
    "    \n",
    "    for node in G.nodes(data=True):\n",
    "        print(node)\n",
    "        break\n",
    "    for edge in G.edges(data=True):\n",
    "        print(edge)\n",
    "        break\n",
    "\n",
    "    print(\"Number of edges is {}\".format(G.number_of_edges()))\n",
    "    print(\"Number of nodes is {}\".format(G.number_of_nodes()))\n",
    "    \n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gnn.linkpred_model import LinkPredModel\n",
    "\n",
    "def load_trained_model(current_run_dir, embedding):\n",
    "    with open(f'{current_run_dir}/best_model_{dataset_nr}_{embedding_method}_args.pkl', 'rb') as f:\n",
    "        loaded_args = pickle.load(f)\n",
    "        \n",
    "    best_model = LinkPredModel(len(embedding['Embedding'][0]), \n",
    "                            loaded_args['hidden_dim'], loaded_args['output_dim'], \n",
    "                            loaded_args['layers'], loaded_args['aggr'], \n",
    "                            loaded_args['dropout'], loaded_args['device']).to(loaded_args['device'])\n",
    "    best_model.load_state_dict(torch.load(f'{run_dir}/best_model_{dataset_nr}_{embedding_method}.pth'))\n",
    "    \n",
    "    return best_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explain predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gnn.gnnexplainer import GNNExplainer, visualize_subgraph\n",
    "\n",
    "num_hops = 1\n",
    "\n",
    "def explain_edge(node_idx1, node_idx2, model, x, edge_index, G):\n",
    "    explainer = GNNExplainer(model,\n",
    "                             epochs=700, num_hops=num_hops, lr=0.01)\n",
    "    \n",
    "    trigger = False\n",
    "    early_stop = 0\n",
    "    size = 20   # change size of the explanation graph TODO: 15\n",
    "    iterations = 150 # number of times GNNExplainer is executed\n",
    "    \n",
    "    while not trigger:\n",
    "        _, edge_mask = explainer.explain_link(node_idx1=node_idx1, node_idx2=node_idx2,\n",
    "                                              x=x, edge_index=edge_index,\n",
    "                                              G=G)\n",
    "        \n",
    "        G2 = copy.deepcopy(G)\n",
    "        \n",
    "        if size is not None:\n",
    "            limit = edge_mask.sort(descending = True)[0][size]\n",
    "            print('Contribution threshold is', limit)\n",
    "        else:\n",
    "            limit = 0.5\n",
    "        \n",
    "        for indx, edge in enumerate(G.edges): \n",
    "            if edge_mask[indx] < limit:\n",
    "                G2.remove_edge(edge[0], edge[1])\n",
    "        \n",
    "        if nx.has_path(G2, node_idx1, node_idx2) and torch.is_nonzero(limit):\n",
    "            trigger = True\n",
    "            \n",
    "        early_stop += 1\n",
    "        \n",
    "        if early_stop == iterations and not trigger:\n",
    "            print('No good explanation found after {} iterations'.format(early_stop))\n",
    "            trigger = True\n",
    "        elif trigger: \n",
    "            print('Explanation found!')\n",
    "            return edge_mask, limit\n",
    "        \n",
    "    return None, None\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'HP:0011675': ['231', '1576'],\n",
       " 'HP:0001638': ['1576'],\n",
       " 'HP:0001635': ['231', '1576'],\n",
       " 'HP:0002791': ['5345'],\n",
       " 'HP:0003236': ['1576'],\n",
       " 'HP:0003323': ['1576'],\n",
       " 'HP:0003307': ['1576'],\n",
       " 'HP:0001644': ['1576'],\n",
       " 'HP:0002650': ['5252']}"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(f'{dataset_output_dir}/symptom_drug_pair_overlapping_{threshold_nr_runs}_runs_{dataset_nr}_{embedding_method}{fixed_emb}.pkl', 'rb') as f:\n",
    "    overlapping_symptom_drug_pairs = pickle.load(f)\n",
    "\n",
    "overlapping_symptoms_drugs = {}\n",
    "\n",
    "for overlapping_pair in overlapping_symptom_drug_pairs:\n",
    "    symptom = overlapping_pair[0]\n",
    "    drug = overlapping_pair[1]\n",
    "    \n",
    "    if symptom in overlapping_symptoms_drugs:\n",
    "        overlapping_symptoms_drugs[symptom].append(drug)\n",
    "    else:\n",
    "        overlapping_symptoms_drugs[symptom] = [drug]\n",
    "        \n",
    "overlapping_symptoms_drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_occurrence_df(c, label_name):\n",
    "    c_perc = [{label_name: i, 'Percentage': c[i] / c.total() * 100.0, 'Appearances': c[i]} for i in c]\n",
    "    c_df = pd.DataFrame.from_dict(c_perc)\n",
    "    c_df['Percentage'] = c_df['Percentage'].transform(lambda x: '{:,.2f}%'.format(x))\n",
    "    c_df = c_df.sort_values(by = ['Appearances'], ascending = False)\n",
    "    return c_df\n",
    "\n",
    "def count_occurrences(G):    \n",
    "    node_types = []\n",
    "    for n, attr in G.nodes(data=True):\n",
    "        node_label = attr['y2']\n",
    "        node_types.append(node_labels_dict[node_label])\n",
    "    \n",
    "    edge_types = []\n",
    "    metapath_triplets = []\n",
    "    for n1, n2, attr in G.edges(data=True):\n",
    "        edge_label = attr['label']\n",
    "        edge_types.append(edge_label)\n",
    "        \n",
    "        n1_class = G.nodes[n1]['y2']\n",
    "        n2_class = G.nodes[n2]['y2']\n",
    "        \n",
    "        metapath = tuple([node_labels_dict[n1_class], edge_label, node_labels_dict[n2_class]])\n",
    "        metapath_triplets.append(metapath)\n",
    "        \n",
    "     \n",
    "    node_types_df = get_occurrence_df(Counter(node_types), 'Node Type')\n",
    "    print(node_types_df)\n",
    "    \n",
    "    edge_types_df = get_occurrence_df(Counter(edge_types), 'Edge Type')\n",
    "    print(edge_types_df)\n",
    "    \n",
    "    metapaths_df = get_occurrence_df(Counter(metapath_triplets), 'Metapath')\n",
    "    print(metapaths_df)\n",
    "    \n",
    "    return node_types_df, edge_types_df, metapaths_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_graph_to_save(G):\n",
    "    G_save = copy.deepcopy(G)\n",
    "\n",
    "    for node in G_save.nodes(data=True):\n",
    "        key = node[1]['y2']\n",
    "        node[1]['type'] = node_labels_dict[key]\n",
    "        \n",
    "    for node in G_save.nodes():\n",
    "        G_save.nodes[node].pop('y')\n",
    "        G_save.nodes[node].pop('y2')\n",
    "        \n",
    "    for edge in G_save.edges():\n",
    "        G_save.edges[edge].pop('att')\n",
    "        G_save.edges[edge].pop('edge_color')\n",
    "\n",
    "    for node in G_save.nodes(data=True):\n",
    "        print(node)\n",
    "    for edge in G_save.edges(data=True):\n",
    "        print(edge)\n",
    "        \n",
    "    return G_save\n",
    "\n",
    "def visualize_explanation(nr, explanation, edge_index, edge_labels_dict):\n",
    "    nodes_idxs = torch.Tensor([explanation['node_idx1'], explanation['node_idx2']]).long()\n",
    "    \n",
    "    drug_id = explanation['drug']\n",
    "    drug_name = nodes.loc[nodes['id'] == drug_id]['label'].iloc[0]\n",
    "    \n",
    "    symptom_id = explanation['symptom']\n",
    "    symptom_name = nodes.loc[nodes['id'] == symptom_id]['label'].iloc[0]\n",
    "    \n",
    "    run_name = explanation['run']\n",
    "    \n",
    "    explanation_title = f\"Explanation for link between {drug_name} ({drug_id}) and {symptom_name} ({symptom_id}) from run {run_name}\"\n",
    "\n",
    "    plt.figure(figsize=(10, 10))\n",
    "    plt.title(explanation_title)\n",
    "    ax, G_sub = visualize_subgraph(nodes_idxs, edge_index, explanation['found_edge_mask'], \n",
    "                                   nodes=nodes, node_labels_dict=node_labels_dict, y=torch.Tensor(nodes.semantic_id), \n",
    "                                   seed=667, num_hops=num_hops, \n",
    "                                   threshold=explanation['found_limit'], \n",
    "                                   node_label='label', edge_labels=edge_labels_dict, \n",
    "                                   show_inactive=False, remove_unconnected=True)\n",
    "    \n",
    "    plt.savefig(f'{expl_output_dir}/explanation_{nr}_{run_name}.png', bbox_inches='tight')\n",
    "    plt.show()\n",
    "    \n",
    "    # See https://networkx.org/documentation/stable/release/migration_guide_from_2.x_to_3.0.html#deprecated-code\n",
    "    with open(f'{expl_output_dir}/explanation_{nr}_{run_name}_graph.gpickle', 'wb') as f:\n",
    "        pickle.dump(get_graph_to_save(G_sub), f, pickle.HIGHEST_PROTOCOL)\n",
    "        \n",
    "    print(f'Explanation graph saved at {expl_output_dir}/explanation_{nr}_{run_name}_graph.gpickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_node_idx(id, nodes): \n",
    "    return nodes[nodes['id'] == id].index.values.astype(int)[0]\n",
    "\n",
    "def is_overlapping_pair(symptom_id, drug_id):\n",
    "    if symptom_id in overlapping_symptoms_drugs:\n",
    "        if drug_id in overlapping_symptoms_drugs[symptom_id]:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "found_explations_list = []\n",
    "\n",
    "for run_dir in run_folders_paths:\n",
    "    print(run_dir)\n",
    "    run_name = run_dir.split('\\\\')[-1]\n",
    "    \n",
    "    symptoms_drugs = pd.read_pickle(f'{run_dir}/pred/candidates_per_symptom_{dataset_nr}_{embedding_method}.pkl')   \n",
    "    \n",
    "    if embedding_method == 'e2v':\n",
    "        embedding = loadEdge2VecEmbedding(current_run_dir=run_dir)\n",
    "    else:\n",
    "        embedding = loadMetapath2VecEmbedding(current_run_dir=run_dir)\n",
    "        \n",
    "    G = create_graph(embedding=embedding)\n",
    "    edge_labels_dict = dict([((n1, n2), G.edges[(n1,n2)]['edge_label']) for n1, n2 in G.edges])\n",
    "    \n",
    "    best_model = load_trained_model(current_run_dir=run_dir, embedding=embedding)\n",
    "    \n",
    "    x = torch.Tensor(embedding['Embedding'])\n",
    "    edge_index = torch.Tensor(np.array(G.edges).transpose()).type(torch.int64).long()\n",
    "    \n",
    "    explanation_nr = 0\n",
    "    \n",
    "    for ind, edge in symptoms_drugs.iterrows(): \n",
    "        symptom_id = edge['Symptom']\n",
    "        for drug_id in edge['Candidates']:\n",
    "            if is_overlapping_pair(symptom_id=symptom_id, drug_id=drug_id):\n",
    "            \n",
    "                node_idx1 = get_node_idx(symptom_id, nodes)\n",
    "                print('node1:')\n",
    "                print(nodes.loc[[node_idx1]])\n",
    "\n",
    "                node_idx2 = get_node_idx(str(drug_id), nodes)\n",
    "                print('node2:')\n",
    "                print(nodes.loc[[node_idx2]])\n",
    "                \n",
    "                found_edge_mask, found_limit = explain_edge(node_idx1=node_idx1, node_idx2=node_idx2,\n",
    "                                                            model=best_model, x=x, edge_index=edge_index, G=G)\n",
    "                \n",
    "                if found_edge_mask is not None and found_limit is not None:\n",
    "                    explanation_nr += 1\n",
    "                    \n",
    "                    found_explanation= {'symptom': symptom_id, 'drug': drug_id, \n",
    "                                        'node_idx1': node_idx1, 'node_idx2': node_idx2, \n",
    "                                        'found_edge_mask': found_edge_mask, 'found_limit': found_limit, \n",
    "                                        'run': run_name}\n",
    "                    \n",
    "                    found_explations_list.append(found_explanation)\n",
    "                    \n",
    "                    visualize_explanation(explanation_nr, found_explanation, edge_index, edge_labels_dict)\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xaifo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
