{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import pickle\n",
    "import copy\n",
    "\n",
    "from gensim.models import KeyedVectors\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "\n",
    "from deepsnap.graph import Graph\n",
    "\n",
    "from torch_geometric.nn import SAGEConv\n",
    "from torch_geometric.explain import Explainer, GNNExplainer, ModelConfig\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Necessary Data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the node features, load the embeddings yielded from Edge2Vec."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Node</th>\n",
       "      <th>Embedding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>[0.025715737, -0.21057369, 0.03792795, 0.25484...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>[0.18812153, -0.1237438, -0.007631695, 0.13901...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>[0.56500244, 0.06161292, 0.7965068, 0.28490803...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>[0.40217963, -0.4195659, -0.0022494257, 0.4107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>[0.18838538, -0.45605302, 0.12164134, 0.125727...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10232</th>\n",
       "      <td>10232</td>\n",
       "      <td>[0.52268314, -0.26845437, -0.17985809, 0.13895...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10233</th>\n",
       "      <td>10233</td>\n",
       "      <td>[0.44407392, -0.21608466, 0.30382442, 0.208148...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10234</th>\n",
       "      <td>10234</td>\n",
       "      <td>[0.33063313, -0.20261967, 0.14750157, 0.319578...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10235</th>\n",
       "      <td>10235</td>\n",
       "      <td>[0.3033455, -0.19828539, -0.02041113, 0.050759...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10236</th>\n",
       "      <td>10236</td>\n",
       "      <td>[0.7833079, -0.07569242, 0.527525, 0.4939633, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10237 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Node                                          Embedding\n",
       "0          0  [0.025715737, -0.21057369, 0.03792795, 0.25484...\n",
       "1          1  [0.18812153, -0.1237438, -0.007631695, 0.13901...\n",
       "2          2  [0.56500244, 0.06161292, 0.7965068, 0.28490803...\n",
       "3          3  [0.40217963, -0.4195659, -0.0022494257, 0.4107...\n",
       "4          4  [0.18838538, -0.45605302, 0.12164134, 0.125727...\n",
       "...      ...                                                ...\n",
       "10232  10232  [0.52268314, -0.26845437, -0.17985809, 0.13895...\n",
       "10233  10233  [0.44407392, -0.21608466, 0.30382442, 0.208148...\n",
       "10234  10234  [0.33063313, -0.20261967, 0.14750157, 0.319578...\n",
       "10235  10235  [0.3033455, -0.19828539, -0.02041113, 0.050759...\n",
       "10236  10236  [0.7833079, -0.07569242, 0.527525, 0.4939633, ...\n",
       "\n",
       "[10237 rows x 2 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loaded_node_embedding = KeyedVectors.load('output/w2v.dvectors', mmap='r')\n",
    "\n",
    "node_embeddings = pd.DataFrame(columns = ['Node', 'Embedding'])\n",
    "for idx, key in enumerate(loaded_node_embedding.index_to_key):\n",
    "    node_embeddings.loc[int(key)] = pd.Series({'Node':int(key), 'Embedding':list(loaded_node_embedding[key])})\n",
    "    \n",
    "node_features = node_embeddings.sort_values('Node')\n",
    "\n",
    "node_features"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For additional information of nodes, load csv file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>MP:0020358</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>abnormal inhibitory synapse morphology</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>HP:0000670</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>Carious teeth</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000017522</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000017522</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ZFIN:ZDB-GENE-070705-188</td>\n",
       "      <td>gene</td>\n",
       "      <td>chrm1b</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>HP:0000544</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>External ophthalmoplegia</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10232</th>\n",
       "      <td>10232</td>\n",
       "      <td>MGI:2176879</td>\n",
       "      <td>genotype</td>\n",
       "      <td>Dmd&lt;mdx&gt;/Dmd&lt;mdx&gt;; Utrn&lt;tm1Ked&gt;/Utrn&lt;tm1Ked&gt; [...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10233</th>\n",
       "      <td>10233</td>\n",
       "      <td>ZP:0008070</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>skeletal muscle cell dystrophic, abnormal</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10234</th>\n",
       "      <td>10234</td>\n",
       "      <td>ClinVarVariant:577742</td>\n",
       "      <td>variant</td>\n",
       "      <td>NM_004006.2(DMD):c.8655C&gt;A (p.Tyr2885Ter)</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10235</th>\n",
       "      <td>10235</td>\n",
       "      <td>MP:0004087</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>abnormal muscle fiber morphology</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10236</th>\n",
       "      <td>10236</td>\n",
       "      <td>ENSEMBL:ENSGALG00000005079</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSGALG00000005079</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10237 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index_id                          id   semantic   \n",
       "0             0                  MP:0020358  phenotype  \\\n",
       "1             1                  HP:0000670  phenotype   \n",
       "2             2  ENSEMBL:ENSCAFG00000017522       gene   \n",
       "3             3    ZFIN:ZDB-GENE-070705-188       gene   \n",
       "4             4                  HP:0000544  phenotype   \n",
       "...         ...                         ...        ...   \n",
       "10232     10232                 MGI:2176879   genotype   \n",
       "10233     10233                  ZP:0008070  phenotype   \n",
       "10234     10234       ClinVarVariant:577742    variant   \n",
       "10235     10235                  MP:0004087  phenotype   \n",
       "10236     10236  ENSEMBL:ENSGALG00000005079       gene   \n",
       "\n",
       "                                                   label  semantic_id  \n",
       "0                 abnormal inhibitory synapse morphology            8  \n",
       "1                                          Carious teeth            8  \n",
       "2                             ENSEMBL:ENSCAFG00000017522            5  \n",
       "3                                                 chrm1b            5  \n",
       "4                               External ophthalmoplegia            8  \n",
       "...                                                  ...          ...  \n",
       "10232  Dmd<mdx>/Dmd<mdx>; Utrn<tm1Ked>/Utrn<tm1Ked> [...            6  \n",
       "10233          skeletal muscle cell dystrophic, abnormal            8  \n",
       "10234          NM_004006.2(DMD):c.8655C>A (p.Tyr2885Ter)           10  \n",
       "10235                   abnormal muscle fiber morphology            8  \n",
       "10236                         ENSEMBL:ENSGALG00000005079            5  \n",
       "\n",
       "[10237 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = pd.read_csv('output/indexed_nodes.csv')\n",
    "nodes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the edges, load csv file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FlyBase:FBgn0011676</td>\n",
       "      <td>Nos</td>\n",
       "      <td>5</td>\n",
       "      <td>9594</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000009820</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000009820</td>\n",
       "      <td>5</td>\n",
       "      <td>7362</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:12012</td>\n",
       "      <td>TPM3</td>\n",
       "      <td>5</td>\n",
       "      <td>1813</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>HGNC:11946</td>\n",
       "      <td>TNNI2</td>\n",
       "      <td>5</td>\n",
       "      <td>7360</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:15516</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>5</td>\n",
       "      <td>3228</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0002650</td>\n",
       "      <td>Scoliosis</td>\n",
       "      <td>8</td>\n",
       "      <td>2102</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WormBase:WBGene00000065</td>\n",
       "      <td>act-3</td>\n",
       "      <td>5</td>\n",
       "      <td>3272</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSOANG00000007850</td>\n",
       "      <td>ENSEMBL:ENSOANG00000007850</td>\n",
       "      <td>5</td>\n",
       "      <td>2284</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:6485</td>\n",
       "      <td>LAMA5</td>\n",
       "      <td>5</td>\n",
       "      <td>4641</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>HGNC:329</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>5</td>\n",
       "      <td>2205</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85949</th>\n",
       "      <td>dictyBase:DDB_G0276459</td>\n",
       "      <td>pakB</td>\n",
       "      <td>5</td>\n",
       "      <td>3746</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>WormBase:WBGene00003911</td>\n",
       "      <td>pak-1</td>\n",
       "      <td>5</td>\n",
       "      <td>6906</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85950</th>\n",
       "      <td>ZFIN:ZDB-GENE-030113-2</td>\n",
       "      <td>ttn.2</td>\n",
       "      <td>5</td>\n",
       "      <td>6355</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>ZP:0106692</td>\n",
       "      <td>Z disc myocardium disorganized, abnormal</td>\n",
       "      <td>8</td>\n",
       "      <td>6931</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85951</th>\n",
       "      <td>ENSEMBL:ENSGALG00000015211</td>\n",
       "      <td>DTNA</td>\n",
       "      <td>5</td>\n",
       "      <td>828</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSECAG00000020912</td>\n",
       "      <td>ENSEMBL:ENSECAG00000020912</td>\n",
       "      <td>5</td>\n",
       "      <td>6757</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85952</th>\n",
       "      <td>ENSEMBL:ENSECAG00000000207</td>\n",
       "      <td>ENSEMBL:ENSECAG00000000207</td>\n",
       "      <td>5</td>\n",
       "      <td>8915</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>WormBase:WBGene00000064</td>\n",
       "      <td>act-2</td>\n",
       "      <td>5</td>\n",
       "      <td>5490</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85953</th>\n",
       "      <td>WormBase:WBGene00010776</td>\n",
       "      <td>pix-1</td>\n",
       "      <td>5</td>\n",
       "      <td>4984</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSMMUG00000013138</td>\n",
       "      <td>ENSEMBL:ENSMMUG00000013138</td>\n",
       "      <td>5</td>\n",
       "      <td>1743</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85954 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             head                  label_head  class_head   \n",
       "0             FlyBase:FBgn0011676                         Nos           5  \\\n",
       "1                      HGNC:12012                        TPM3           5   \n",
       "2                      HGNC:15516                       XYLT1           5   \n",
       "3         WormBase:WBGene00000065                       act-3           5   \n",
       "4                       HGNC:6485                       LAMA5           5   \n",
       "...                           ...                         ...         ...   \n",
       "85949      dictyBase:DDB_G0276459                        pakB           5   \n",
       "85950      ZFIN:ZDB-GENE-030113-2                       ttn.2           5   \n",
       "85951  ENSEMBL:ENSGALG00000015211                        DTNA           5   \n",
       "85952  ENSEMBL:ENSECAG00000000207  ENSEMBL:ENSECAG00000000207           5   \n",
       "85953     WormBase:WBGene00010776                       pix-1           5   \n",
       "\n",
       "       index_head                        relation                        tail   \n",
       "0            9594  in orthology relationship with  ENSEMBL:ENSCAFG00000009820  \\\n",
       "1            1813                  interacts with                  HGNC:11946   \n",
       "2            3228                causes condition                  HP:0002650   \n",
       "3            3272  in orthology relationship with  ENSEMBL:ENSOANG00000007850   \n",
       "4            4641                  interacts with                    HGNC:329   \n",
       "...           ...                             ...                         ...   \n",
       "85949        3746  in orthology relationship with     WormBase:WBGene00003911   \n",
       "85950        6355                causes condition                  ZP:0106692   \n",
       "85951         828  in orthology relationship with  ENSEMBL:ENSECAG00000020912   \n",
       "85952        8915  in orthology relationship with     WormBase:WBGene00000064   \n",
       "85953        4984  in orthology relationship with  ENSEMBL:ENSMMUG00000013138   \n",
       "\n",
       "                                     label_tail  class_tail  index_tail  type  \n",
       "0                    ENSEMBL:ENSCAFG00000009820           5        7362     0  \n",
       "1                                         TNNI2           5        7360     1  \n",
       "2                                     Scoliosis           8        2102     2  \n",
       "3                    ENSEMBL:ENSOANG00000007850           5        2284     0  \n",
       "4                                          AGRN           5        2205     1  \n",
       "...                                         ...         ...         ...   ...  \n",
       "85949                                     pak-1           5        6906     0  \n",
       "85950  Z disc myocardium disorganized, abnormal           8        6931     2  \n",
       "85951                ENSEMBL:ENSECAG00000020912           5        6757     0  \n",
       "85952                                     act-2           5        5490     0  \n",
       "85953                ENSEMBL:ENSMMUG00000013138           5        1743     0  \n",
       "\n",
       "[85954 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = pd.read_csv('output/indexed_edges.csv')\n",
    "edges"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load nodes with their features and edges into graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, {'node_feature': tensor([ 0.0257, -0.2106,  0.0379,  0.2548, -0.0710, -0.1161,  0.0487, -0.0889,\n",
      "        -0.2617,  0.0058,  0.1430, -0.2625, -0.0045,  0.0088, -0.0159,  0.2041,\n",
      "        -0.0750, -0.0746, -0.1348,  0.5121,  0.3134,  0.3092, -0.0215, -0.1648,\n",
      "        -0.1561,  0.3388, -0.3543, -0.0308, -0.0959, -0.1503, -0.0089,  0.0727,\n",
      "        -0.2139, -0.1520, -0.0370,  0.2890, -0.0060,  0.0132,  0.1780,  0.0230,\n",
      "         0.0030,  0.1089, -0.0066,  0.0677,  0.0310,  0.0451,  0.0139,  0.0874,\n",
      "        -0.0229,  0.2928,  0.1286,  0.0255,  0.1482, -0.0255,  0.1047,  0.1834,\n",
      "         0.0289, -0.0020, -0.1158,  0.1742, -0.0059, -0.2629,  0.0610,  0.0691])})\n",
      "(2, 835, {})\n",
      "Number of edges is 85840\n",
      "Number of nodes is 10237\n"
     ]
    }
   ],
   "source": [
    "G = nx.DiGraph()\n",
    "for ind, node in node_features.iterrows(): \n",
    "    G.add_node(int(node['Node']), node_feature=torch.Tensor(node['Embedding']))\n",
    "for ind, edge in edges.iterrows(): \n",
    "    G.add_edge(int(edge['index_head']), int(edge['index_tail']))\n",
    "  \n",
    "for node in G.nodes(data=True):\n",
    "    print(node)\n",
    "    break\n",
    "for edge in G.edges(data=True):\n",
    "    print(edge)\n",
    "    break\n",
    "\n",
    "print(\"Number of edges is {}\".format(G.number_of_edges()))\n",
    "print(\"Number of nodes is {}\".format(G.number_of_nodes()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch.Size([10237, 64])\n",
      "torch.Size([2, 85840])\n"
     ]
    }
   ],
   "source": [
    "DeepG = Graph(G)\n",
    "\n",
    "print(DeepG.node_feature.shape)\n",
    "print(DeepG.edge_index.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[ 0.0257, -0.2106,  0.0379,  ..., -0.2629,  0.0610,  0.0691],\n",
       "        [ 0.1881, -0.1237, -0.0076,  ..., -0.0790, -0.0277, -0.0832],\n",
       "        [ 0.5650,  0.0616,  0.7965,  ..., -0.4596, -0.2780,  0.3672],\n",
       "        ...,\n",
       "        [ 0.3306, -0.2026,  0.1475,  ..., -0.0057,  0.0220, -0.0708],\n",
       "        [ 0.3033, -0.1983, -0.0204,  ...,  0.0915,  0.1807,  0.2646],\n",
       "        [ 0.7833, -0.0757,  0.5275,  ..., -0.4399,  0.0885, -0.2627]])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = DeepG.node_feature\n",
    "x"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load arguments and parameters of trained GNN model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<All keys matched successfully>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('output/best_model_args.pkl', 'rb') as f:\n",
    "    loaded_args = pickle.load(f)\n",
    "\n",
    "# Important! Changed the forward function such that it allows input of less edges that are predicted\n",
    "# and it does not return newly calculated x\n",
    "\n",
    "class LinkPredModel(torch.nn.Module):\n",
    "    \"\"\"\n",
    "        Architecture contains Batch Normalization layers (https://towardsdatascience.com/batch-norm-explained-visually-how-it-works-and-why-neural-networks-need-it-b18919692739) \n",
    "        between the SAGEConvolutional layers.\n",
    "    \"\"\"\n",
    "    def __init__(self, input_size, hidden_size, out_size, num_layers, aggr, dropout, device):\n",
    "        super(LinkPredModel, self).__init__()\n",
    "\n",
    "        self.convs = torch.nn.ModuleList()\n",
    "        self.bns = torch.nn.ModuleList()\n",
    "        \n",
    "        self.convs.append(SAGEConv(input_size, hidden_size, normalize=True, aggr=aggr)) # input node embedding features, output is size of hidden layers\n",
    "        self.bns.append(nn.BatchNorm1d(hidden_size))\n",
    "        for _ in range(num_layers - 2):\n",
    "            self.convs.append(SAGEConv(hidden_size, hidden_size, normalize=True, aggr=aggr))\n",
    "            self.bns.append(nn.BatchNorm1d(hidden_size))\n",
    "        self.convs.append(SAGEConv(hidden_size, out_size, normalize=True, aggr=aggr))\n",
    "        self.bns.append(nn.BatchNorm1d(out_size))\n",
    "\n",
    "        self.dropout = dropout\n",
    "        self.loss_fn = torch.nn.BCEWithLogitsLoss()\n",
    "        \n",
    "        self.device = device\n",
    "\n",
    "\n",
    "    def forward(self, x, edge_index, edge_label_index, training = False):\n",
    "        x = x.to(self.device)\n",
    "        edge_index = edge_index.to(self.device)\n",
    "        edge_label_index = edge_label_index.to(self.device)\n",
    "\n",
    "        for i in range(len(self.convs) - 1):\n",
    "          x = self.convs[i](x, edge_index)\n",
    "          x = self.bns[i](x)\n",
    "          x = F.leaky_relu(x)\n",
    "          x = F.dropout(x, p=self.dropout, training=training)\n",
    "        \n",
    "        x = self.convs[-1](x, edge_index)  \n",
    "        x = self.bns[-1](x)\n",
    "        \n",
    "\n",
    "        nodes_first, nodes_second = edge_label_index\n",
    "        nodes_first_features = x[nodes_first.long()]\n",
    "        nodes_second_features = x[nodes_second.long()]\n",
    "        \n",
    "        pred = torch.sum(nodes_first_features * nodes_second_features, dim=-1)\n",
    "        \n",
    "        return pred\n",
    "    \n",
    "    def loss(self, pred, label):\n",
    "        return self.loss_fn(pred, label)\n",
    "\n",
    "num_node_features = len(node_features['Embedding'][0])\n",
    "\n",
    "best_model = LinkPredModel(num_node_features, loaded_args['hidden_dim'], loaded_args['output_dim'], \n",
    "                           loaded_args['layers'], loaded_args['aggr'], loaded_args['dropout'], \n",
    "                           loaded_args['device']).to(loaded_args['device'])\n",
    "\n",
    "best_model.load_state_dict(torch.load('output/best_model.pth'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GNNExplainer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_config = ModelConfig(\n",
    "    mode='binary_classification',\n",
    "    task_level='edge',\n",
    "    return_type='raw',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "explainer = Explainer(\n",
    "    model=best_model,\n",
    "    explanation_type='model',\n",
    "    algorithm=GNNExplainer(epochs=700),\n",
    "    node_mask_type='attributes',\n",
    "    edge_mask_type='object',\n",
    "    model_config=model_config,\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the node indices that are relevant to the edge of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "node1:\n",
      "     index_id          id   semantic               label  semantic_id\n",
      "184       184  HP:0003560  phenotype  Muscular dystrophy            8\n",
      "node2:\n",
      "       index_id   id semantic       label  semantic_id\n",
      "10137     10137  522     drug  carvedilol            4\n"
     ]
    }
   ],
   "source": [
    "def get_node_idx(id, nodes): \n",
    "    return nodes[nodes['id'] == id].index.values.astype(int)[0]\n",
    "\n",
    "node_idx1 = get_node_idx('HP:0003560', nodes)\n",
    "print('node1:')\n",
    "print(nodes.loc[[node_idx1]])\n",
    "\n",
    "node_idx2 = get_node_idx('522', nodes)\n",
    "print('node2:')\n",
    "print(nodes.loc[[node_idx2]])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a tensor as input of the model. Only predict the label of a single edge instead of all edges in current dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([  184, 10137], dtype=torch.int32)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_label_index = torch.tensor([node_idx1, node_idx2])\n",
    "edge_label_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Iteration 0 ---\n",
      "A total of 3687 edges contributing to prediction.\n",
      "Maximum contribution value is 0.030153825879096985\n",
      "Based on desired graph size, contribution limit is set to 0.02365393005311489\n",
      "--- Iteration 1 ---\n",
      "A total of 3687 edges contributing to prediction.\n",
      "Maximum contribution value is 0.029727665707468987\n",
      "Based on desired graph size, contribution limit is set to 0.023980459198355675\n",
      "--- Iteration 2 ---\n",
      "A total of 3687 edges contributing to prediction.\n",
      "Maximum contribution value is 0.029797092080116272\n",
      "Based on desired graph size, contribution limit is set to 0.02377462200820446\n",
      "--- Iteration 3 ---\n",
      "A total of 3687 edges contributing to prediction.\n",
      "Maximum contribution value is 0.029377831146121025\n",
      "Based on desired graph size, contribution limit is set to 0.023571254685521126\n",
      "--- Iteration 4 ---\n",
      "A total of 3687 edges contributing to prediction.\n",
      "Maximum contribution value is 0.030073348432779312\n",
      "Based on desired graph size, contribution limit is set to 0.0237234178930521\n",
      "--- Iteration 5 ---\n",
      "A total of 3687 edges contributing to prediction.\n",
      "Maximum contribution value is 0.029522787779569626\n",
      "Based on desired graph size, contribution limit is set to 0.023629792034626007\n",
      "--- Iteration 6 ---\n",
      "A total of 3687 edges contributing to prediction.\n",
      "Maximum contribution value is 0.029795540496706963\n",
      "Based on desired graph size, contribution limit is set to 0.023874536156654358\n",
      "--- Iteration 7 ---\n",
      "A total of 3687 edges contributing to prediction.\n",
      "Maximum contribution value is 0.029592392966151237\n",
      "Based on desired graph size, contribution limit is set to 0.023723876103758812\n",
      "--- Iteration 8 ---\n",
      "A total of 3687 edges contributing to prediction.\n",
      "Maximum contribution value is 0.029410824179649353\n",
      "Based on desired graph size, contribution limit is set to 0.023741718381643295\n",
      "--- Iteration 9 ---\n",
      "A total of 3687 edges contributing to prediction.\n",
      "Maximum contribution value is 0.0304348636418581\n",
      "Based on desired graph size, contribution limit is set to 0.024054870009422302\n",
      "No good explanations found after 10 iterations...\n"
     ]
    }
   ],
   "source": [
    "done = False\n",
    "current_iterations = 0\n",
    "total_iterations = 10\n",
    "graph_size = 15 \n",
    "\n",
    "while not done:\n",
    "    print(f'--- Iteration {current_iterations} ---')\n",
    "\n",
    "    explanation = explainer(\n",
    "        x=DeepG.node_feature,\n",
    "        edge_index=DeepG.edge_index,\n",
    "        edge_label_index=edge_label_index,\n",
    "    )\n",
    "    \n",
    "    edge_mask = explanation.edge_mask\n",
    "    \n",
    "    print(f'A total of {torch.count_nonzero(edge_mask)} edges contributing to prediction.')\n",
    "    \n",
    "    sorted_contributions = edge_mask.sort(descending=True)[0]\n",
    "    print(f'Maximum contribution value is {sorted_contributions[0]}')\n",
    "    limit = sorted_contributions[graph_size]\n",
    "    print(f'Based on desired graph size, contribution limit is set to {limit}')\n",
    "    \n",
    "    pruned_G = copy.deepcopy(G)\n",
    "    for indx, edge in enumerate(G.edges):\n",
    "        if edge_mask[indx] < limit:\n",
    "            pruned_G.remove_edge(edge[0], edge[1])\n",
    "    \n",
    "    done = nx.has_path(pruned_G, node_idx1, node_idx2)\n",
    "    \n",
    "    current_iterations += 1\n",
    "    \n",
    "    if done:\n",
    "        print(f'A good explanation has been found!')\n",
    "    elif current_iterations == total_iterations and not done:\n",
    "        done = True\n",
    "        print(f'No good explanations found after {current_iterations} iterations...')\n",
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xaifognn",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
