{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This Jupyter notebook contains the GNNExplainer method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import torch\n",
    "\n",
    "import pickle\n",
    "import copy\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from gensim.models import KeyedVectors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_nr = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load all data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>WormBase:WBGene00000389</td>\n",
       "      <td>ORTH</td>\n",
       "      <td>cdc-25.4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ZP:0018675</td>\n",
       "      <td>DISO</td>\n",
       "      <td>right side lateral plate mesoderm mislocalised...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ZFIN:ZDB-GENE-040426-1197</td>\n",
       "      <td>ORTH</td>\n",
       "      <td>tbc1d5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>DRUG</td>\n",
       "      <td>(S)-nicardipine</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>RGD:3443</td>\n",
       "      <td>ORTH</td>\n",
       "      <td>Ptk2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10029</th>\n",
       "      <td>10029</td>\n",
       "      <td>MP:0009763</td>\n",
       "      <td>DISO</td>\n",
       "      <td>increased sensitivity to induced morbidity/mor...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10030</th>\n",
       "      <td>10030</td>\n",
       "      <td>MP:0011057</td>\n",
       "      <td>DISO</td>\n",
       "      <td>absent brain ependyma motile cilia</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10031</th>\n",
       "      <td>10031</td>\n",
       "      <td>MP:0001412</td>\n",
       "      <td>DISO</td>\n",
       "      <td>excessive scratching</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10032</th>\n",
       "      <td>10032</td>\n",
       "      <td>WBPhenotype:0004023</td>\n",
       "      <td>DISO</td>\n",
       "      <td>frequency of body bend variant</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10033</th>\n",
       "      <td>10033</td>\n",
       "      <td>ClinVarVariant:858516</td>\n",
       "      <td>VARI</td>\n",
       "      <td>NM_004006.3(DMD):c.9852G&gt;A (p.Trp3284Ter)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10034 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index_id                         id semantic  \\\n",
       "0             0    WormBase:WBGene00000389     ORTH   \n",
       "1             1                 ZP:0018675     DISO   \n",
       "2             2  ZFIN:ZDB-GENE-040426-1197     ORTH   \n",
       "3             3                          5     DRUG   \n",
       "4             4                   RGD:3443     ORTH   \n",
       "...         ...                        ...      ...   \n",
       "10029     10029                 MP:0009763     DISO   \n",
       "10030     10030                 MP:0011057     DISO   \n",
       "10031     10031                 MP:0001412     DISO   \n",
       "10032     10032        WBPhenotype:0004023     DISO   \n",
       "10033     10033      ClinVarVariant:858516     VARI   \n",
       "\n",
       "                                                   label  semantic_id  \n",
       "0                                               cdc-25.4            5  \n",
       "1      right side lateral plate mesoderm mislocalised...            1  \n",
       "2                                                 tbc1d5            5  \n",
       "3                                        (S)-nicardipine            2  \n",
       "4                                                   Ptk2            5  \n",
       "...                                                  ...          ...  \n",
       "10029  increased sensitivity to induced morbidity/mor...            1  \n",
       "10030                 absent brain ependyma motile cilia            1  \n",
       "10031                               excessive scratching            1  \n",
       "10032                     frequency of body bend variant            1  \n",
       "10033          NM_004006.3(DMD):c.9852G>A (p.Trp3284Ter)            7  \n",
       "\n",
       "[10034 rows x 5 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = pd.read_csv(f'output/indexed_nodes_{dataset_nr}.csv')\n",
    "nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the Edge2Vec embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Node</th>\n",
       "      <th>Embedding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>[0.80579096, 0.039609067, -0.17131972, 0.68869...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>[0.13173065, -0.30342472, 0.44653553, 0.617166...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>[0.62159157, -0.75673, 0.62674034, -0.35293385...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>[0.44389758, -0.25133932, -0.03261666, -0.1510...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>[0.44021487, -0.30697218, -0.24358109, 0.24869...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10029</th>\n",
       "      <td>10029</td>\n",
       "      <td>[0.35744143, -0.30719844, 0.43343344, -0.10684...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10030</th>\n",
       "      <td>10030</td>\n",
       "      <td>[0.5525602, -0.011062832, 0.14208537, 0.322823...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10031</th>\n",
       "      <td>10031</td>\n",
       "      <td>[0.61204404, -0.2899803, 0.26087046, -0.607966...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10032</th>\n",
       "      <td>10032</td>\n",
       "      <td>[0.29751068, -0.5034509, -0.01351818, -0.03332...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10033</th>\n",
       "      <td>10033</td>\n",
       "      <td>[0.3058808, -0.17289351, 0.2881391, 0.605097, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10034 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Node                                          Embedding\n",
       "0          0  [0.80579096, 0.039609067, -0.17131972, 0.68869...\n",
       "1          1  [0.13173065, -0.30342472, 0.44653553, 0.617166...\n",
       "2          2  [0.62159157, -0.75673, 0.62674034, -0.35293385...\n",
       "3          3  [0.44389758, -0.25133932, -0.03261666, -0.1510...\n",
       "4          4  [0.44021487, -0.30697218, -0.24358109, 0.24869...\n",
       "...      ...                                                ...\n",
       "10029  10029  [0.35744143, -0.30719844, 0.43343344, -0.10684...\n",
       "10030  10030  [0.5525602, -0.011062832, 0.14208537, 0.322823...\n",
       "10031  10031  [0.61204404, -0.2899803, 0.26087046, -0.607966...\n",
       "10032  10032  [0.29751068, -0.5034509, -0.01351818, -0.03332...\n",
       "10033  10033  [0.3058808, -0.17289351, 0.2881391, 0.605097, ...\n",
       "\n",
       "[10034 rows x 2 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_feat = KeyedVectors.load(f'output/w2v_{dataset_nr}.dvectors', mmap='r')\n",
    "\n",
    "e2v_embedding = pd.DataFrame(columns = ['Node', 'Embedding'])\n",
    "for idx, key in enumerate(node_feat.index_to_key):\n",
    "    e2v_embedding.loc[int(key)] = pd.Series({'Node':int(key), 'Embedding':list(node_feat[key])})\n",
    "    \n",
    "e2v_embedding = e2v_embedding.sort_values('Node')\n",
    "e2v_embedding"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZFIN:ZDB-GENE-050626-112</td>\n",
       "      <td>myl4</td>\n",
       "      <td>5</td>\n",
       "      <td>5279</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:FBgn0085464</td>\n",
       "      <td>CG34435</td>\n",
       "      <td>5</td>\n",
       "      <td>6825</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZFIN:ZDB-GENE-050626-112</td>\n",
       "      <td>myl4</td>\n",
       "      <td>5</td>\n",
       "      <td>5279</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>HGNC:7585</td>\n",
       "      <td>MYL4</td>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZFIN:ZDB-GENE-050626-112</td>\n",
       "      <td>myl4</td>\n",
       "      <td>5</td>\n",
       "      <td>5279</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>FlyBase:FBgn0002772</td>\n",
       "      <td>Mlc1</td>\n",
       "      <td>5</td>\n",
       "      <td>8901</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZFIN:ZDB-GENE-050626-112</td>\n",
       "      <td>myl4</td>\n",
       "      <td>5</td>\n",
       "      <td>5279</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>NCBIGene:396472</td>\n",
       "      <td>MYL4</td>\n",
       "      <td>3</td>\n",
       "      <td>9508</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZFIN:ZDB-GENE-050626-112</td>\n",
       "      <td>myl4</td>\n",
       "      <td>5</td>\n",
       "      <td>5279</td>\n",
       "      <td>in 1 to 1 orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSECAG00000020967</td>\n",
       "      <td>ENSEMBL:ENSECAG00000020967</td>\n",
       "      <td>5</td>\n",
       "      <td>8807</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82908</th>\n",
       "      <td>4810</td>\n",
       "      <td>ibrutinib</td>\n",
       "      <td>2</td>\n",
       "      <td>1618</td>\n",
       "      <td>targets</td>\n",
       "      <td>HGNC:11283</td>\n",
       "      <td>SRC</td>\n",
       "      <td>3</td>\n",
       "      <td>3279</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82909</th>\n",
       "      <td>522</td>\n",
       "      <td>carvedilol</td>\n",
       "      <td>2</td>\n",
       "      <td>184</td>\n",
       "      <td>targets</td>\n",
       "      <td>HGNC:620</td>\n",
       "      <td>APP</td>\n",
       "      <td>3</td>\n",
       "      <td>547</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82910</th>\n",
       "      <td>OMIM:300377.0013</td>\n",
       "      <td>DMD, EX18DEL</td>\n",
       "      <td>1</td>\n",
       "      <td>2822</td>\n",
       "      <td>is allele of</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>3</td>\n",
       "      <td>6612</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82911</th>\n",
       "      <td>Coriell:GM05113</td>\n",
       "      <td>NIGMS-GM05113</td>\n",
       "      <td>4</td>\n",
       "      <td>8105</td>\n",
       "      <td>has role in modeling</td>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>1</td>\n",
       "      <td>6315</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82912</th>\n",
       "      <td>OMIM:300377.0039</td>\n",
       "      <td>DMD, 1-BP INS, NT1554</td>\n",
       "      <td>1</td>\n",
       "      <td>7364</td>\n",
       "      <td>is allele of</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>3</td>\n",
       "      <td>6612</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82913 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           head             label_head  class_head  \\\n",
       "0      ZFIN:ZDB-GENE-050626-112                   myl4           5   \n",
       "1      ZFIN:ZDB-GENE-050626-112                   myl4           5   \n",
       "2      ZFIN:ZDB-GENE-050626-112                   myl4           5   \n",
       "3      ZFIN:ZDB-GENE-050626-112                   myl4           5   \n",
       "4      ZFIN:ZDB-GENE-050626-112                   myl4           5   \n",
       "...                         ...                    ...         ...   \n",
       "82908                      4810              ibrutinib           2   \n",
       "82909                       522             carvedilol           2   \n",
       "82910          OMIM:300377.0013           DMD, EX18DEL           1   \n",
       "82911           Coriell:GM05113          NIGMS-GM05113           4   \n",
       "82912          OMIM:300377.0039  DMD, 1-BP INS, NT1554           1   \n",
       "\n",
       "       index_head                               relation  \\\n",
       "0            5279         in orthology relationship with   \n",
       "1            5279         in orthology relationship with   \n",
       "2            5279         in orthology relationship with   \n",
       "3            5279         in orthology relationship with   \n",
       "4            5279  in 1 to 1 orthology relationship with   \n",
       "...           ...                                    ...   \n",
       "82908        1618                                targets   \n",
       "82909         184                                targets   \n",
       "82910        2822                           is allele of   \n",
       "82911        8105                   has role in modeling   \n",
       "82912        7364                           is allele of   \n",
       "\n",
       "                             tail                   label_tail  class_tail  \\\n",
       "0             FlyBase:FBgn0085464                      CG34435           5   \n",
       "1                       HGNC:7585                         MYL4           3   \n",
       "2             FlyBase:FBgn0002772                         Mlc1           5   \n",
       "3                 NCBIGene:396472                         MYL4           3   \n",
       "4      ENSEMBL:ENSECAG00000020967   ENSEMBL:ENSECAG00000020967           5   \n",
       "...                           ...                          ...         ...   \n",
       "82908                  HGNC:11283                          SRC           3   \n",
       "82909                    HGNC:620                          APP           3   \n",
       "82910                   HGNC:2928                          DMD           3   \n",
       "82911               MONDO:0010679  Duchenne muscular dystrophy           1   \n",
       "82912                   HGNC:2928                          DMD           3   \n",
       "\n",
       "       index_tail  type  \n",
       "0            6825     0  \n",
       "1              27     0  \n",
       "2            8901     0  \n",
       "3            9508     0  \n",
       "4            8807     1  \n",
       "...           ...   ...  \n",
       "82908        3279    14  \n",
       "82909         547    14  \n",
       "82910        6612    17  \n",
       "82911        6315    15  \n",
       "82912        6612    17  \n",
       "\n",
       "[82913 rows x 10 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = pd.read_csv(f'output/indexed_edges_{dataset_nr}.csv')\n",
    "edges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initialize a graph with edges and nodes including the embedding features of each node."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, {'node_feature': tensor([ 8.0579e-01,  3.9609e-02, -1.7132e-01,  6.8870e-01,  3.8687e-01,\n",
      "        -1.1756e+00,  4.3465e-01, -5.1411e-01, -8.6061e-01, -1.5018e-01,\n",
      "        -5.6375e-01,  5.2018e-02,  4.2212e-01, -1.4776e-01, -1.6446e-01,\n",
      "         9.3733e-01, -6.3037e-01, -2.5282e-01,  1.1521e-01,  5.3040e-01,\n",
      "         7.8000e-01,  3.6201e-01, -2.6033e-01, -1.1161e+00,  6.4173e-01,\n",
      "         2.9151e-01, -1.0554e+00, -8.0453e-01, -6.6271e-01,  2.2767e-02,\n",
      "         1.0419e-01,  1.0334e-01,  5.3798e-01, -1.7435e-01,  4.3009e-01,\n",
      "         1.6690e-01, -4.3212e-01,  1.6400e-01, -2.9133e-01,  1.0707e+00,\n",
      "        -3.5062e-02,  1.2267e-01,  3.4321e-01, -8.6810e-02,  1.0513e+00,\n",
      "        -5.3982e-01, -2.1837e-01, -1.5077e-01,  4.7775e-04, -1.2586e-02,\n",
      "        -7.0311e-01,  1.1769e-01,  2.7447e-01, -2.5308e-02,  6.0167e-01,\n",
      "         1.3919e-01,  3.6101e-02, -5.8889e-03, -4.1151e-01,  2.1858e-01,\n",
      "        -7.8684e-01, -1.7285e-01,  6.9146e-01, -7.3086e-02])})\n",
      "(0, 6308, {'edge_label': 'in orthology relationship with'})\n",
      "Number of edges is 82899\n",
      "Number of nodes is 10034\n"
     ]
    }
   ],
   "source": [
    "G = nx.DiGraph() # TODO: changed from Graph\n",
    "for ind, node in e2v_embedding.iterrows(): \n",
    "    G.add_node(int(node['Node']), node_feature=torch.Tensor(node['Embedding']))\n",
    "for ind, edge in edges.iterrows(): \n",
    "    G.add_edge(int(edge['index_head']), int(edge['index_tail']), edge_label = edge['relation'])\n",
    "  \n",
    "for node in G.nodes(data=True):\n",
    "    print(node)\n",
    "    break\n",
    "for edge in G.edges(data=True):\n",
    "    print(edge)\n",
    "    break\n",
    "\n",
    "print(\"Number of edges is {}\".format(G.number_of_edges()))\n",
    "print(\"Number of nodes is {}\".format(G.number_of_nodes()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{(0, 6308): 'in orthology relationship with',\n",
       " (0, 9835): 'in 1 to 1 orthology relationship with',\n",
       " (0, 1681): 'in orthology relationship with',\n",
       " (0, 3670): 'in 1 to 1 orthology relationship with',\n",
       " (0, 5156): 'in orthology relationship with',\n",
       " (0, 2619): 'in orthology relationship with',\n",
       " (0, 2226): 'in orthology relationship with',\n",
       " (0, 10028): 'in 1 to 1 orthology relationship with',\n",
       " (0, 363): 'in orthology relationship with',\n",
       " (0, 8710): 'in orthology relationship with',\n",
       " (0, 8473): 'in orthology relationship with',\n",
       " (0, 8615): 'in orthology relationship with',\n",
       " (0, 5765): 'interacts with',\n",
       " (0, 5949): 'in orthology relationship with',\n",
       " (0, 3961): 'in orthology relationship with',\n",
       " (0, 5248): 'in 1 to 1 orthology relationship with',\n",
       " (0, 8756): 'in orthology relationship with',\n",
       " (0, 2139): 'in orthology relationship with',\n",
       " (0, 9308): 'has phenotype',\n",
       " (0, 71): 'has phenotype',\n",
       " (0, 1856): 'in orthology relationship with',\n",
       " (0, 6666): 'in orthology relationship with',\n",
       " (2, 1154): 'in 1 to 1 orthology relationship with',\n",
       " (2, 7669): 'in 1 to 1 orthology relationship with',\n",
       " (2, 8729): 'in 1 to 1 orthology relationship with',\n",
       " (2, 2397): 'in 1 to 1 orthology relationship with',\n",
       " (2, 7575): 'in 1 to 1 orthology relationship with',\n",
       " (2, 3801): 'expressed in',\n",
       " (2, 5176): 'in 1 to 1 orthology relationship with',\n",
       " (2, 7443): 'in orthology relationship with',\n",
       " (2, 7124): 'in 1 to 1 orthology relationship with',\n",
       " (2, 5822): 'in 1 to 1 orthology relationship with',\n",
       " (2, 2555): 'in 1 to 1 orthology relationship with',\n",
       " (2, 2701): 'in 1 to 1 orthology relationship with',\n",
       " (2, 2759): 'in 1 to 1 orthology relationship with',\n",
       " (2, 3940): 'in 1 to 1 orthology relationship with',\n",
       " (2, 4957): 'in 1 to 1 orthology relationship with',\n",
       " (2, 7504): 'in 1 to 1 orthology relationship with',\n",
       " (2, 711): 'in 1 to 1 orthology relationship with',\n",
       " (2, 5483): 'expressed in',\n",
       " (2, 5480): 'in 1 to 1 orthology relationship with',\n",
       " (2, 8689): 'in 1 to 1 orthology relationship with',\n",
       " (3, 3972): 'targets',\n",
       " (4, 1579): 'in 1 to 1 orthology relationship with',\n",
       " (4, 5717): 'in 1 to 1 orthology relationship with',\n",
       " (4, 8052): 'in 1 to 1 orthology relationship with',\n",
       " (4, 444): 'in 1 to 1 orthology relationship with',\n",
       " (4, 9798): 'is part of',\n",
       " (4, 7774): 'enables',\n",
       " (4, 6186): 'in 1 to 1 orthology relationship with',\n",
       " (4, 8185): 'in 1 to 1 orthology relationship with',\n",
       " (4, 3852): 'in 1 to 1 orthology relationship with',\n",
       " (4, 2609): 'in 1 to 1 orthology relationship with',\n",
       " (4, 8851): 'enables',\n",
       " (4, 9155): 'is part of',\n",
       " (4, 2236): 'in 1 to 1 orthology relationship with',\n",
       " (4, 5050): 'is part of',\n",
       " (4, 8863): 'in 1 to 1 orthology relationship with',\n",
       " (4, 61): 'in 1 to 1 orthology relationship with',\n",
       " (4, 9792): 'in 1 to 1 orthology relationship with',\n",
       " (5, 6612): 'has affected feature',\n",
       " (5, 4648): 'contributes to condition',\n",
       " (6, 9062): 'targets',\n",
       " (6, 6995): 'targets',\n",
       " (6, 8702): 'targets',\n",
       " (6, 4940): 'targets',\n",
       " (6, 3972): 'targets',\n",
       " (9, 6315): 'None',\n",
       " (11, 6612): 'has affected feature',\n",
       " (11, 6315): 'pathogenic for condition',\n",
       " (13, 8501): 'targets',\n",
       " (15, 3625): 'in 1 to 1 orthology relationship with',\n",
       " (15, 4329): 'in 1 to 1 orthology relationship with',\n",
       " (15, 314): 'in 1 to 1 orthology relationship with',\n",
       " (15, 5016): 'in orthology relationship with',\n",
       " (15, 5781): 'in 1 to 1 orthology relationship with',\n",
       " (15, 2105): 'in orthology relationship with',\n",
       " (15, 1531): 'in orthology relationship with',\n",
       " (15, 7157): 'in 1 to 1 orthology relationship with',\n",
       " (15, 5632): 'in 1 to 1 orthology relationship with',\n",
       " (15, 7581): 'in 1 to 1 orthology relationship with',\n",
       " (15, 9069): 'in 1 to 1 orthology relationship with',\n",
       " (15, 8523): 'in 1 to 1 orthology relationship with',\n",
       " (15, 4316): 'in 1 to 1 orthology relationship with',\n",
       " (16, 4153): 'in 1 to 1 orthology relationship with',\n",
       " (16, 5757): 'in 1 to 1 orthology relationship with',\n",
       " (16, 9658): 'in 1 to 1 orthology relationship with',\n",
       " (16, 3630): 'in 1 to 1 orthology relationship with',\n",
       " (16, 3712): 'in 1 to 1 orthology relationship with',\n",
       " (16, 4089): 'in 1 to 1 orthology relationship with',\n",
       " (16, 9461): 'in 1 to 1 orthology relationship with',\n",
       " (16, 6189): 'in 1 to 1 orthology relationship with',\n",
       " (16, 4559): 'in orthology relationship with',\n",
       " (16, 7727): 'in 1 to 1 orthology relationship with',\n",
       " (16, 4820): 'in 1 to 1 orthology relationship with',\n",
       " (16, 7994): 'in 1 to 1 orthology relationship with',\n",
       " (16, 8308): 'in orthology relationship with',\n",
       " (16, 1120): 'in orthology relationship with',\n",
       " (16, 507): 'in 1 to 1 orthology relationship with',\n",
       " (16, 4884): 'in 1 to 1 orthology relationship with',\n",
       " (16, 3272): 'in 1 to 1 orthology relationship with',\n",
       " (19, 5107): 'None',\n",
       " (19, 6315): 'has role in modeling',\n",
       " (19, 1893): 'has phenotype',\n",
       " (19, 4460): 'has phenotype',\n",
       " (20, 2788): 'has phenotype',\n",
       " (20, 2241): 'has phenotype',\n",
       " (20, 8469): 'None',\n",
       " (20, 6315): 'has role in modeling',\n",
       " (20, 8368): 'has phenotype',\n",
       " (23, 6315): 'pathogenic for condition',\n",
       " (23, 6612): 'has affected feature',\n",
       " (24, 8561): 'in 1 to 1 orthology relationship with',\n",
       " (24, 5022): 'has phenotype',\n",
       " (24, 8226): 'in 1 to 1 orthology relationship with',\n",
       " (24, 4839): 'has phenotype',\n",
       " (24, 3067): 'has phenotype',\n",
       " (24, 3231): 'involved in',\n",
       " (24, 5867): 'has phenotype',\n",
       " (24, 5077): 'involved in',\n",
       " (24, 8340): 'has phenotype',\n",
       " (24, 1999): 'in 1 to 1 orthology relationship with',\n",
       " (24, 656): 'has phenotype',\n",
       " (24, 2391): 'has phenotype',\n",
       " (24, 7328): 'has phenotype',\n",
       " (24, 3347): 'in 1 to 1 orthology relationship with',\n",
       " (24, 3214): 'in 1 to 1 orthology relationship with',\n",
       " (24, 9065): 'has phenotype',\n",
       " (24, 31): 'has phenotype',\n",
       " (24, 7825): 'has phenotype',\n",
       " (24, 478): 'has phenotype',\n",
       " (24, 948): 'in 1 to 1 orthology relationship with',\n",
       " (24, 8599): 'has phenotype',\n",
       " (24, 5638): 'has phenotype',\n",
       " (24, 7245): 'in 1 to 1 orthology relationship with',\n",
       " (24, 7523): 'has phenotype',\n",
       " (24, 2171): 'has phenotype',\n",
       " (24, 1994): 'enables',\n",
       " (24, 2527): 'has phenotype',\n",
       " (24, 3671): 'has phenotype',\n",
       " (24, 8851): 'enables',\n",
       " (24, 7774): 'enables',\n",
       " (24, 3087): 'in 1 to 1 orthology relationship with',\n",
       " (24, 9806): 'has phenotype',\n",
       " (24, 9876): 'in 1 to 1 orthology relationship with',\n",
       " (24, 667): 'has phenotype',\n",
       " (24, 6008): 'has phenotype',\n",
       " (24, 3336): 'has phenotype',\n",
       " (24, 7904): 'has phenotype',\n",
       " (24, 5280): 'has phenotype',\n",
       " (24, 6409): 'has phenotype',\n",
       " (24, 5819): 'has phenotype',\n",
       " (24, 3990): 'in 1 to 1 orthology relationship with',\n",
       " (24, 7969): 'in 1 to 1 orthology relationship with',\n",
       " (24, 7316): 'enables',\n",
       " (24, 5207): 'has phenotype',\n",
       " (24, 7377): 'has phenotype',\n",
       " (24, 10029): 'has phenotype',\n",
       " (24, 5977): 'has phenotype',\n",
       " (24, 8820): 'in 1 to 1 orthology relationship with',\n",
       " (24, 1598): 'has phenotype',\n",
       " (24, 297): 'has phenotype',\n",
       " (24, 738): 'has phenotype',\n",
       " (24, 9777): 'has phenotype',\n",
       " (24, 2564): 'has phenotype',\n",
       " (24, 284): 'has phenotype',\n",
       " (24, 4093): 'has phenotype',\n",
       " (24, 1379): 'has phenotype',\n",
       " (24, 2035): 'enables',\n",
       " (24, 191): 'in 1 to 1 orthology relationship with',\n",
       " (24, 5231): 'in 1 to 1 orthology relationship with',\n",
       " (24, 7296): 'in 1 to 1 orthology relationship with',\n",
       " (24, 3225): 'has phenotype',\n",
       " (24, 2654): 'has phenotype',\n",
       " (24, 9798): 'is part of',\n",
       " (24, 3613): 'has phenotype',\n",
       " (24, 863): 'has phenotype',\n",
       " (24, 9082): 'has phenotype',\n",
       " (25, 6612): 'has affected feature',\n",
       " (25, 6315): 'pathogenic for condition',\n",
       " (26, 3983): 'in 1 to 1 orthology relationship with',\n",
       " (26, 4140): 'in 1 to 1 orthology relationship with',\n",
       " (26, 6507): 'in orthology relationship with',\n",
       " (26, 6752): 'in 1 to 1 orthology relationship with',\n",
       " (26, 3265): 'in 1 to 1 orthology relationship with',\n",
       " (26, 2280): 'in 1 to 1 orthology relationship with',\n",
       " (26, 7811): 'in orthology relationship with',\n",
       " (26, 5581): 'in orthology relationship with',\n",
       " (26, 7796): 'in 1 to 1 orthology relationship with',\n",
       " (26, 8501): 'in 1 to 1 orthology relationship with',\n",
       " (26, 4118): 'in 1 to 1 orthology relationship with',\n",
       " (26, 7874): 'in orthology relationship with',\n",
       " (26, 193): 'in 1 to 1 orthology relationship with',\n",
       " (26, 7672): 'in 1 to 1 orthology relationship with',\n",
       " (26, 7771): 'in orthology relationship with',\n",
       " (26, 3298): 'in orthology relationship with',\n",
       " (26, 2632): 'in 1 to 1 orthology relationship with',\n",
       " (26, 9083): 'in 1 to 1 orthology relationship with',\n",
       " (26, 8114): 'in 1 to 1 orthology relationship with',\n",
       " (27, 5622): 'in 1 to 1 orthology relationship with',\n",
       " (27, 154): 'interacts with',\n",
       " (27, 5448): 'interacts with',\n",
       " (27, 9266): 'in 1 to 1 orthology relationship with',\n",
       " (27, 4254): 'interacts with',\n",
       " (27, 7020): 'in orthology relationship with',\n",
       " (27, 2601): 'expressed in',\n",
       " (27, 6533): 'interacts with',\n",
       " (27, 5811): 'interacts with',\n",
       " (27, 1250): 'interacts with',\n",
       " (27, 8208): 'interacts with',\n",
       " (27, 660): 'in 1 to 1 orthology relationship with',\n",
       " (27, 4867): 'interacts with',\n",
       " (27, 6963): 'in 1 to 1 orthology relationship with',\n",
       " (27, 7581): 'interacts with',\n",
       " (27, 5756): 'involved in',\n",
       " (27, 7037): 'interacts with',\n",
       " (27, 8400): 'in 1 to 1 orthology relationship with',\n",
       " (27, 8807): 'in 1 to 1 orthology relationship with',\n",
       " (27, 3231): 'involved in',\n",
       " (27, 1568): 'in orthology relationship with',\n",
       " (27, 9155): 'is part of',\n",
       " (27, 2302): 'in orthology relationship with',\n",
       " (27, 7950): 'involved in',\n",
       " (27, 9754): 'interacts with',\n",
       " (27, 5279): 'in orthology relationship with',\n",
       " (27, 1999): 'interacts with',\n",
       " (27, 1131): 'in 1 to 1 orthology relationship with',\n",
       " (27, 6440): 'interacts with',\n",
       " (27, 8264): 'interacts with',\n",
       " (27, 9508): 'in orthology relationship with',\n",
       " (27, 9373): 'interacts with',\n",
       " (27, 9608): 'interacts with',\n",
       " (27, 6825): 'in orthology relationship with',\n",
       " (27, 9119): 'interacts with',\n",
       " (27, 4077): 'interacts with',\n",
       " (27, 5410): 'in orthology relationship with',\n",
       " (27, 6400): 'in 1 to 1 orthology relationship with',\n",
       " (27, 9104): 'in 1 to 1 orthology relationship with',\n",
       " (27, 8978): 'interacts with',\n",
       " (27, 3693): 'interacts with',\n",
       " (27, 753): 'in orthology relationship with',\n",
       " (27, 746): 'interacts with',\n",
       " (27, 3146): 'interacts with',\n",
       " (27, 229): 'in orthology relationship with',\n",
       " (27, 5344): 'in 1 to 1 orthology relationship with',\n",
       " (27, 2568): 'in 1 to 1 orthology relationship with',\n",
       " (27, 9520): 'in 1 to 1 orthology relationship with',\n",
       " (27, 1663): 'interacts with',\n",
       " (27, 1072): 'interacts with',\n",
       " (27, 6449): 'in 1 to 1 orthology relationship with',\n",
       " (27, 4176): 'in 1 to 1 orthology relationship with',\n",
       " (27, 7197): 'in 1 to 1 orthology relationship with',\n",
       " (27, 2274): 'interacts with',\n",
       " (27, 5664): 'interacts with',\n",
       " (27, 8901): 'in orthology relationship with',\n",
       " (27, 7798): 'interacts with',\n",
       " (27, 6612): 'interacts with',\n",
       " (27, 8865): 'interacts with',\n",
       " (27, 1330): 'in orthology relationship with',\n",
       " (27, 9483): 'interacts with',\n",
       " (27, 7126): 'in 1 to 1 orthology relationship with',\n",
       " (27, 5617): 'interacts with',\n",
       " (27, 904): 'in 1 to 1 orthology relationship with',\n",
       " (28, 5333): 'in 1 to 1 orthology relationship with',\n",
       " (28, 3669): 'in 1 to 1 orthology relationship with',\n",
       " (28, 5370): 'in 1 to 1 orthology relationship with',\n",
       " (28, 6457): 'in 1 to 1 orthology relationship with',\n",
       " (28, 6351): 'in 1 to 1 orthology relationship with',\n",
       " (28, 1653): 'in 1 to 1 orthology relationship with',\n",
       " (28, 3873): 'has phenotype',\n",
       " (28, 6742): 'in 1 to 1 orthology relationship with',\n",
       " (28, 6809): 'in 1 to 1 orthology relationship with',\n",
       " (28, 2533): 'in 1 to 1 orthology relationship with',\n",
       " (28, 276): 'has phenotype',\n",
       " (28, 4872): 'has phenotype',\n",
       " (28, 6012): 'in 1 to 1 orthology relationship with',\n",
       " (28, 6652): 'in 1 to 1 orthology relationship with',\n",
       " (28, 4808): 'in 1 to 1 orthology relationship with',\n",
       " (28, 8596): 'in 1 to 1 orthology relationship with',\n",
       " (28, 7394): 'has phenotype',\n",
       " (28, 80): 'has phenotype',\n",
       " (28, 71): 'has phenotype',\n",
       " (28, 4476): 'in 1 to 1 orthology relationship with',\n",
       " (28, 6770): 'in 1 to 1 orthology relationship with',\n",
       " (28, 107): 'has phenotype',\n",
       " (28, 8525): 'has phenotype',\n",
       " (28, 7094): 'in 1 to 1 orthology relationship with',\n",
       " (28, 5258): 'has phenotype',\n",
       " (28, 4620): 'has phenotype',\n",
       " (28, 5852): 'has phenotype',\n",
       " (28, 9606): 'in 1 to 1 orthology relationship with',\n",
       " (28, 1113): 'in 1 to 1 orthology relationship with',\n",
       " (28, 5993): 'in 1 to 1 orthology relationship with',\n",
       " (28, 4439): 'in 1 to 1 orthology relationship with',\n",
       " (28, 1919): 'has phenotype',\n",
       " (28, 7964): 'interacts with',\n",
       " (28, 6255): 'has phenotype',\n",
       " (30, 4016): 'targets',\n",
       " (32, 6368): 'interacts with',\n",
       " (32, 5357): 'has phenotype',\n",
       " (32, 5638): 'has phenotype',\n",
       " (32, 5595): 'in 1 to 1 orthology relationship with',\n",
       " (32, 1384): 'has phenotype',\n",
       " (32, 9346): 'is part of',\n",
       " (32, 2312): 'interacts with',\n",
       " (32, 5050): 'is part of',\n",
       " (32, 8989): 'has phenotype',\n",
       " (32, 5911): 'interacts with',\n",
       " (32, 7707): 'interacts with',\n",
       " (32, 9798): 'is part of',\n",
       " (32, 6892): 'is part of',\n",
       " (32, 5932): 'in 1 to 1 orthology relationship with',\n",
       " (32, 1776): 'interacts with',\n",
       " (32, 3067): 'has phenotype',\n",
       " (32, 2291): 'colocalizes with',\n",
       " (32, 2788): 'has phenotype',\n",
       " (32, 142): 'has phenotype',\n",
       " (32, 5738): 'in orthology relationship with',\n",
       " (32, 4990): 'interacts with',\n",
       " (32, 842): 'interacts with',\n",
       " (32, 4446): 'interacts with',\n",
       " (32, 9651): 'has phenotype',\n",
       " (32, 8896): 'colocalizes with',\n",
       " (32, 7202): 'in 1 to 1 orthology relationship with',\n",
       " (32, 1433): 'in 1 to 1 orthology relationship with',\n",
       " (32, 4069): 'has phenotype',\n",
       " (32, 6679): 'in 1 to 1 orthology relationship with',\n",
       " (32, 3163): 'in orthology relationship with',\n",
       " (32, 6286): 'has phenotype',\n",
       " (32, 4095): 'in 1 to 1 orthology relationship with',\n",
       " (32, 4992): 'in 1 to 1 orthology relationship with',\n",
       " (32, 6572): 'has phenotype',\n",
       " (32, 863): 'has phenotype',\n",
       " (32, 8851): 'enables',\n",
       " (32, 1816): 'has phenotype',\n",
       " (32, 933): 'interacts with',\n",
       " (32, 4893): 'has phenotype',\n",
       " (32, 5826): 'enables',\n",
       " (32, 5063): 'interacts with',\n",
       " (32, 9806): 'has phenotype',\n",
       " (32, 3538): 'has phenotype',\n",
       " (32, 1580): 'has phenotype',\n",
       " (32, 321): 'has phenotype',\n",
       " (32, 4794): 'has phenotype',\n",
       " (32, 5667): 'has phenotype',\n",
       " (32, 6200): 'has phenotype',\n",
       " (32, 5616): 'in 1 to 1 orthology relationship with',\n",
       " (32, 8459): 'is part of',\n",
       " (32, 3183): 'has phenotype',\n",
       " (32, 4125): 'has phenotype',\n",
       " (32, 575): 'has phenotype',\n",
       " (32, 731): 'has phenotype',\n",
       " (32, 5547): 'has phenotype',\n",
       " (32, 5266): 'in orthology relationship with',\n",
       " (32, 418): 'has phenotype',\n",
       " (32, 7207): 'has phenotype',\n",
       " (32, 3841): 'in 1 to 1 orthology relationship with',\n",
       " (32, 3096): 'has phenotype',\n",
       " (32, 2087): 'has phenotype',\n",
       " (32, 6558): 'interacts with',\n",
       " (32, 1994): 'enables',\n",
       " (32, 6552): 'has phenotype',\n",
       " (32, 4897): 'in 1 to 1 orthology relationship with',\n",
       " (32, 3275): 'interacts with',\n",
       " (32, 1525): 'has phenotype',\n",
       " (32, 586): 'has phenotype',\n",
       " (32, 2784): 'has phenotype',\n",
       " (32, 9787): 'has phenotype',\n",
       " (32, 2154): 'is part of',\n",
       " (32, 789): 'has phenotype',\n",
       " (32, 8552): 'interacts with',\n",
       " (32, 5369): 'interacts with',\n",
       " (32, 2391): 'has phenotype',\n",
       " (32, 7897): 'has phenotype',\n",
       " (32, 7475): 'has phenotype',\n",
       " (32, 4863): 'is part of',\n",
       " (32, 1990): 'has phenotype',\n",
       " (32, 3419): 'has phenotype',\n",
       " (32, 2402): 'has phenotype',\n",
       " (32, 7515): 'has phenotype',\n",
       " (32, 9230): 'interacts with',\n",
       " (32, 3937): 'interacts with',\n",
       " (32, 7105): 'in 1 to 1 orthology relationship with',\n",
       " (32, 8093): 'has phenotype',\n",
       " (32, 4602): 'colocalizes with',\n",
       " (32, 6448): 'in 1 to 1 orthology relationship with',\n",
       " (32, 9378): 'interacts with',\n",
       " (32, 1290): 'has phenotype',\n",
       " (32, 6864): 'has phenotype',\n",
       " (32, 7774): 'enables',\n",
       " (32, 5229): 'has phenotype',\n",
       " (32, 7074): 'has phenotype',\n",
       " (32, 8958): 'is part of',\n",
       " (32, 7085): 'has phenotype',\n",
       " (32, 444): 'colocalizes with',\n",
       " (32, 8469): 'interacts with',\n",
       " (32, 1532): 'in 1 to 1 orthology relationship with',\n",
       " (32, 3691): 'has phenotype',\n",
       " (32, 5569): 'has phenotype',\n",
       " (32, 3822): 'interacts with',\n",
       " (32, 4389): 'interacts with',\n",
       " (32, 5744): 'has phenotype',\n",
       " (32, 9892): 'has phenotype',\n",
       " (32, 8423): 'has phenotype',\n",
       " (32, 5380): 'interacts with',\n",
       " (32, 2364): 'interacts with',\n",
       " (32, 290): 'has phenotype',\n",
       " (32, 695): 'expressed in',\n",
       " (32, 9322): 'has phenotype',\n",
       " (32, 7552): 'has phenotype',\n",
       " (32, 2141): 'enables',\n",
       " (32, 333): 'colocalizes with',\n",
       " (32, 2505): 'has phenotype',\n",
       " (32, 7728): 'has phenotype',\n",
       " (32, 1205): 'has phenotype',\n",
       " (32, 9458): 'has phenotype',\n",
       " (33, 6315): 'pathogenic for condition',\n",
       " (33, 6612): 'has affected feature',\n",
       " (34, 6333): 'in 1 to 1 orthology relationship with',\n",
       " (34, 9420): 'in 1 to 1 orthology relationship with',\n",
       " (34, 3933): 'in 1 to 1 orthology relationship with',\n",
       " (34, 1939): 'in 1 to 1 orthology relationship with',\n",
       " (34, 1771): 'in orthology relationship with',\n",
       " (34, 4377): 'in 1 to 1 orthology relationship with',\n",
       " (34, 1749): 'in 1 to 1 orthology relationship with',\n",
       " (34, 4049): 'in 1 to 1 orthology relationship with',\n",
       " (34, 1262): 'in orthology relationship with',\n",
       " (34, 7631): 'in 1 to 1 orthology relationship with',\n",
       " (34, 4657): 'in orthology relationship with',\n",
       " (34, 9195): 'in 1 to 1 orthology relationship with',\n",
       " (36, 6315): 'pathogenic for condition',\n",
       " (36, 6612): 'has affected feature',\n",
       " (37, 6612): 'has affected feature',\n",
       " (37, 6315): 'likely pathogenic for condition',\n",
       " (38, 4547): 'in 1 to 1 orthology relationship with',\n",
       " (38, 3146): 'in 1 to 1 orthology relationship with',\n",
       " (38, 7919): 'in 1 to 1 orthology relationship with',\n",
       " (38, 2043): 'in orthology relationship with',\n",
       " (41, 5317): 'targets',\n",
       " (42, 6315): 'pathogenic for condition',\n",
       " (42, 6612): 'has affected feature',\n",
       " (43, 9838): 'in orthology relationship with',\n",
       " (43, 7412): 'in orthology relationship with',\n",
       " (43, 5941): 'has phenotype',\n",
       " (43, 7026): 'in orthology relationship with',\n",
       " (43, 1331): 'in orthology relationship with',\n",
       " (43, 349): 'has phenotype',\n",
       " (43, 7856): 'has phenotype',\n",
       " (43, 2084): 'has phenotype',\n",
       " (43, 6608): 'in orthology relationship with',\n",
       " (43, 8713): 'has phenotype',\n",
       " (43, 5655): 'has phenotype',\n",
       " (43, 742): 'in orthology relationship with',\n",
       " (43, 8851): 'enables',\n",
       " (43, 2885): 'in orthology relationship with',\n",
       " (43, 7485): 'in orthology relationship with',\n",
       " (43, 9404): 'has phenotype',\n",
       " (43, 6516): 'interacts with',\n",
       " (43, 8326): 'has phenotype',\n",
       " (43, 5016): 'interacts with',\n",
       " (43, 3498): 'has phenotype',\n",
       " (43, 8980): 'interacts with',\n",
       " (43, 5665): 'in orthology relationship with',\n",
       " (43, 668): 'has phenotype',\n",
       " (43, 1514): 'in orthology relationship with',\n",
       " (43, 1904): 'in orthology relationship with',\n",
       " (43, 7620): 'in orthology relationship with',\n",
       " (43, 9283): 'interacts with',\n",
       " (43, 1788): 'has phenotype',\n",
       " (43, 7675): 'in orthology relationship with',\n",
       " (43, 2657): 'in orthology relationship with',\n",
       " (43, 436): 'has phenotype',\n",
       " (43, 2132): 'in 1 to 1 orthology relationship with',\n",
       " (43, 3004): 'interacts with',\n",
       " (43, 1854): 'interacts with',\n",
       " (44, 6612): 'has affected feature',\n",
       " (44, 6315): 'pathogenic for condition',\n",
       " (45, 7447): 'in orthology relationship with',\n",
       " (45, 7632): 'in orthology relationship with',\n",
       " (45, 6615): 'in orthology relationship with',\n",
       " (45, 6859): 'in orthology relationship with',\n",
       " (45, 3861): 'in orthology relationship with',\n",
       " (45, 1996): 'in orthology relationship with',\n",
       " (45, 7636): 'in orthology relationship with',\n",
       " (45, 8463): 'in orthology relationship with',\n",
       " (45, 7044): 'in orthology relationship with',\n",
       " (45, 3491): 'in orthology relationship with',\n",
       " (45, 9155): 'is part of',\n",
       " (46, 2688): 'targets',\n",
       " (46, 565): 'targets',\n",
       " (55, 9155): 'is part of',\n",
       " (55, 423): 'in orthology relationship with',\n",
       " (55, 8493): 'in orthology relationship with',\n",
       " (55, 4427): 'in orthology relationship with',\n",
       " (55, 9291): 'in orthology relationship with',\n",
       " (55, 3325): 'in orthology relationship with',\n",
       " (55, 6768): 'in orthology relationship with',\n",
       " (55, 324): 'in orthology relationship with',\n",
       " (55, 6075): 'in orthology relationship with',\n",
       " (55, 6452): 'in orthology relationship with',\n",
       " (55, 6122): 'in orthology relationship with',\n",
       " (55, 8995): 'in orthology relationship with',\n",
       " (55, 2081): 'in orthology relationship with',\n",
       " (55, 6602): 'in orthology relationship with',\n",
       " (55, 5783): 'in orthology relationship with',\n",
       " (55, 1764): 'in orthology relationship with',\n",
       " (55, 5658): 'in orthology relationship with',\n",
       " (55, 6650): 'in orthology relationship with',\n",
       " (55, 2719): 'in orthology relationship with',\n",
       " (55, 4696): 'in orthology relationship with',\n",
       " (55, 9948): 'in orthology relationship with',\n",
       " (56, 4555): 'in 1 to 1 orthology relationship with',\n",
       " (56, 6856): 'in 1 to 1 orthology relationship with',\n",
       " (56, 5192): 'in 1 to 1 orthology relationship with',\n",
       " (56, 7215): 'in 1 to 1 orthology relationship with',\n",
       " (56, 2531): 'in 1 to 1 orthology relationship with',\n",
       " (56, 5592): 'in 1 to 1 orthology relationship with',\n",
       " (56, 6334): 'in 1 to 1 orthology relationship with',\n",
       " (56, 5570): 'in 1 to 1 orthology relationship with',\n",
       " (56, 2605): 'in 1 to 1 orthology relationship with',\n",
       " (56, 9015): 'in 1 to 1 orthology relationship with',\n",
       " (56, 4446): 'in 1 to 1 orthology relationship with',\n",
       " (56, 1789): 'in 1 to 1 orthology relationship with',\n",
       " (57, 6811): 'has phenotype',\n",
       " (57, 4315): 'in orthology relationship with',\n",
       " (57, 6788): 'has phenotype',\n",
       " (57, 9050): 'has phenotype',\n",
       " (57, 1416): 'has phenotype',\n",
       " (57, 432): 'has phenotype',\n",
       " (57, 9846): 'has phenotype',\n",
       " (57, 4872): 'has phenotype',\n",
       " (57, 1810): 'has phenotype',\n",
       " (57, 4465): 'interacts with',\n",
       " (57, 8509): 'has phenotype',\n",
       " (57, 365): 'has phenotype',\n",
       " (57, 5295): 'in orthology relationship with',\n",
       " (57, 7531): 'has phenotype',\n",
       " (57, 4352): 'has phenotype',\n",
       " (57, 4977): 'interacts with',\n",
       " (57, 3279): 'in orthology relationship with',\n",
       " (57, 6340): 'in orthology relationship with',\n",
       " (57, 4022): 'in orthology relationship with',\n",
       " (57, 1239): 'interacts with',\n",
       " (57, 2057): 'has phenotype',\n",
       " (57, 8851): 'enables',\n",
       " (57, 7528): 'has phenotype',\n",
       " (57, 1613): 'interacts with',\n",
       " (57, 5849): 'has phenotype',\n",
       " (57, 195): 'in orthology relationship with',\n",
       " (57, 8052): 'interacts with',\n",
       " (57, 5455): 'interacts with',\n",
       " (57, 3006): 'in 1 to 1 orthology relationship with',\n",
       " (57, 6038): 'interacts with',\n",
       " (57, 4931): 'has phenotype',\n",
       " (57, 4564): 'interacts with',\n",
       " (57, 1523): 'in orthology relationship with',\n",
       " (57, 1123): 'interacts with',\n",
       " (57, 2102): 'in orthology relationship with',\n",
       " (57, 1635): 'has phenotype',\n",
       " (57, 9859): 'in orthology relationship with',\n",
       " (57, 3681): 'interacts with',\n",
       " (57, 3400): 'interacts with',\n",
       " (57, 7546): 'in orthology relationship with',\n",
       " (57, 4866): 'interacts with',\n",
       " (57, 2358): 'has phenotype',\n",
       " (57, 9317): 'in orthology relationship with',\n",
       " (57, 4218): 'interacts with',\n",
       " (57, 7438): 'in orthology relationship with',\n",
       " (57, 4383): 'has phenotype',\n",
       " (57, 8869): 'has phenotype',\n",
       " (57, 1457): 'interacts with',\n",
       " (57, 3987): 'has phenotype',\n",
       " (57, 2324): 'has phenotype',\n",
       " (57, 1668): 'in orthology relationship with',\n",
       " (57, 9193): 'in orthology relationship with',\n",
       " (57, 3077): 'interacts with',\n",
       " (57, 4206): 'in orthology relationship with',\n",
       " (57, 903): 'has phenotype',\n",
       " (57, 9308): 'has phenotype',\n",
       " (57, 9436): 'interacts with',\n",
       " (61, 3852): 'in 1 to 1 orthology relationship with',\n",
       " (61, 6186): 'in 1 to 1 orthology relationship with',\n",
       " (61, 8052): 'in 1 to 1 orthology relationship with',\n",
       " (61, 5717): 'in 1 to 1 orthology relationship with',\n",
       " (61, 2609): 'in 1 to 1 orthology relationship with',\n",
       " (61, 1579): 'in 1 to 1 orthology relationship with',\n",
       " (61, 2236): 'in 1 to 1 orthology relationship with',\n",
       " (61, 4): 'in 1 to 1 orthology relationship with',\n",
       " (61, 444): 'in 1 to 1 orthology relationship with',\n",
       " (63, 5317): 'targets',\n",
       " (63, 2688): 'targets',\n",
       " (65, 4940): 'targets',\n",
       " (65, 9062): 'targets',\n",
       " (65, 3972): 'targets',\n",
       " (66, 8769): 'interacts with',\n",
       " (66, 3114): 'in orthology relationship with',\n",
       " (66, 5483): 'expressed in',\n",
       " (66, 2971): 'interacts with',\n",
       " (66, 5616): 'interacts with',\n",
       " (66, 6313): 'in 1 to 1 orthology relationship with',\n",
       " (66, 5833): 'interacts with',\n",
       " (66, 6612): 'interacts with',\n",
       " (66, 3059): 'in 1 to 1 orthology relationship with',\n",
       " (66, 1320): 'interacts with',\n",
       " (66, 2040): 'interacts with',\n",
       " (66, 5132): 'in 1 to 1 orthology relationship with',\n",
       " (66, 7993): 'interacts with',\n",
       " (66, 7931): 'interacts with',\n",
       " (66, 827): 'interacts with',\n",
       " (66, 5137): 'interacts with',\n",
       " (66, 7076): 'in orthology relationship with',\n",
       " (66, 5256): 'in 1 to 1 orthology relationship with',\n",
       " (66, 1750): 'interacts with',\n",
       " (66, 9109): 'interacts with',\n",
       " (66, 2917): 'in 1 to 1 orthology relationship with',\n",
       " (66, 8978): 'interacts with',\n",
       " (66, 8442): 'interacts with',\n",
       " (66, 4915): 'expressed in',\n",
       " (66, 565): 'interacts with',\n",
       " (66, 8851): 'enables',\n",
       " (66, 6478): 'interacts with',\n",
       " (66, 7581): 'interacts with',\n",
       " (66, 9847): 'in orthology relationship with',\n",
       " (66, 3563): 'interacts with',\n",
       " (66, 2331): 'interacts with',\n",
       " (66, 2722): 'interacts with',\n",
       " (66, 3648): 'interacts with',\n",
       " (66, 8432): 'interacts with',\n",
       " (66, 4037): 'interacts with',\n",
       " (66, 6186): 'interacts with',\n",
       " (66, 6406): 'interacts with',\n",
       " (66, 4947): 'in orthology relationship with',\n",
       " (66, 2705): 'interacts with',\n",
       " (66, 3922): 'interacts with',\n",
       " (66, 2157): 'interacts with',\n",
       " (66, 9596): 'interacts with',\n",
       " (66, 2688): 'interacts with',\n",
       " (66, 342): 'in orthology relationship with',\n",
       " (66, 6763): 'interacts with',\n",
       " (66, 7165): 'interacts with',\n",
       " (66, 7774): 'enables',\n",
       " (66, 9448): 'interacts with',\n",
       " (66, 3279): 'interacts with',\n",
       " (66, 4425): 'interacts with',\n",
       " (66, 6381): 'interacts with',\n",
       " (66, 7714): 'involved in',\n",
       " (66, 979): 'interacts with',\n",
       " (66, 8688): 'interacts with',\n",
       " (66, 9250): 'interacts with',\n",
       " (66, 7482): 'in 1 to 1 orthology relationship with',\n",
       " (66, 2714): 'in 1 to 1 orthology relationship with',\n",
       " (66, 1374): 'expressed in',\n",
       " (66, 9346): 'is part of',\n",
       " (66, 7385): 'is part of',\n",
       " (66, 2615): 'in 1 to 1 orthology relationship with',\n",
       " (66, 2613): 'in 1 to 1 orthology relationship with',\n",
       " (66, 3045): 'in 1 to 1 orthology relationship with',\n",
       " (66, 8352): 'in 1 to 1 orthology relationship with',\n",
       " (66, 9774): 'interacts with',\n",
       " (66, 5050): 'is part of',\n",
       " (66, 2154): 'is part of',\n",
       " (66, 1231): 'interacts with',\n",
       " (67, 547): 'targets',\n",
       " (68, 6612): 'has affected feature',\n",
       " (68, 6315): 'pathogenic for condition',\n",
       " (69, 2688): 'targets',\n",
       " (77, 6612): 'has affected feature',\n",
       " (77, 6315): 'pathogenic for condition',\n",
       " (82, 9210): 'in 1 to 1 orthology relationship with',\n",
       " (82, 3577): 'in 1 to 1 orthology relationship with',\n",
       " (82, 3058): 'in 1 to 1 orthology relationship with',\n",
       " (82, 778): 'in 1 to 1 orthology relationship with',\n",
       " (82, 2860): 'in 1 to 1 orthology relationship with',\n",
       " (82, 3263): 'in 1 to 1 orthology relationship with',\n",
       " (82, 5032): 'involved in',\n",
       " (82, 4975): 'in 1 to 1 orthology relationship with',\n",
       " (82, 3051): 'in orthology relationship with',\n",
       " (82, 2453): 'in 1 to 1 orthology relationship with',\n",
       " (82, 9469): 'in 1 to 1 orthology relationship with',\n",
       " (82, 2378): 'in 1 to 1 orthology relationship with',\n",
       " (82, 6582): 'in orthology relationship with',\n",
       " (82, 1642): 'in 1 to 1 orthology relationship with',\n",
       " (82, 8981): 'in 1 to 1 orthology relationship with',\n",
       " (82, 8181): 'in 1 to 1 orthology relationship with',\n",
       " (84, 2688): 'targets',\n",
       " (85, 8401): 'interacts with',\n",
       " (85, 3231): 'involved in',\n",
       " (85, 8595): 'interacts with',\n",
       " (85, 2666): 'in 1 to 1 orthology relationship with',\n",
       " (85, 9777): 'has phenotype',\n",
       " (85, 3848): 'has phenotype',\n",
       " (85, 8297): 'has phenotype',\n",
       " (85, 2919): 'has phenotype',\n",
       " (85, 7243): 'in 1 to 1 orthology relationship with',\n",
       " (85, 9475): 'interacts with',\n",
       " (85, 863): 'has phenotype',\n",
       " (85, 8456): 'has phenotype',\n",
       " (85, 8240): 'interacts with',\n",
       " (85, 2034): 'has phenotype',\n",
       " (85, 6937): 'has phenotype',\n",
       " (85, 7326): 'interacts with',\n",
       " (85, 3643): 'in orthology relationship with',\n",
       " (85, 8426): 'has phenotype',\n",
       " (85, 2654): 'has phenotype',\n",
       " (85, 6183): 'in orthology relationship with',\n",
       " (85, 6517): 'in 1 to 1 orthology relationship with',\n",
       " (85, 9036): 'in 1 to 1 orthology relationship with',\n",
       " (85, 5022): 'has phenotype',\n",
       " (85, 3625): 'interacts with',\n",
       " (85, 4239): 'has phenotype',\n",
       " (85, 9727): 'in orthology relationship with',\n",
       " (85, 8340): 'has phenotype',\n",
       " (85, 4736): 'in 1 to 1 orthology relationship with',\n",
       " (85, 2168): 'interacts with',\n",
       " (85, 8469): 'interacts with',\n",
       " (85, 4077): 'in 1 to 1 orthology relationship with',\n",
       " (85, 3356): 'interacts with',\n",
       " (85, 2811): 'interacts with',\n",
       " (85, 1168): 'in orthology relationship with',\n",
       " (85, 6210): 'interacts with',\n",
       " (85, 4839): 'has phenotype',\n",
       " (85, 2198): 'interacts with',\n",
       " (85, 3595): 'has phenotype',\n",
       " (85, 5768): 'in orthology relationship with',\n",
       " (85, 7570): 'interacts with',\n",
       " (85, 9662): 'has phenotype',\n",
       " (85, 2171): 'has phenotype',\n",
       " (85, 4322): 'in 1 to 1 orthology relationship with',\n",
       " (85, 3230): 'interacts with',\n",
       " (85, 2871): 'in orthology relationship with',\n",
       " (85, 1541): 'in 1 to 1 orthology relationship with',\n",
       " (85, 3488): 'interacts with',\n",
       " (85, 1592): 'has phenotype',\n",
       " (85, 9823): 'has phenotype',\n",
       " (85, 2038): 'in 1 to 1 orthology relationship with',\n",
       " (85, 705): 'in 1 to 1 orthology relationship with',\n",
       " (85, 7904): 'has phenotype',\n",
       " (85, 1609): 'interacts with',\n",
       " (85, 7644): 'in 1 to 1 orthology relationship with',\n",
       " (85, 9756): 'interacts with',\n",
       " (85, 1286): 'interacts with',\n",
       " (85, 1379): 'has phenotype',\n",
       " (85, 8823): 'interacts with',\n",
       " (85, 6784): 'in 1 to 1 orthology relationship with',\n",
       " (85, 6100): 'has phenotype',\n",
       " (85, 5392): 'interacts with',\n",
       " (85, 6852): 'interacts with',\n",
       " (85, 3425): 'has phenotype',\n",
       " (85, 1828): 'interacts with',\n",
       " (85, 9372): 'interacts with',\n",
       " (85, 1825): 'interacts with',\n",
       " (85, 737): 'interacts with',\n",
       " (85, 8862): 'has phenotype',\n",
       " (85, 4407): 'has phenotype',\n",
       " (85, 1598): 'has phenotype',\n",
       " (85, 951): 'in 1 to 1 orthology relationship with',\n",
       " (85, 5848): 'in 1 to 1 orthology relationship with',\n",
       " (85, 3122): 'interacts with',\n",
       " (85, 31): 'has phenotype',\n",
       " (85, 738): 'has phenotype',\n",
       " (85, 3449): 'has phenotype',\n",
       " (86, 4076): 'targets',\n",
       " (87, 3696): 'in 1 to 1 orthology relationship with',\n",
       " (87, 7048): 'in 1 to 1 orthology relationship with',\n",
       " (87, 4855): 'in 1 to 1 orthology relationship with',\n",
       " (87, 6996): 'in 1 to 1 orthology relationship with',\n",
       " (87, 3823): 'in 1 to 1 orthology relationship with',\n",
       " (87, 7142): 'in 1 to 1 orthology relationship with',\n",
       " (87, 405): 'in 1 to 1 orthology relationship with',\n",
       " (87, 7283): 'in 1 to 1 orthology relationship with',\n",
       " (87, 7703): 'in 1 to 1 orthology relationship with',\n",
       " (87, 2427): 'in 1 to 1 orthology relationship with',\n",
       " (87, 4783): 'in 1 to 1 orthology relationship with',\n",
       " (87, 5683): 'in orthology relationship with',\n",
       " (87, 1964): 'in 1 to 1 orthology relationship with',\n",
       " (87, 2586): 'in 1 to 1 orthology relationship with',\n",
       " (87, 5473): 'in orthology relationship with',\n",
       " (87, 6188): 'in 1 to 1 orthology relationship with',\n",
       " (87, 2727): 'in 1 to 1 orthology relationship with',\n",
       " (87, 7826): 'in 1 to 1 orthology relationship with',\n",
       " (88, 5026): 'in orthology relationship with',\n",
       " (88, 5634): 'in orthology relationship with',\n",
       " (88, 9712): 'in orthology relationship with',\n",
       " (88, 8765): 'in 1 to 1 orthology relationship with',\n",
       " (88, 8800): 'in 1 to 1 orthology relationship with',\n",
       " (88, 1204): 'in orthology relationship with',\n",
       " (88, 9666): 'in orthology relationship with',\n",
       " (88, 4561): 'in 1 to 1 orthology relationship with',\n",
       " (88, 9629): 'in 1 to 1 orthology relationship with',\n",
       " (88, 9631): 'in orthology relationship with',\n",
       " (88, 808): 'in 1 to 1 orthology relationship with',\n",
       " (88, 3743): 'in orthology relationship with',\n",
       " (88, 8947): 'in 1 to 1 orthology relationship with',\n",
       " (88, 1040): 'in 1 to 1 orthology relationship with',\n",
       " (88, 10005): 'in 1 to 1 orthology relationship with',\n",
       " (88, 3932): 'in 1 to 1 orthology relationship with',\n",
       " (88, 3627): 'in orthology relationship with',\n",
       " (88, 3399): 'in 1 to 1 orthology relationship with',\n",
       " (88, 4437): 'in orthology relationship with',\n",
       " (89, 3972): 'targets',\n",
       " (93, 9896): 'in 1 to 1 orthology relationship with',\n",
       " (93, 8739): 'in 1 to 1 orthology relationship with',\n",
       " (93, 6080): 'in 1 to 1 orthology relationship with',\n",
       " (93, 7589): 'in orthology relationship with',\n",
       " (93, 6923): 'in 1 to 1 orthology relationship with',\n",
       " (93, 8827): 'in 1 to 1 orthology relationship with',\n",
       " (93, 2861): 'in 1 to 1 orthology relationship with',\n",
       " (93, 574): 'in 1 to 1 orthology relationship with',\n",
       " (93, 4849): 'in 1 to 1 orthology relationship with',\n",
       " (93, 2300): 'in orthology relationship with',\n",
       " (93, 8699): 'in 1 to 1 orthology relationship with',\n",
       " (93, 3633): 'in 1 to 1 orthology relationship with',\n",
       " (97, 565): 'targets',\n",
       " (99, 8432): 'targets',\n",
       " (100, 2155): 'in 1 to 1 orthology relationship with',\n",
       " (100, 1223): 'in 1 to 1 orthology relationship with',\n",
       " (100, 4231): 'in 1 to 1 orthology relationship with',\n",
       " (100, 5679): 'in 1 to 1 orthology relationship with',\n",
       " (100, 2616): 'in 1 to 1 orthology relationship with',\n",
       " (100, 9912): 'in 1 to 1 orthology relationship with',\n",
       " (100, 8323): 'in 1 to 1 orthology relationship with',\n",
       " (100, 8022): 'in 1 to 1 orthology relationship with',\n",
       " (100, 1022): 'in 1 to 1 orthology relationship with',\n",
       " (100, 8805): 'in 1 to 1 orthology relationship with',\n",
       " (100, 1237): 'in 1 to 1 orthology relationship with',\n",
       " (100, 6229): 'in 1 to 1 orthology relationship with',\n",
       " (100, 445): 'in 1 to 1 orthology relationship with',\n",
       " (100, 2544): 'in 1 to 1 orthology relationship with',\n",
       " (100, 2039): 'in 1 to 1 orthology relationship with',\n",
       " (100, 8893): 'in 1 to 1 orthology relationship with',\n",
       " (100, 3980): 'in 1 to 1 orthology relationship with',\n",
       " (100, 8752): 'in 1 to 1 orthology relationship with',\n",
       " (100, 8103): 'in 1 to 1 orthology relationship with',\n",
       " (102, 2688): 'targets',\n",
       " (104, 3972): 'targets',\n",
       " (105, 6612): 'has affected feature',\n",
       " (105, 6315): 'pathogenic for condition',\n",
       " (106, 2592): 'pathogenic for condition',\n",
       " (106, 6612): 'has affected feature',\n",
       " (109, 6612): 'has affected feature',\n",
       " (109, 6315): 'likely pathogenic for condition',\n",
       " (112, 3972): 'targets',\n",
       " (113, 6612): 'has affected feature',\n",
       " (113, 6315): 'pathogenic for condition',\n",
       " (114, 9475): 'in orthology relationship with',\n",
       " (114, 7398): 'in orthology relationship with',\n",
       " (114, 2091): 'in orthology relationship with',\n",
       " (114, 6533): 'in orthology relationship with',\n",
       " (114, 8457): 'in orthology relationship with',\n",
       " (114, 4245): 'in orthology relationship with',\n",
       " (114, 5605): 'in orthology relationship with',\n",
       " (114, 7605): 'in orthology relationship with',\n",
       " (114, 7660): 'in orthology relationship with',\n",
       " (114, 6573): 'in orthology relationship with',\n",
       " (114, 2526): 'in orthology relationship with',\n",
       " (114, 9866): 'in orthology relationship with',\n",
       " (114, 9945): 'in orthology relationship with',\n",
       " (114, 3788): 'in orthology relationship with',\n",
       " (116, 6447): 'in orthology relationship with',\n",
       " (116, 7743): 'in orthology relationship with',\n",
       " (116, 3552): 'in orthology relationship with',\n",
       " (116, 2285): 'in orthology relationship with',\n",
       " (116, 8094): 'in orthology relationship with',\n",
       " (116, 10008): 'in orthology relationship with',\n",
       " (116, 527): 'in orthology relationship with',\n",
       " (116, 8814): 'in orthology relationship with',\n",
       " (116, 7108): 'in orthology relationship with',\n",
       " (116, 692): 'in orthology relationship with',\n",
       " (116, 7116): 'in orthology relationship with',\n",
       " (116, 1614): 'in orthology relationship with',\n",
       " (116, 7300): 'in orthology relationship with',\n",
       " (116, 6709): 'in orthology relationship with',\n",
       " (116, 6982): 'in orthology relationship with',\n",
       " (116, 9836): 'in orthology relationship with',\n",
       " (116, 2582): 'in orthology relationship with',\n",
       " (116, 3601): 'in orthology relationship with',\n",
       " (116, 3378): 'in orthology relationship with',\n",
       " (116, 1344): 'in orthology relationship with',\n",
       " (116, 3661): 'in orthology relationship with',\n",
       " (116, 1972): 'in orthology relationship with',\n",
       " (116, 2970): 'in orthology relationship with',\n",
       " (116, 3270): 'in orthology relationship with',\n",
       " (116, 1863): 'in orthology relationship with',\n",
       " (117, 8501): 'targets',\n",
       " (118, 9759): 'targets',\n",
       " (119, 5859): 'enables',\n",
       " (119, 9487): 'in 1 to 1 orthology relationship with',\n",
       " (119, 6850): 'in 1 to 1 orthology relationship with',\n",
       " (119, 3831): 'in orthology relationship with',\n",
       " (119, 2479): 'in 1 to 1 orthology relationship with',\n",
       " (119, 2083): 'in 1 to 1 orthology relationship with',\n",
       " (119, 1515): 'in 1 to 1 orthology relationship with',\n",
       " (119, 6459): 'in 1 to 1 orthology relationship with',\n",
       " (119, 9409): 'in 1 to 1 orthology relationship with',\n",
       " (119, 8693): 'in 1 to 1 orthology relationship with',\n",
       " (119, 9155): 'is part of',\n",
       " (119, 7184): 'in 1 to 1 orthology relationship with',\n",
       " (119, 8746): 'in 1 to 1 orthology relationship with',\n",
       " (119, 2739): 'in 1 to 1 orthology relationship with',\n",
       " (119, 5411): 'in 1 to 1 orthology relationship with',\n",
       " (119, 3108): 'in 1 to 1 orthology relationship with',\n",
       " (119, 9890): 'in 1 to 1 orthology relationship with',\n",
       " (120, 2179): 'in orthology relationship with',\n",
       " (120, 5253): 'in orthology relationship with',\n",
       " (120, 7518): 'in orthology relationship with',\n",
       " (120, 6630): 'in orthology relationship with',\n",
       " (120, 8709): 'in orthology relationship with',\n",
       " (120, 2496): 'in orthology relationship with',\n",
       " (120, 8851): 'enables',\n",
       " (120, 3563): 'in orthology relationship with',\n",
       " (120, 2142): 'in orthology relationship with',\n",
       " (120, 6634): 'in orthology relationship with',\n",
       " (120, 4376): 'in orthology relationship with',\n",
       " (120, 3156): 'in orthology relationship with',\n",
       " (120, 473): 'in orthology relationship with',\n",
       " (120, 7960): 'in orthology relationship with',\n",
       " (120, 7731): 'in orthology relationship with',\n",
       " (120, 9906): 'in orthology relationship with',\n",
       " (120, 3922): 'in orthology relationship with',\n",
       " (120, 9292): 'in orthology relationship with',\n",
       " (120, 9543): 'in orthology relationship with',\n",
       " (120, 4612): 'in orthology relationship with',\n",
       " (120, 7165): 'in orthology relationship with',\n",
       " (120, 4465): 'in orthology relationship with',\n",
       " (120, 6218): 'in orthology relationship with',\n",
       " (120, 9155): 'is part of',\n",
       " (120, 2711): 'in orthology relationship with',\n",
       " (120, 8688): 'in orthology relationship with',\n",
       " (120, 6478): 'in orthology relationship with',\n",
       " (120, 6909): 'in orthology relationship with',\n",
       " (120, 1677): 'in orthology relationship with',\n",
       " (120, 3514): 'in orthology relationship with',\n",
       " (120, 3757): 'in orthology relationship with',\n",
       " (120, 9798): 'is part of',\n",
       " (121, 6420): 'in orthology relationship with',\n",
       " (121, 8927): 'in orthology relationship with',\n",
       " (121, 1473): 'in orthology relationship with',\n",
       " (121, 3591): 'in orthology relationship with',\n",
       " (121, 7570): 'in orthology relationship with',\n",
       " (121, 5967): 'expressed in',\n",
       " (121, 5483): 'expressed in',\n",
       " (121, 4212): 'in orthology relationship with',\n",
       " (121, 4111): 'in orthology relationship with',\n",
       " (121, 914): 'in orthology relationship with',\n",
       " (121, 3693): 'in orthology relationship with',\n",
       " (121, 9328): 'in orthology relationship with',\n",
       " (121, 6355): 'in orthology relationship with',\n",
       " (121, 4226): 'in orthology relationship with',\n",
       " (121, 570): 'in orthology relationship with',\n",
       " (121, 8661): 'in orthology relationship with',\n",
       " (121, 8921): 'in orthology relationship with',\n",
       " (122, 2688): 'targets',\n",
       " (124, 6315): 'pathogenic for condition',\n",
       " (124, 6612): 'has affected feature',\n",
       " (125, 6612): 'has affected feature',\n",
       " (125, 6315): 'pathogenic for condition',\n",
       " (128, 6612): 'has affected feature',\n",
       " (128, 6315): 'pathogenic for condition',\n",
       " (130, 9759): 'targets',\n",
       " (131, 5317): 'targets',\n",
       " (131, 2688): 'targets',\n",
       " (135, 6612): 'has affected feature',\n",
       " (135, 6315): 'pathogenic for condition',\n",
       " (138, 6315): 'likely pathogenic for condition',\n",
       " (138, 6612): 'has affected feature',\n",
       " (145, 2688): 'targets',\n",
       " (146, 6315): 'pathogenic for condition',\n",
       " (146, 6612): 'has affected feature',\n",
       " (147, 3299): 'in 1 to 1 orthology relationship with',\n",
       " (147, 2443): 'in 1 to 1 orthology relationship with',\n",
       " (147, 8459): 'is part of',\n",
       " (147, 3927): 'in orthology relationship with',\n",
       " (147, 8763): 'in 1 to 1 orthology relationship with',\n",
       " (147, 2350): 'in 1 to 1 orthology relationship with',\n",
       " (147, 4569): 'in 1 to 1 orthology relationship with',\n",
       " (147, 3010): 'in 1 to 1 orthology relationship with',\n",
       " (147, 9382): 'in 1 to 1 orthology relationship with',\n",
       " (147, 4602): 'in 1 to 1 orthology relationship with',\n",
       " (147, 489): 'in 1 to 1 orthology relationship with',\n",
       " (147, 7045): 'in 1 to 1 orthology relationship with',\n",
       " (147, 7191): 'in 1 to 1 orthology relationship with',\n",
       " (147, 5168): 'in 1 to 1 orthology relationship with',\n",
       " (147, 2901): 'in 1 to 1 orthology relationship with',\n",
       " (147, 6625): 'in 1 to 1 orthology relationship with',\n",
       " (147, 5859): 'enables',\n",
       " (151, 7372): 'in orthology relationship with',\n",
       " (151, 8432): 'in 1 to 1 orthology relationship with',\n",
       " (151, 9432): 'in orthology relationship with',\n",
       " (151, 1555): 'in 1 to 1 orthology relationship with',\n",
       " (151, 4029): 'in orthology relationship with',\n",
       " (151, 5766): 'in 1 to 1 orthology relationship with',\n",
       " (151, 7119): 'in 1 to 1 orthology relationship with',\n",
       " (151, 4914): 'in 1 to 1 orthology relationship with',\n",
       " (151, 8535): 'in 1 to 1 orthology relationship with',\n",
       " (151, 4017): 'in 1 to 1 orthology relationship with',\n",
       " (151, 6878): 'in 1 to 1 orthology relationship with',\n",
       " (151, 1123): 'in orthology relationship with',\n",
       " (151, 4977): 'in 1 to 1 orthology relationship with',\n",
       " (151, 1612): 'in 1 to 1 orthology relationship with',\n",
       " (151, 681): 'in orthology relationship with',\n",
       " ...}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_labels_dict = dict([((n1, n2), G.edges[(n1,n2)]['edge_label']) for n1, n2 in G.edges])\n",
    "edge_labels_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load trained model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<All keys matched successfully>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from gnn.linkpred_model import LinkPredModel\n",
    "\n",
    "with open(f'output/best_model_{dataset_nr}_args.pkl', 'rb') as f:\n",
    "    loaded_args = pickle.load(f)\n",
    "    \n",
    "best_model = LinkPredModel(len(e2v_embedding['Embedding'][0]), \n",
    "                           loaded_args['hidden_dim'], loaded_args['output_dim'], \n",
    "                           loaded_args['layers'], loaded_args['aggr'], \n",
    "                           loaded_args['dropout'], loaded_args['device']).to(loaded_args['device'])\n",
    "best_model.load_state_dict(torch.load(f'output/best_model_{dataset_nr}.pth'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explain predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch.Size([10034, 64])\n"
     ]
    }
   ],
   "source": [
    "x = torch.Tensor(e2v_embedding['Embedding'])\n",
    "print(x.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch.Size([2, 82899])\n"
     ]
    }
   ],
   "source": [
    "edge_index = torch.Tensor(np.array(G.edges).transpose()).type(torch.int64).long()\n",
    "print(edge_index.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gnn.gnnexplainer import GNNExplainer, visualize_subgraph\n",
    "\n",
    "def explain_edge(node_idx1, node_idx2):\n",
    "    explainer = GNNExplainer(best_model,\n",
    "                             epochs=700, num_hops=1, lr=0.01)\n",
    "    \n",
    "    trigger = False\n",
    "    early_stop = 0\n",
    "    size = 15   # change size of the explanation graph\n",
    "    iterations = 50 # number of times GNNExplainer is executed\n",
    "    \n",
    "    while not trigger:\n",
    "        _, edge_mask = explainer.explain_link(node_idx1=node_idx1, node_idx2=node_idx2,\n",
    "                                              x=x, edge_index=edge_index,\n",
    "                                              G=G)\n",
    "        \n",
    "        G2 = copy.deepcopy(G)\n",
    "        \n",
    "        if size is not None:\n",
    "            limit = edge_mask.sort(descending = True)[0][size]\n",
    "            print('Contribution threshold is', limit)\n",
    "        else:\n",
    "            limit = 0.5\n",
    "        \n",
    "        for indx, edge in enumerate(G.edges): \n",
    "            if edge_mask[indx] < limit:\n",
    "                G2.remove_edge(edge[0], edge[1])\n",
    "        \n",
    "        trigger = nx.has_path(G2, node_idx1, node_idx2)\n",
    "        early_stop += 1\n",
    "        \n",
    "        if early_stop == iterations and not trigger:\n",
    "            print('No good explanation found after {} iterations'.format(early_stop))\n",
    "            trigger = True\n",
    "        elif trigger: \n",
    "            print('Explanation found!')\n",
    "            \n",
    "            plt.figure(figsize=(10, 10))\n",
    "            ax, G_sub = visualize_subgraph([node_idx1, node_idx2], \n",
    "                                           edge_index, edge_mask, \n",
    "                                           nodes=nodes, y=torch.Tensor(nodes.semantic_id), \n",
    "                                           seed=667, num_hops=2, threshold=limit, \n",
    "                                           node_label='label', edge_labels=edge_labels_dict, \n",
    "                                           show_inactive=False, remove_unconnected=True)\n",
    "            plt.show()\n",
    "            \n",
    "            return True\n",
    "        \n",
    "    return False\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Symptom</th>\n",
       "      <th>Candidates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HP:0011675</td>\n",
       "      <td>[2974, 4285, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HP:0002515</td>\n",
       "      <td>[5302, 4187, 4285]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HP:0003236</td>\n",
       "      <td>[2974, 4285, 5345]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HP:0002093</td>\n",
       "      <td>[4285, 5302, 5345]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HP:0003707</td>\n",
       "      <td>[1529, 5345, 4187]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HP:0001256</td>\n",
       "      <td>[4285, 5345, 5302]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HP:0003701</td>\n",
       "      <td>[4285, 5302, 5345]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HP:0003202</td>\n",
       "      <td>[4285, 2835, 624]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HP:0003560</td>\n",
       "      <td>[2612, 624, 4285]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HP:0003391</td>\n",
       "      <td>[5302, 4285, 4187]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HP:0001635</td>\n",
       "      <td>[231, 4285, 4818]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HP:0001328</td>\n",
       "      <td>[5345, 5302, 4187]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HP:0003323</td>\n",
       "      <td>[624, 2612, 2974]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HP:0001371</td>\n",
       "      <td>[5302, 5345, 4187]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HP:0002650</td>\n",
       "      <td>[4187, 5302, 2974]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HP:0003115</td>\n",
       "      <td>[4285, 2974, 5345]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>[4285, 1529, 1942]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HP:0008981</td>\n",
       "      <td>[5302, 4187, 5345]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HP:0001638</td>\n",
       "      <td>[624, 2612, 4285]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HP:0003307</td>\n",
       "      <td>[4285, 5302, 4187]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HP:0000750</td>\n",
       "      <td>[4285, 1529, 969]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HP:0001265</td>\n",
       "      <td>[4285, 1529, 5302]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HP:0001644</td>\n",
       "      <td>[2974, 4187, 5345]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HP:0001270</td>\n",
       "      <td>[5302, 4285, 5345]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HP:0001290</td>\n",
       "      <td>[4285, 5302, 1529]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>[5345, 606, 2974]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HP:0002791</td>\n",
       "      <td>[5345, 5302, 4187]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Symptom          Candidates\n",
       "0   HP:0011675     [2974, 4285, 6]\n",
       "1   HP:0002515  [5302, 4187, 4285]\n",
       "2   HP:0003236  [2974, 4285, 5345]\n",
       "3   HP:0002093  [4285, 5302, 5345]\n",
       "4   HP:0003707  [1529, 5345, 4187]\n",
       "5   HP:0001256  [4285, 5345, 5302]\n",
       "6   HP:0003701  [4285, 5302, 5345]\n",
       "7   HP:0003202   [4285, 2835, 624]\n",
       "8   HP:0003560   [2612, 624, 4285]\n",
       "9   HP:0003391  [5302, 4285, 4187]\n",
       "10  HP:0001635   [231, 4285, 4818]\n",
       "11  HP:0001328  [5345, 5302, 4187]\n",
       "12  HP:0003323   [624, 2612, 2974]\n",
       "13  HP:0001371  [5302, 5345, 4187]\n",
       "14  HP:0002650  [4187, 5302, 2974]\n",
       "15  HP:0003115  [4285, 2974, 5345]\n",
       "16  HP:0001263  [4285, 1529, 1942]\n",
       "17  HP:0008981  [5302, 4187, 5345]\n",
       "18  HP:0001638   [624, 2612, 4285]\n",
       "19  HP:0003307  [4285, 5302, 4187]\n",
       "20  HP:0000750   [4285, 1529, 969]\n",
       "21  HP:0001265  [4285, 1529, 5302]\n",
       "22  HP:0001644  [2974, 4187, 5345]\n",
       "23  HP:0001270  [5302, 4285, 5345]\n",
       "24  HP:0001290  [4285, 5302, 1529]\n",
       "25  HP:0100543   [5345, 606, 2974]\n",
       "26  HP:0002791  [5345, 5302, 4187]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import ast\n",
    "\n",
    "symptoms_drugs = pd.read_csv(f'output/symptom_drugs_{dataset_nr}.csv', converters={1:ast.literal_eval})\n",
    "symptoms_drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "node1:\n",
      "      index_id          id semantic       label  semantic_id\n",
      "9512      9512  HP:0011675     DISO  Arrhythmia            1\n",
      "node2:\n",
      "      index_id    id semantic       label  semantic_id\n",
      "1019      1019  2974     DRUG  doxacurium            2\n",
      "Explain for edge between nodes 9512 and 1019\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Explain edge between nodes 9512 and 1019: 100%|██████████| 700/700 [00:10<00:00, 65.01it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contribution threshold is tensor(0.0086)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Explain edge between nodes 9512 and 1019: 100%|██████████| 700/700 [00:10<00:00, 67.15it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contribution threshold is tensor(0.0086)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Explain edge between nodes 9512 and 1019: 100%|██████████| 700/700 [00:09<00:00, 71.55it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contribution threshold is tensor(0.0086)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Explain edge between nodes 9512 and 1019: 100%|██████████| 700/700 [00:10<00:00, 64.99it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contribution threshold is tensor(0.0086)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Explain edge between nodes 9512 and 1019: 100%|██████████| 700/700 [00:11<00:00, 60.91it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contribution threshold is tensor(0.0086)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Explain edge between nodes 9512 and 1019: 100%|██████████| 700/700 [00:10<00:00, 68.76it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contribution threshold is tensor(0.0086)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Explain edge between nodes 9512 and 1019: 100%|██████████| 700/700 [00:09<00:00, 75.91it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contribution threshold is tensor(0.0086)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Explain edge between nodes 9512 and 1019: 100%|██████████| 700/700 [00:09<00:00, 77.45it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contribution threshold is tensor(0.0086)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Explain edge between nodes 9512 and 1019: 100%|██████████| 700/700 [00:09<00:00, 72.82it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contribution threshold is tensor(0.0086)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[35], line 19\u001b[0m\n\u001b[0;32m     16\u001b[0m \u001b[39mprint\u001b[39m(nodes\u001b[39m.\u001b[39mloc[[node_idx2]])\n\u001b[0;32m     18\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m'\u001b[39m\u001b[39mExplain for edge between nodes\u001b[39m\u001b[39m'\u001b[39m, node_idx1, \u001b[39m'\u001b[39m\u001b[39mand\u001b[39m\u001b[39m'\u001b[39m, node_idx2)\n\u001b[1;32m---> 19\u001b[0m explained \u001b[39m=\u001b[39m explain_edge(node_idx1, node_idx2)\n\u001b[0;32m     21\u001b[0m \u001b[39mif\u001b[39;00m explained:\n\u001b[0;32m     22\u001b[0m     found_explanation\u001b[39m.\u001b[39mappend({\u001b[39m'\u001b[39m\u001b[39msymptom\u001b[39m\u001b[39m'\u001b[39m: symptom_id, \u001b[39m'\u001b[39m\u001b[39mdrug\u001b[39m\u001b[39m'\u001b[39m: drug_id})\n",
      "Cell \u001b[1;32mIn[32], line 27\u001b[0m, in \u001b[0;36mexplain_edge\u001b[1;34m(node_idx1, node_idx2)\u001b[0m\n\u001b[0;32m     25\u001b[0m \u001b[39mfor\u001b[39;00m indx, edge \u001b[39min\u001b[39;00m \u001b[39menumerate\u001b[39m(G\u001b[39m.\u001b[39medges): \n\u001b[0;32m     26\u001b[0m     \u001b[39mif\u001b[39;00m edge_mask[indx] \u001b[39m<\u001b[39m limit:\n\u001b[1;32m---> 27\u001b[0m         G2\u001b[39m.\u001b[39;49mremove_edge(edge[\u001b[39m0\u001b[39;49m], edge[\u001b[39m1\u001b[39;49m])\n\u001b[0;32m     29\u001b[0m trigger \u001b[39m=\u001b[39m nx\u001b[39m.\u001b[39mhas_path(G2, node_idx1, node_idx2)\n\u001b[0;32m     30\u001b[0m early_stop \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "def get_node_idx(id, nodes): \n",
    "    return nodes[nodes['id'] == id].index.values.astype(int)[0]\n",
    "\n",
    "found_explanation = []\n",
    "\n",
    "for ind, edge in symptoms_drugs.iterrows(): \n",
    "    symptom_id = edge['Symptom']\n",
    "    \n",
    "    for drug_id in edge['Candidates']:\n",
    "        node_idx1 = get_node_idx(symptom_id, nodes)\n",
    "        print('node1:')\n",
    "        print(nodes.loc[[node_idx1]])\n",
    "\n",
    "        node_idx2 = get_node_idx(str(drug_id), nodes)\n",
    "        print('node2:')\n",
    "        print(nodes.loc[[node_idx2]])\n",
    "        \n",
    "        explained = explain_edge(node_idx1, node_idx2)\n",
    "        \n",
    "        if explained:\n",
    "            found_explanation.append({'symptom': symptom_id, 'drug': drug_id})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "found_explanation"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xaifo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
