{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import pickle\n",
    "import copy\n",
    "\n",
    "from gensim.models import KeyedVectors\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "\n",
    "from deepsnap.graph import Graph\n",
    "\n",
    "from torch_geometric.nn import SAGEConv\n",
    "from torch_geometric.explain import Explainer, GNNExplainer, ModelConfig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_nr = 2"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Necessary Data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the node features, load the embeddings yielded from Edge2Vec."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Node</th>\n",
       "      <th>Embedding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>[0.4429384, -0.64671946, -0.045503706, 0.04694...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>[0.7495496, -0.5790302, 0.16192597, 0.03131027...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>[-0.21050918, -0.28908953, 0.118277326, -0.162...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>[0.2527442, -0.2967686, 0.29164207, 0.8072669,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>[0.41442072, 0.06997513, 0.4540072, -0.0370618...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10270</th>\n",
       "      <td>10270</td>\n",
       "      <td>[0.21862394, 0.11319974, 0.31551972, 0.1497048...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10271</th>\n",
       "      <td>10271</td>\n",
       "      <td>[0.8747792, -0.7461259, 0.22221777, -0.0425329...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10272</th>\n",
       "      <td>10272</td>\n",
       "      <td>[0.5863251, -0.8054981, 0.27565393, 0.516578, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10273</th>\n",
       "      <td>10273</td>\n",
       "      <td>[0.6474387, -0.38251266, 0.43746328, 0.3413055...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10274</th>\n",
       "      <td>10274</td>\n",
       "      <td>[0.56232107, -0.5135573, -0.036642436, 0.09458...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10275 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Node                                          Embedding\n",
       "0          0  [0.4429384, -0.64671946, -0.045503706, 0.04694...\n",
       "1          1  [0.7495496, -0.5790302, 0.16192597, 0.03131027...\n",
       "2          2  [-0.21050918, -0.28908953, 0.118277326, -0.162...\n",
       "3          3  [0.2527442, -0.2967686, 0.29164207, 0.8072669,...\n",
       "4          4  [0.41442072, 0.06997513, 0.4540072, -0.0370618...\n",
       "...      ...                                                ...\n",
       "10270  10270  [0.21862394, 0.11319974, 0.31551972, 0.1497048...\n",
       "10271  10271  [0.8747792, -0.7461259, 0.22221777, -0.0425329...\n",
       "10272  10272  [0.5863251, -0.8054981, 0.27565393, 0.516578, ...\n",
       "10273  10273  [0.6474387, -0.38251266, 0.43746328, 0.3413055...\n",
       "10274  10274  [0.56232107, -0.5135573, -0.036642436, 0.09458...\n",
       "\n",
       "[10275 rows x 2 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loaded_node_embedding = KeyedVectors.load(f'output/w2v_{dataset_nr}.dvectors', mmap='r')\n",
    "\n",
    "node_embeddings = pd.DataFrame(columns = ['Node', 'Embedding'])\n",
    "for idx, key in enumerate(loaded_node_embedding.index_to_key):\n",
    "    node_embeddings.loc[int(key)] = pd.Series({'Node':int(key), 'Embedding':list(loaded_node_embedding[key])})\n",
    "    \n",
    "node_features = node_embeddings.sort_values('Node')\n",
    "\n",
    "node_features"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For additional information of nodes, load csv file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>MP:0004187</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>cardia bifida</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ZP:0100138</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>muscle tendon junction myotome increased amoun...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1346525</td>\n",
       "      <td>gene</td>\n",
       "      <td>Sgcd</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>OMIM:300377.0044</td>\n",
       "      <td>variant</td>\n",
       "      <td>DMD, LYS770TER</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ZP:0002210</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>posterior lateral line neuromast primordium mi...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10270</th>\n",
       "      <td>10270</td>\n",
       "      <td>ZP:0014934</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>atrioventricular valve development process qua...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10271</th>\n",
       "      <td>10271</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000011207</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000011207</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10272</th>\n",
       "      <td>10272</td>\n",
       "      <td>ENSEMBL:ENSXETG00000039922</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSXETG00000039922</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10273</th>\n",
       "      <td>10273</td>\n",
       "      <td>ENSEMBL:ENSACAG00000010058</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSACAG00000010058</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10274</th>\n",
       "      <td>10274</td>\n",
       "      <td>WBPhenotype:0000019</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>pharyngeal pumping reduced</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10275 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index_id                          id   semantic   \n",
       "0             0                  MP:0004187  phenotype  \\\n",
       "1             1                  ZP:0100138  phenotype   \n",
       "2             2                 MGI:1346525       gene   \n",
       "3             3            OMIM:300377.0044    variant   \n",
       "4             4                  ZP:0002210  phenotype   \n",
       "...         ...                         ...        ...   \n",
       "10270     10270                  ZP:0014934  phenotype   \n",
       "10271     10271  ENSEMBL:ENSCAFG00000011207       gene   \n",
       "10272     10272  ENSEMBL:ENSXETG00000039922       gene   \n",
       "10273     10273  ENSEMBL:ENSACAG00000010058       gene   \n",
       "10274     10274         WBPhenotype:0000019  phenotype   \n",
       "\n",
       "                                                   label  semantic_id  \n",
       "0                                          cardia bifida            9  \n",
       "1      muscle tendon junction myotome increased amoun...            9  \n",
       "2                                                   Sgcd            5  \n",
       "3                                         DMD, LYS770TER           11  \n",
       "4      posterior lateral line neuromast primordium mi...            9  \n",
       "...                                                  ...          ...  \n",
       "10270  atrioventricular valve development process qua...            9  \n",
       "10271                         ENSEMBL:ENSCAFG00000011207            5  \n",
       "10272                         ENSEMBL:ENSXETG00000039922            5  \n",
       "10273                         ENSEMBL:ENSACAG00000010058            5  \n",
       "10274                         pharyngeal pumping reduced            9  \n",
       "\n",
       "[10275 rows x 5 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = pd.read_csv(f'output/indexed_nodes_{dataset_nr}.csv')\n",
    "nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Node</th>\n",
       "      <th>Embedding</th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>[0.4429384, -0.64671946, -0.045503706, 0.04694...</td>\n",
       "      <td>0</td>\n",
       "      <td>MP:0004187</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>cardia bifida</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>[0.7495496, -0.5790302, 0.16192597, 0.03131027...</td>\n",
       "      <td>1</td>\n",
       "      <td>ZP:0100138</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>muscle tendon junction myotome increased amoun...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>[-0.21050918, -0.28908953, 0.118277326, -0.162...</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1346525</td>\n",
       "      <td>gene</td>\n",
       "      <td>Sgcd</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>[0.2527442, -0.2967686, 0.29164207, 0.8072669,...</td>\n",
       "      <td>3</td>\n",
       "      <td>OMIM:300377.0044</td>\n",
       "      <td>variant</td>\n",
       "      <td>DMD, LYS770TER</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>[0.41442072, 0.06997513, 0.4540072, -0.0370618...</td>\n",
       "      <td>4</td>\n",
       "      <td>ZP:0002210</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>posterior lateral line neuromast primordium mi...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10270</th>\n",
       "      <td>10270</td>\n",
       "      <td>[0.21862394, 0.11319974, 0.31551972, 0.1497048...</td>\n",
       "      <td>10270</td>\n",
       "      <td>ZP:0014934</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>atrioventricular valve development process qua...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10271</th>\n",
       "      <td>10271</td>\n",
       "      <td>[0.8747792, -0.7461259, 0.22221777, -0.0425329...</td>\n",
       "      <td>10271</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000011207</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000011207</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10272</th>\n",
       "      <td>10272</td>\n",
       "      <td>[0.5863251, -0.8054981, 0.27565393, 0.516578, ...</td>\n",
       "      <td>10272</td>\n",
       "      <td>ENSEMBL:ENSXETG00000039922</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSXETG00000039922</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10273</th>\n",
       "      <td>10273</td>\n",
       "      <td>[0.6474387, -0.38251266, 0.43746328, 0.3413055...</td>\n",
       "      <td>10273</td>\n",
       "      <td>ENSEMBL:ENSACAG00000010058</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSACAG00000010058</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10274</th>\n",
       "      <td>10274</td>\n",
       "      <td>[0.56232107, -0.5135573, -0.036642436, 0.09458...</td>\n",
       "      <td>10274</td>\n",
       "      <td>WBPhenotype:0000019</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>pharyngeal pumping reduced</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10275 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Node                                          Embedding  index_id   \n",
       "0          0  [0.4429384, -0.64671946, -0.045503706, 0.04694...         0  \\\n",
       "1          1  [0.7495496, -0.5790302, 0.16192597, 0.03131027...         1   \n",
       "2          2  [-0.21050918, -0.28908953, 0.118277326, -0.162...         2   \n",
       "3          3  [0.2527442, -0.2967686, 0.29164207, 0.8072669,...         3   \n",
       "4          4  [0.41442072, 0.06997513, 0.4540072, -0.0370618...         4   \n",
       "...      ...                                                ...       ...   \n",
       "10270  10270  [0.21862394, 0.11319974, 0.31551972, 0.1497048...     10270   \n",
       "10271  10271  [0.8747792, -0.7461259, 0.22221777, -0.0425329...     10271   \n",
       "10272  10272  [0.5863251, -0.8054981, 0.27565393, 0.516578, ...     10272   \n",
       "10273  10273  [0.6474387, -0.38251266, 0.43746328, 0.3413055...     10273   \n",
       "10274  10274  [0.56232107, -0.5135573, -0.036642436, 0.09458...     10274   \n",
       "\n",
       "                               id   semantic   \n",
       "0                      MP:0004187  phenotype  \\\n",
       "1                      ZP:0100138  phenotype   \n",
       "2                     MGI:1346525       gene   \n",
       "3                OMIM:300377.0044    variant   \n",
       "4                      ZP:0002210  phenotype   \n",
       "...                           ...        ...   \n",
       "10270                  ZP:0014934  phenotype   \n",
       "10271  ENSEMBL:ENSCAFG00000011207       gene   \n",
       "10272  ENSEMBL:ENSXETG00000039922       gene   \n",
       "10273  ENSEMBL:ENSACAG00000010058       gene   \n",
       "10274         WBPhenotype:0000019  phenotype   \n",
       "\n",
       "                                                   label  semantic_id  \n",
       "0                                          cardia bifida            9  \n",
       "1      muscle tendon junction myotome increased amoun...            9  \n",
       "2                                                   Sgcd            5  \n",
       "3                                         DMD, LYS770TER           11  \n",
       "4      posterior lateral line neuromast primordium mi...            9  \n",
       "...                                                  ...          ...  \n",
       "10270  atrioventricular valve development process qua...            9  \n",
       "10271                         ENSEMBL:ENSCAFG00000011207            5  \n",
       "10272                         ENSEMBL:ENSXETG00000039922            5  \n",
       "10273                         ENSEMBL:ENSACAG00000010058            5  \n",
       "10274                         pharyngeal pumping reduced            9  \n",
       "\n",
       "[10275 rows x 7 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_features = pd.merge(node_features, nodes,\n",
    "                         left_on='Node', right_on='index_id',\n",
    "                         how='inner')\n",
    "node_features"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the edges, load csv file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>pat-2</td>\n",
       "      <td>5</td>\n",
       "      <td>1542</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00006789</td>\n",
       "      <td>unc-54</td>\n",
       "      <td>5</td>\n",
       "      <td>6544</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000015555</td>\n",
       "      <td>LAMC1</td>\n",
       "      <td>5</td>\n",
       "      <td>9268</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ZFIN:ZDB-GENE-021226-3</td>\n",
       "      <td>lamc1</td>\n",
       "      <td>5</td>\n",
       "      <td>5387</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "      <td>5</td>\n",
       "      <td>2204</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85987</th>\n",
       "      <td>458</td>\n",
       "      <td>scopolamine butylbromide</td>\n",
       "      <td>4</td>\n",
       "      <td>5945</td>\n",
       "      <td>targets</td>\n",
       "      <td>P11229</td>\n",
       "      <td>Muscarinic acetylcholine receptor M1</td>\n",
       "      <td>6</td>\n",
       "      <td>5919</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85988</th>\n",
       "      <td>OMIM:300377.0080</td>\n",
       "      <td>DMD, IVS62, A-G, -285</td>\n",
       "      <td>11</td>\n",
       "      <td>1578</td>\n",
       "      <td>is allele of</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>5</td>\n",
       "      <td>3310</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85989</th>\n",
       "      <td>5297</td>\n",
       "      <td>dacomitinib</td>\n",
       "      <td>4</td>\n",
       "      <td>8798</td>\n",
       "      <td>targets</td>\n",
       "      <td>P12931</td>\n",
       "      <td>Proto-oncogene tyrosine-protein kinase Src</td>\n",
       "      <td>6</td>\n",
       "      <td>2379</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85990</th>\n",
       "      <td>ClinVarVariant:981988</td>\n",
       "      <td>NC_000023.11:g.(31875374_31929595)_(31968515_3...</td>\n",
       "      <td>11</td>\n",
       "      <td>8189</td>\n",
       "      <td>has affected feature</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>5</td>\n",
       "      <td>3310</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85991</th>\n",
       "      <td>ClinVarVariant:584105</td>\n",
       "      <td>NC_000023.11:g.(?_32595737)_(32699313_?)del</td>\n",
       "      <td>11</td>\n",
       "      <td>9866</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85992 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          head   \n",
       "0      WormBase:WBGene00006787  \\\n",
       "1      WormBase:WBGene00006787   \n",
       "2      WormBase:WBGene00006787   \n",
       "3      WormBase:WBGene00006787   \n",
       "4      WormBase:WBGene00006787   \n",
       "...                        ...   \n",
       "85987                      458   \n",
       "85988         OMIM:300377.0080   \n",
       "85989                     5297   \n",
       "85990    ClinVarVariant:981988   \n",
       "85991    ClinVarVariant:584105   \n",
       "\n",
       "                                              label_head  class_head   \n",
       "0                                                 unc-52           5  \\\n",
       "1                                                 unc-52           5   \n",
       "2                                                 unc-52           5   \n",
       "3                                                 unc-52           5   \n",
       "4                                                 unc-52           5   \n",
       "...                                                  ...         ...   \n",
       "85987                           scopolamine butylbromide           4   \n",
       "85988                              DMD, IVS62, A-G, -285          11   \n",
       "85989                                        dacomitinib           4   \n",
       "85990  NC_000023.11:g.(31875374_31929595)_(31968515_3...          11   \n",
       "85991        NC_000023.11:g.(?_32595737)_(32699313_?)del          11   \n",
       "\n",
       "       index_head                        relation                        tail   \n",
       "0             304                  interacts with     WormBase:WBGene00003929  \\\n",
       "1             304                  interacts with     WormBase:WBGene00006789   \n",
       "2             304  in orthology relationship with  ENSEMBL:ENSSSCG00000015555   \n",
       "3             304  in orthology relationship with      ZFIN:ZDB-GENE-021226-3   \n",
       "4             304  in orthology relationship with  ENSEMBL:ENSOANG00000001050   \n",
       "...           ...                             ...                         ...   \n",
       "85987        5945                         targets                      P11229   \n",
       "85988        1578                    is allele of                   HGNC:2928   \n",
       "85989        8798                         targets                      P12931   \n",
       "85990        8189            has affected feature                   HGNC:2928   \n",
       "85991        9866                causes condition               MONDO:0010679   \n",
       "\n",
       "                                       label_tail  class_tail  index_tail   \n",
       "0                                           pat-2           5        1542  \\\n",
       "1                                          unc-54           5        6544   \n",
       "2                                           LAMC1           5        9268   \n",
       "3                                           lamc1           5        5387   \n",
       "4                      ENSEMBL:ENSOANG00000001050           5        2204   \n",
       "...                                           ...         ...         ...   \n",
       "85987        Muscarinic acetylcholine receptor M1           6        5919   \n",
       "85988                                         DMD           5        3310   \n",
       "85989  Proto-oncogene tyrosine-protein kinase Src           6        2379   \n",
       "85990                                         DMD           5        3310   \n",
       "85991                 Duchenne muscular dystrophy           3        1913   \n",
       "\n",
       "       type  \n",
       "0         0  \n",
       "1         0  \n",
       "2         1  \n",
       "3         1  \n",
       "4         1  \n",
       "...     ...  \n",
       "85987    17  \n",
       "85988    15  \n",
       "85989    17  \n",
       "85990    11  \n",
       "85991     2  \n",
       "\n",
       "[85992 rows x 10 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = pd.read_csv(f'output/indexed_edges_{dataset_nr}.csv')\n",
    "edges"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load nodes with their features and edges into graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, {'node_feature': tensor([ 0.4429, -0.6467, -0.0455,  0.0469,  0.3007, -0.4822,  0.2423,  0.0707,\n",
      "        -0.5349,  0.4175,  0.7552,  0.0628, -0.3813, -0.3389,  0.5671,  0.4919,\n",
      "         0.0854,  0.0053, -0.2314,  0.8349,  0.5443,  0.7359, -0.1399, -0.2049,\n",
      "        -0.1793,  0.0314, -0.7152,  0.1764,  0.3586, -0.4770,  0.0838, -0.3013,\n",
      "         0.1751, -0.7785, -0.1280,  0.4082,  0.5959,  0.4058,  0.5182,  0.5561,\n",
      "         0.3995, -0.1704,  0.3317, -0.1183, -0.1047,  0.2789,  0.3293, -0.1368,\n",
      "         0.1259,  0.1704,  0.7724,  0.4460,  0.6070, -0.3315,  0.5776,  0.4367,\n",
      "         0.2875, -0.7309,  0.2535,  0.5972,  0.4410, -0.4248, -0.3355,  0.3258])})\n",
      "(2, 7693, {})\n",
      "Number of edges is 85878\n",
      "Number of nodes is 10275\n"
     ]
    }
   ],
   "source": [
    "G = nx.DiGraph()\n",
    "for ind, node in node_features.iterrows(): \n",
    "    G.add_node(int(node['Node']), node_feature = torch.Tensor(node['Embedding']))\n",
    "for ind, edge in edges.iterrows(): \n",
    "    G.add_edge(int(edge['index_head']), int(edge['index_tail']))\n",
    "  \n",
    "for node in G.nodes(data=True):\n",
    "    print(node)\n",
    "    break\n",
    "for edge in G.edges(data=True):\n",
    "    print(edge)\n",
    "    break\n",
    "\n",
    "print(\"Number of edges is {}\".format(G.number_of_edges()))\n",
    "print(\"Number of nodes is {}\".format(G.number_of_nodes()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, {'node_label': 'cardia bifida', 'node_semantic': 'phenotype'})\n",
      "(2, 7693, {'edge_label': 'found in'})\n",
      "Number of edges is 85878\n",
      "Number of nodes is 10275\n"
     ]
    }
   ],
   "source": [
    "G_attributes = nx.DiGraph()\n",
    "for ind, node in node_features.iterrows(): \n",
    "    G_attributes.add_node(int(node['Node']), node_label = node['label'], node_semantic = node['semantic'])\n",
    "for ind, edge in edges.iterrows(): \n",
    "    G_attributes.add_edge(int(edge['index_head']), int(edge['index_tail']), edge_label = edge['relation'])\n",
    "  \n",
    "for node in G_attributes.nodes(data=True):\n",
    "    print(node)\n",
    "    break\n",
    "for edge in G_attributes.edges(data=True):\n",
    "    print(edge)\n",
    "    break\n",
    "\n",
    "print(\"Number of edges is {}\".format(G_attributes.number_of_edges()))\n",
    "print(\"Number of nodes is {}\".format(G_attributes.number_of_nodes()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch.Size([10275, 64])\n",
      "torch.Size([2, 85878])\n"
     ]
    }
   ],
   "source": [
    "DeepG = Graph(G)\n",
    "\n",
    "print(DeepG.node_feature.shape)\n",
    "print(DeepG.edge_index.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[ 0.4429, -0.6467, -0.0455,  ..., -0.4248, -0.3355,  0.3258],\n",
       "        [ 0.7495, -0.5790,  0.1619,  ..., -0.4995,  0.4305,  0.0857],\n",
       "        [-0.2105, -0.2891,  0.1183,  ...,  0.3423, -0.0471,  0.2484],\n",
       "        ...,\n",
       "        [ 0.5863, -0.8055,  0.2757,  ..., -0.5787, -0.5317, -0.0063],\n",
       "        [ 0.6474, -0.3825,  0.4375,  ...,  0.3179, -0.5550, -0.1071],\n",
       "        [ 0.5623, -0.5136, -0.0366,  ..., -0.7215, -0.4545, -0.0579]])"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = DeepG.node_feature\n",
    "x"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load arguments and parameters of trained GNN model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<All keys matched successfully>"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(f'output/best_model_{dataset_nr}_args.pkl', 'rb') as f:\n",
    "    loaded_args = pickle.load(f)\n",
    "\n",
    "# Important! Changed the forward function such that it allows input of less edges that are predicted\n",
    "# and it does not return newly calculated x\n",
    "\n",
    "class LinkPredModel(torch.nn.Module):\n",
    "    \"\"\"\n",
    "        Architecture contains Batch Normalization layers (https://towardsdatascience.com/batch-norm-explained-visually-how-it-works-and-why-neural-networks-need-it-b18919692739) \n",
    "        between the SAGEConvolutional layers.\n",
    "    \"\"\"\n",
    "    def __init__(self, input_size, hidden_size, out_size, num_layers, aggr, dropout, device):\n",
    "        super(LinkPredModel, self).__init__()\n",
    "\n",
    "        self.convs = torch.nn.ModuleList()\n",
    "        self.bns = torch.nn.ModuleList()\n",
    "        \n",
    "        self.convs.append(SAGEConv(input_size, hidden_size, normalize=True, aggr=aggr)) # input node embedding features, output is size of hidden layers\n",
    "        self.bns.append(nn.BatchNorm1d(hidden_size))\n",
    "        for _ in range(num_layers - 2):\n",
    "            self.convs.append(SAGEConv(hidden_size, hidden_size, normalize=True, aggr=aggr))\n",
    "            self.bns.append(nn.BatchNorm1d(hidden_size))\n",
    "        self.convs.append(SAGEConv(hidden_size, out_size, normalize=True, aggr=aggr))\n",
    "        self.bns.append(nn.BatchNorm1d(out_size))\n",
    "\n",
    "        self.dropout = dropout\n",
    "        self.loss_fn = torch.nn.BCEWithLogitsLoss()\n",
    "        \n",
    "        self.device = device\n",
    "\n",
    "\n",
    "    def forward(self, x, edge_index, edge_label_index, training = False):\n",
    "        x = x.to(self.device)\n",
    "        edge_index = edge_index.to(self.device)\n",
    "        edge_label_index = edge_label_index.to(self.device)\n",
    "\n",
    "        for i in range(len(self.convs) - 1):\n",
    "          x = self.convs[i](x, edge_index)\n",
    "          x = self.bns[i](x)\n",
    "          x = F.leaky_relu(x)\n",
    "          x = F.dropout(x, p=self.dropout, training=training)\n",
    "        \n",
    "        x = self.convs[-1](x, edge_index)  \n",
    "        x = self.bns[-1](x)\n",
    "        \n",
    "\n",
    "        nodes_first, nodes_second = edge_label_index\n",
    "        nodes_first_features = x[nodes_first.long()]\n",
    "        nodes_second_features = x[nodes_second.long()]\n",
    "        \n",
    "        pred = torch.sum(nodes_first_features * nodes_second_features, dim=-1)\n",
    "        \n",
    "        return pred\n",
    "    \n",
    "    def loss(self, pred, label):\n",
    "        return self.loss_fn(pred, label)\n",
    "\n",
    "num_node_features = len(node_features['Embedding'][0])\n",
    "\n",
    "best_model = LinkPredModel(num_node_features, loaded_args['hidden_dim'], loaded_args['output_dim'], \n",
    "                           loaded_args['layers'], loaded_args['aggr'], loaded_args['dropout'], \n",
    "                           loaded_args['device']).to(loaded_args['device'])\n",
    "\n",
    "best_model.load_state_dict(torch.load(f'output/best_model_{dataset_nr}.pth'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GNNExplainer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_config = ModelConfig(\n",
    "    mode='binary_classification',\n",
    "    task_level='edge',\n",
    "    return_type='raw',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "explainer = Explainer(\n",
    "    model=best_model,\n",
    "    explanation_type='model',\n",
    "    algorithm=GNNExplainer(epochs=200),\n",
    "    node_mask_type='attributes',\n",
    "    edge_mask_type='object',\n",
    "    model_config=model_config,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Symptom</th>\n",
       "      <th>Candidates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HP:0003707</td>\n",
       "      <td>[2774, 176, 1043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HP:0003236</td>\n",
       "      <td>[1043, 176, 1971]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HP:0001256</td>\n",
       "      <td>[65, 1273, 1971]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HP:0001265</td>\n",
       "      <td>[1273, 1971, 176]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HP:0001290</td>\n",
       "      <td>[1971, 1043, 2630]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HP:0003307</td>\n",
       "      <td>[176, 2774, 1043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HP:0003202</td>\n",
       "      <td>[1273, 1043, 2774]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>[1273, 1150, 832]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HP:0002791</td>\n",
       "      <td>[1273, 1646, 271]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HP:0001371</td>\n",
       "      <td>[1043, 522, 1273]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HP:0003701</td>\n",
       "      <td>[1273, 1971, 176]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HP:0001270</td>\n",
       "      <td>[1273, 176, 1646]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HP:0003560</td>\n",
       "      <td>[1043, 2774, 1273]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HP:0008981</td>\n",
       "      <td>[176, 2774, 5345]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HP:0002650</td>\n",
       "      <td>[1043, 1273, 5345]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HP:0001638</td>\n",
       "      <td>[176, 2774, 1273]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HP:0003323</td>\n",
       "      <td>[223, 176, 1971]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HP:0001635</td>\n",
       "      <td>[1576, 522, 313]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HP:0000750</td>\n",
       "      <td>[1273, 1646, 1920]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HP:0003115</td>\n",
       "      <td>[1726, 1043, 223]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HP:0011675</td>\n",
       "      <td>[231, 2774, 65]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HP:0003391</td>\n",
       "      <td>[2774, 176, 1273]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HP:0002093</td>\n",
       "      <td>[1971, 1273, 65]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>[1273, 1043, 1646]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HP:0001644</td>\n",
       "      <td>[176, 223, 1043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HP:0002515</td>\n",
       "      <td>[176, 1273, 1043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HP:0001328</td>\n",
       "      <td>[65, 1273, 1043]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Symptom          Candidates\n",
       "0   HP:0003707   [2774, 176, 1043]\n",
       "1   HP:0003236   [1043, 176, 1971]\n",
       "2   HP:0001256    [65, 1273, 1971]\n",
       "3   HP:0001265   [1273, 1971, 176]\n",
       "4   HP:0001290  [1971, 1043, 2630]\n",
       "5   HP:0003307   [176, 2774, 1043]\n",
       "6   HP:0003202  [1273, 1043, 2774]\n",
       "7   HP:0001263   [1273, 1150, 832]\n",
       "8   HP:0002791   [1273, 1646, 271]\n",
       "9   HP:0001371   [1043, 522, 1273]\n",
       "10  HP:0003701   [1273, 1971, 176]\n",
       "11  HP:0001270   [1273, 176, 1646]\n",
       "12  HP:0003560  [1043, 2774, 1273]\n",
       "13  HP:0008981   [176, 2774, 5345]\n",
       "14  HP:0002650  [1043, 1273, 5345]\n",
       "15  HP:0001638   [176, 2774, 1273]\n",
       "16  HP:0003323    [223, 176, 1971]\n",
       "17  HP:0001635    [1576, 522, 313]\n",
       "18  HP:0000750  [1273, 1646, 1920]\n",
       "19  HP:0003115   [1726, 1043, 223]\n",
       "20  HP:0011675     [231, 2774, 65]\n",
       "21  HP:0003391   [2774, 176, 1273]\n",
       "22  HP:0002093    [1971, 1273, 65]\n",
       "23  HP:0100543  [1273, 1043, 1646]\n",
       "24  HP:0001644    [176, 223, 1043]\n",
       "25  HP:0002515   [176, 1273, 1043]\n",
       "26  HP:0001328    [65, 1273, 1043]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import ast\n",
    "\n",
    "symptoms_drugs = pd.read_csv(f'output/symptom_drugs_{dataset_nr}.csv', converters={1:ast.literal_eval})\n",
    "symptoms_drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "def explain_edge(node_idx1, node_idx2):\n",
    "    edge_label_index = torch.tensor([node_idx1, node_idx2]) # Create a tensor as input of the model. Only predict the label of a single edge instead of all edges in current dataset.\n",
    "    \n",
    "    done = False\n",
    "    current_iterations = 0\n",
    "    total_iterations = 30\n",
    "    graph_size = 30\n",
    "\n",
    "    while not done:\n",
    "        print(f'--- Iteration {current_iterations} ---')\n",
    "\n",
    "        explanation = explainer(\n",
    "            x=DeepG.node_feature,\n",
    "            edge_index=DeepG.edge_index,\n",
    "            edge_label_index=edge_label_index,\n",
    "        )\n",
    "        \n",
    "        edge_mask = explanation.edge_mask\n",
    "        \n",
    "        print(f'A total of {torch.count_nonzero(edge_mask)} edges contributing to prediction.')\n",
    "        \n",
    "        sorted_contributions = edge_mask.sort(descending=True)[0]\n",
    "        print(f'Maximum contribution value is {sorted_contributions[0]}')\n",
    "        limit = sorted_contributions[graph_size]\n",
    "        print(f'Based on desired graph size, contribution limit is set to {limit}')\n",
    "        \n",
    "        pruned_G = copy.deepcopy(G_attributes)\n",
    "        for indx, edge in enumerate(G.edges):\n",
    "            if edge_mask[indx] < limit:\n",
    "                pruned_G.remove_edge(edge[0], edge[1])\n",
    "        \n",
    "        done = nx.has_path(pruned_G, node_idx1, node_idx2)\n",
    "        \n",
    "        pruned_G.remove_nodes_from(list(nx.isolates(pruned_G)))\n",
    "        print('Pruned graph has edge total of:', pruned_G.number_of_edges())\n",
    "        print('Pruned graph has node total of:', pruned_G.number_of_nodes())\n",
    "        \n",
    "        current_iterations += 1\n",
    "        \n",
    "        if done:\n",
    "            print(f'A good explanation has been found!')\n",
    "            nx.write_gexf(pruned_G, \"output/explanation.gexf\")\n",
    "            return True\n",
    "            \n",
    "        elif current_iterations == total_iterations and not done:\n",
    "            done = True\n",
    "            print(f'No good explanations found after {current_iterations} iterations...')\n",
    "    \n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Explain for edge between nodes 1351 and 7315\n",
      "node1:\n",
      "      index_id          id   semantic                          label   \n",
      "5727      5727  HP:0003707  phenotype  Calf muscle pseudohypertrophy  \\\n",
      "\n",
      "      semantic_id  \n",
      "5727            9  \n",
      "node2:\n",
      "      index_id    id semantic        label  semantic_id\n",
      "9663      9663  2774     drug  tropicamide            4\n",
      "--- Iteration 0 ---\n",
      "A total of 4260 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17691539227962494\n",
      "Based on desired graph size, contribution limit is set to 0.14854896068572998\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 39\n",
      "--- Iteration 1 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17125378549098969\n",
      "Based on desired graph size, contribution limit is set to 0.14849451184272766\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 2 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.16578319668769836\n",
      "Based on desired graph size, contribution limit is set to 0.14848966896533966\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 3 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.165847510099411\n",
      "Based on desired graph size, contribution limit is set to 0.14839567244052887\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 4 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1772894561290741\n",
      "Based on desired graph size, contribution limit is set to 0.1485842615365982\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 5 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17121239006519318\n",
      "Based on desired graph size, contribution limit is set to 0.14844630658626556\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 6 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.16789506375789642\n",
      "Based on desired graph size, contribution limit is set to 0.1484932005405426\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 7 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17275497317314148\n",
      "Based on desired graph size, contribution limit is set to 0.14872944355010986\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 8 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17815163731575012\n",
      "Based on desired graph size, contribution limit is set to 0.1484563946723938\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 9 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17751052975654602\n",
      "Based on desired graph size, contribution limit is set to 0.14852192997932434\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 10 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1691892147064209\n",
      "Based on desired graph size, contribution limit is set to 0.14860963821411133\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 11 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17634932696819305\n",
      "Based on desired graph size, contribution limit is set to 0.1482432335615158\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 12 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.16694007813930511\n",
      "Based on desired graph size, contribution limit is set to 0.1483328938484192\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 13 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17112864553928375\n",
      "Based on desired graph size, contribution limit is set to 0.14842896163463593\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 14 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1572033315896988\n",
      "Based on desired graph size, contribution limit is set to 0.1484317034482956\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 36\n",
      "--- Iteration 15 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17503722012043\n",
      "Based on desired graph size, contribution limit is set to 0.14858505129814148\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 16 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1602732390165329\n",
      "Based on desired graph size, contribution limit is set to 0.14839565753936768\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 17 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1685551255941391\n",
      "Based on desired graph size, contribution limit is set to 0.14829984307289124\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 18 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1675790697336197\n",
      "Based on desired graph size, contribution limit is set to 0.14856790006160736\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 19 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.16699855029582977\n",
      "Based on desired graph size, contribution limit is set to 0.14851728081703186\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 20 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1679367572069168\n",
      "Based on desired graph size, contribution limit is set to 0.14839157462120056\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 21 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17075294256210327\n",
      "Based on desired graph size, contribution limit is set to 0.14835286140441895\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 22 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17001330852508545\n",
      "Based on desired graph size, contribution limit is set to 0.1486179232597351\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 23 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.16797436773777008\n",
      "Based on desired graph size, contribution limit is set to 0.14831241965293884\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 24 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.16781438887119293\n",
      "Based on desired graph size, contribution limit is set to 0.14842067658901215\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 25 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.16614706814289093\n",
      "Based on desired graph size, contribution limit is set to 0.14843013882637024\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 26 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17334577441215515\n",
      "Based on desired graph size, contribution limit is set to 0.14852622151374817\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 27 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.16902609169483185\n",
      "Based on desired graph size, contribution limit is set to 0.1486344337463379\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 28 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17497922480106354\n",
      "Based on desired graph size, contribution limit is set to 0.14842309057712555\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 29 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17262913286685944\n",
      "Based on desired graph size, contribution limit is set to 0.14835967123508453\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "No good explanations found after 30 iterations...\n",
      "Explain for edge between nodes 5727 and 9663\n",
      "node1:\n",
      "      index_id          id   semantic                          label   \n",
      "5727      5727  HP:0003707  phenotype  Calf muscle pseudohypertrophy  \\\n",
      "\n",
      "      semantic_id  \n",
      "5727            9  \n",
      "node2:\n",
      "     index_id   id semantic       label  semantic_id\n",
      "772       772  176     drug  amiodarone            4\n",
      "--- Iteration 0 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19227959215641022\n",
      "Based on desired graph size, contribution limit is set to 0.1480652093887329\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 1 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19845569133758545\n",
      "Based on desired graph size, contribution limit is set to 0.14819222688674927\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 2 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19902122020721436\n",
      "Based on desired graph size, contribution limit is set to 0.14817167818546295\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 3 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.20023366808891296\n",
      "Based on desired graph size, contribution limit is set to 0.1481882482767105\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 4 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19320204854011536\n",
      "Based on desired graph size, contribution limit is set to 0.14793652296066284\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 5 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1974249929189682\n",
      "Based on desired graph size, contribution limit is set to 0.14821481704711914\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 6 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.21376244723796844\n",
      "Based on desired graph size, contribution limit is set to 0.1481621414422989\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 7 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19694195687770844\n",
      "Based on desired graph size, contribution limit is set to 0.14800599217414856\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 8 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19691242277622223\n",
      "Based on desired graph size, contribution limit is set to 0.1482628434896469\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 9 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.20006895065307617\n",
      "Based on desired graph size, contribution limit is set to 0.14823921024799347\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 32\n",
      "--- Iteration 10 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.20423822104930878\n",
      "Based on desired graph size, contribution limit is set to 0.14822649955749512\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 11 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.20117944478988647\n",
      "Based on desired graph size, contribution limit is set to 0.14838097989559174\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 12 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.20161305367946625\n",
      "Based on desired graph size, contribution limit is set to 0.14814524352550507\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 13 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19273142516613007\n",
      "Based on desired graph size, contribution limit is set to 0.14821738004684448\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 14 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.17786934971809387\n",
      "Based on desired graph size, contribution limit is set to 0.14819616079330444\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 15 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19136443734169006\n",
      "Based on desired graph size, contribution limit is set to 0.14827430248260498\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 16 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19851063191890717\n",
      "Based on desired graph size, contribution limit is set to 0.1481267213821411\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 32\n",
      "--- Iteration 17 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1822528839111328\n",
      "Based on desired graph size, contribution limit is set to 0.1480887234210968\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 18 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.18976901471614838\n",
      "Based on desired graph size, contribution limit is set to 0.1480596512556076\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 19 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.18671807646751404\n",
      "Based on desired graph size, contribution limit is set to 0.14823538064956665\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 20 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19777093827724457\n",
      "Based on desired graph size, contribution limit is set to 0.14807669818401337\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 21 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1889718770980835\n",
      "Based on desired graph size, contribution limit is set to 0.14829501509666443\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 22 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.18975701928138733\n",
      "Based on desired graph size, contribution limit is set to 0.14822813868522644\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 23 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.188750222325325\n",
      "Based on desired graph size, contribution limit is set to 0.14805303514003754\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 24 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.20559988915920258\n",
      "Based on desired graph size, contribution limit is set to 0.14810773730278015\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 25 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19752097129821777\n",
      "Based on desired graph size, contribution limit is set to 0.14831896126270294\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 26 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.18706540763378143\n",
      "Based on desired graph size, contribution limit is set to 0.1482982337474823\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 27 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1915573626756668\n",
      "Based on desired graph size, contribution limit is set to 0.14779317378997803\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 28 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19910356402397156\n",
      "Based on desired graph size, contribution limit is set to 0.14821089804172516\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 29 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19597455859184265\n",
      "Based on desired graph size, contribution limit is set to 0.14816951751708984\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "No good explanations found after 30 iterations...\n",
      "Explain for edge between nodes 5727 and 772\n",
      "node1:\n",
      "      index_id          id   semantic                          label   \n",
      "5727      5727  HP:0003707  phenotype  Calf muscle pseudohypertrophy  \\\n",
      "\n",
      "      semantic_id  \n",
      "5727            9  \n",
      "node2:\n",
      "      index_id    id semantic       label  semantic_id\n",
      "7315      7315  1043     drug  ergotamine            4\n",
      "--- Iteration 0 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19142159819602966\n",
      "Based on desired graph size, contribution limit is set to 0.1479046642780304\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 32\n",
      "--- Iteration 1 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1843545138835907\n",
      "Based on desired graph size, contribution limit is set to 0.14753152430057526\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 2 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19073806703090668\n",
      "Based on desired graph size, contribution limit is set to 0.1477396935224533\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 30\n",
      "--- Iteration 3 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.19636370241641998\n",
      "Based on desired graph size, contribution limit is set to 0.14772555232048035\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 4 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.18825729191303253\n",
      "Based on desired graph size, contribution limit is set to 0.14774690568447113\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 5 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1960357129573822\n",
      "Based on desired graph size, contribution limit is set to 0.14775505661964417\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 6 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.20547515153884888\n",
      "Based on desired graph size, contribution limit is set to 0.14800012111663818\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 7 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1967364251613617\n",
      "Based on desired graph size, contribution limit is set to 0.14799077808856964\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 33\n",
      "--- Iteration 8 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.18340854346752167\n",
      "Based on desired graph size, contribution limit is set to 0.14771902561187744\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 9 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.18049746751785278\n",
      "Based on desired graph size, contribution limit is set to 0.14776833355426788\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 35\n",
      "--- Iteration 10 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1874285340309143\n",
      "Based on desired graph size, contribution limit is set to 0.14769534766674042\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 34\n",
      "--- Iteration 11 ---\n",
      "A total of 2958 edges contributing to prediction.\n",
      "Maximum contribution value is 0.1890588402748108\n",
      "Based on desired graph size, contribution limit is set to 0.14773298799991608\n",
      "Pruned graph has edge total of: 31\n",
      "Pruned graph has node total of: 32\n",
      "--- Iteration 12 ---\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[83], line 21\u001b[0m\n\u001b[0;32m     18\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m'\u001b[39m\u001b[39mnode2:\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m     19\u001b[0m \u001b[39mprint\u001b[39m(nodes\u001b[39m.\u001b[39mloc[[node_idx2]])\n\u001b[1;32m---> 21\u001b[0m explained \u001b[39m=\u001b[39m explain_edge(node_idx1, node_idx2)\n\u001b[0;32m     22\u001b[0m \u001b[39mif\u001b[39;00m explained:\n\u001b[0;32m     23\u001b[0m     found_explanation\u001b[39m.\u001b[39mappend({\u001b[39m'\u001b[39m\u001b[39msymptom\u001b[39m\u001b[39m'\u001b[39m: symptom_id, \u001b[39m'\u001b[39m\u001b[39mdrug\u001b[39m\u001b[39m'\u001b[39m: drug_id})\n",
      "Cell \u001b[1;32mIn[82], line 12\u001b[0m, in \u001b[0;36mexplain_edge\u001b[1;34m(node_idx1, node_idx2)\u001b[0m\n\u001b[0;32m      9\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mnot\u001b[39;00m done:\n\u001b[0;32m     10\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m--- Iteration \u001b[39m\u001b[39m{\u001b[39;00mcurrent_iterations\u001b[39m}\u001b[39;00m\u001b[39m ---\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m---> 12\u001b[0m     explanation \u001b[39m=\u001b[39m explainer(\n\u001b[0;32m     13\u001b[0m         x\u001b[39m=\u001b[39;49mDeepG\u001b[39m.\u001b[39;49mnode_feature,\n\u001b[0;32m     14\u001b[0m         edge_index\u001b[39m=\u001b[39;49mDeepG\u001b[39m.\u001b[39;49medge_index,\n\u001b[0;32m     15\u001b[0m         edge_label_index\u001b[39m=\u001b[39;49medge_label_index,\n\u001b[0;32m     16\u001b[0m     )\n\u001b[0;32m     18\u001b[0m     edge_mask \u001b[39m=\u001b[39m explanation\u001b[39m.\u001b[39medge_mask\n\u001b[0;32m     20\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mA total of \u001b[39m\u001b[39m{\u001b[39;00mtorch\u001b[39m.\u001b[39mcount_nonzero(edge_mask)\u001b[39m}\u001b[39;00m\u001b[39m edges contributing to prediction.\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch_geometric\\explain\\explainer.py:198\u001b[0m, in \u001b[0;36mExplainer.__call__\u001b[1;34m(self, x, edge_index, target, index, **kwargs)\u001b[0m\n\u001b[0;32m    195\u001b[0m training \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmodel\u001b[39m.\u001b[39mtraining\n\u001b[0;32m    196\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmodel\u001b[39m.\u001b[39meval()\n\u001b[1;32m--> 198\u001b[0m explanation \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39malgorithm(\n\u001b[0;32m    199\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmodel,\n\u001b[0;32m    200\u001b[0m     x,\n\u001b[0;32m    201\u001b[0m     edge_index,\n\u001b[0;32m    202\u001b[0m     target\u001b[39m=\u001b[39mtarget,\n\u001b[0;32m    203\u001b[0m     index\u001b[39m=\u001b[39mindex,\n\u001b[0;32m    204\u001b[0m     \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs,\n\u001b[0;32m    205\u001b[0m )\n\u001b[0;32m    207\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmodel\u001b[39m.\u001b[39mtrain(training)\n\u001b[0;32m    209\u001b[0m \u001b[39m# Add explainer objectives to the `Explanation` object:\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch\\nn\\modules\\module.py:1501\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1496\u001b[0m \u001b[39m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[0;32m   1497\u001b[0m \u001b[39m# this function, and just call forward.\u001b[39;00m\n\u001b[0;32m   1498\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m (\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_backward_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_backward_pre_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_forward_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_forward_pre_hooks\n\u001b[0;32m   1499\u001b[0m         \u001b[39mor\u001b[39;00m _global_backward_pre_hooks \u001b[39mor\u001b[39;00m _global_backward_hooks\n\u001b[0;32m   1500\u001b[0m         \u001b[39mor\u001b[39;00m _global_forward_hooks \u001b[39mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[1;32m-> 1501\u001b[0m     \u001b[39mreturn\u001b[39;00m forward_call(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m   1502\u001b[0m \u001b[39m# Do not call functions when jit is used\u001b[39;00m\n\u001b[0;32m   1503\u001b[0m full_backward_hooks, non_full_backward_hooks \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch_geometric\\explain\\algorithm\\gnn_explainer.py:75\u001b[0m, in \u001b[0;36mGNNExplainer.forward\u001b[1;34m(self, model, x, edge_index, target, index, **kwargs)\u001b[0m\n\u001b[0;32m     71\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(x, \u001b[39mdict\u001b[39m) \u001b[39mor\u001b[39;00m \u001b[39misinstance\u001b[39m(edge_index, \u001b[39mdict\u001b[39m):\n\u001b[0;32m     72\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mHeterogeneous graphs not yet supported in \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m     73\u001b[0m                      \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__class__\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__name__\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m---> 75\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_train(model, x, edge_index, target\u001b[39m=\u001b[39mtarget, index\u001b[39m=\u001b[39mindex, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m     77\u001b[0m node_mask \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_post_process_mask(\n\u001b[0;32m     78\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mnode_mask,\n\u001b[0;32m     79\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhard_node_mask,\n\u001b[0;32m     80\u001b[0m     apply_sigmoid\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m,\n\u001b[0;32m     81\u001b[0m )\n\u001b[0;32m     82\u001b[0m edge_mask \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_post_process_mask(\n\u001b[0;32m     83\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39medge_mask,\n\u001b[0;32m     84\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhard_edge_mask,\n\u001b[0;32m     85\u001b[0m     apply_sigmoid\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m,\n\u001b[0;32m     86\u001b[0m )\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch_geometric\\explain\\algorithm\\gnn_explainer.py:120\u001b[0m, in \u001b[0;36mGNNExplainer._train\u001b[1;34m(self, model, x, edge_index, target, index, **kwargs)\u001b[0m\n\u001b[0;32m    117\u001b[0m optimizer\u001b[39m.\u001b[39mzero_grad()\n\u001b[0;32m    119\u001b[0m h \u001b[39m=\u001b[39m x \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mnode_mask \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39melse\u001b[39;00m x \u001b[39m*\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mnode_mask\u001b[39m.\u001b[39msigmoid()\n\u001b[1;32m--> 120\u001b[0m y_hat, y \u001b[39m=\u001b[39m model(h, edge_index, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs), target\n\u001b[0;32m    122\u001b[0m \u001b[39mif\u001b[39;00m index \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m    123\u001b[0m     y_hat, y \u001b[39m=\u001b[39m y_hat[index], y[index]\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch\\nn\\modules\\module.py:1501\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1496\u001b[0m \u001b[39m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[0;32m   1497\u001b[0m \u001b[39m# this function, and just call forward.\u001b[39;00m\n\u001b[0;32m   1498\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m (\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_backward_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_backward_pre_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_forward_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_forward_pre_hooks\n\u001b[0;32m   1499\u001b[0m         \u001b[39mor\u001b[39;00m _global_backward_pre_hooks \u001b[39mor\u001b[39;00m _global_backward_hooks\n\u001b[0;32m   1500\u001b[0m         \u001b[39mor\u001b[39;00m _global_forward_hooks \u001b[39mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[1;32m-> 1501\u001b[0m     \u001b[39mreturn\u001b[39;00m forward_call(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m   1502\u001b[0m \u001b[39m# Do not call functions when jit is used\u001b[39;00m\n\u001b[0;32m   1503\u001b[0m full_backward_hooks, non_full_backward_hooks \u001b[39m=\u001b[39m [], []\n",
      "Cell \u001b[1;32mIn[68], line 43\u001b[0m, in \u001b[0;36mLinkPredModel.forward\u001b[1;34m(self, x, edge_index, edge_label_index, training)\u001b[0m\n\u001b[0;32m     40\u001b[0m   x \u001b[39m=\u001b[39m F\u001b[39m.\u001b[39mleaky_relu(x)\n\u001b[0;32m     41\u001b[0m   x \u001b[39m=\u001b[39m F\u001b[39m.\u001b[39mdropout(x, p\u001b[39m=\u001b[39m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdropout, training\u001b[39m=\u001b[39mtraining)\n\u001b[1;32m---> 43\u001b[0m x \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mconvs[\u001b[39m-\u001b[39;49m\u001b[39m1\u001b[39;49m](x, edge_index)  \n\u001b[0;32m     44\u001b[0m x \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mbns[\u001b[39m-\u001b[39m\u001b[39m1\u001b[39m](x)\n\u001b[0;32m     47\u001b[0m nodes_first, nodes_second \u001b[39m=\u001b[39m edge_label_index\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch\\nn\\modules\\module.py:1501\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1496\u001b[0m \u001b[39m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[0;32m   1497\u001b[0m \u001b[39m# this function, and just call forward.\u001b[39;00m\n\u001b[0;32m   1498\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m (\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_backward_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_backward_pre_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_forward_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_forward_pre_hooks\n\u001b[0;32m   1499\u001b[0m         \u001b[39mor\u001b[39;00m _global_backward_pre_hooks \u001b[39mor\u001b[39;00m _global_backward_hooks\n\u001b[0;32m   1500\u001b[0m         \u001b[39mor\u001b[39;00m _global_forward_hooks \u001b[39mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[1;32m-> 1501\u001b[0m     \u001b[39mreturn\u001b[39;00m forward_call(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m   1502\u001b[0m \u001b[39m# Do not call functions when jit is used\u001b[39;00m\n\u001b[0;32m   1503\u001b[0m full_backward_hooks, non_full_backward_hooks \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch_geometric\\nn\\conv\\sage_conv.py:131\u001b[0m, in \u001b[0;36mSAGEConv.forward\u001b[1;34m(self, x, edge_index, size)\u001b[0m\n\u001b[0;32m    128\u001b[0m     x \u001b[39m=\u001b[39m (\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mlin(x[\u001b[39m0\u001b[39m])\u001b[39m.\u001b[39mrelu(), x[\u001b[39m1\u001b[39m])\n\u001b[0;32m    130\u001b[0m \u001b[39m# propagate_type: (x: OptPairTensor)\u001b[39;00m\n\u001b[1;32m--> 131\u001b[0m out \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mpropagate(edge_index, x\u001b[39m=\u001b[39;49mx, size\u001b[39m=\u001b[39;49msize)\n\u001b[0;32m    132\u001b[0m out \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mlin_l(out)\n\u001b[0;32m    134\u001b[0m x_r \u001b[39m=\u001b[39m x[\u001b[39m1\u001b[39m]\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch_geometric\\nn\\conv\\message_passing.py:484\u001b[0m, in \u001b[0;36mMessagePassing.propagate\u001b[1;34m(self, edge_index, size, **kwargs)\u001b[0m\n\u001b[0;32m    481\u001b[0m     \u001b[39mif\u001b[39;00m res \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m    482\u001b[0m         aggr_kwargs \u001b[39m=\u001b[39m res[\u001b[39m0\u001b[39m] \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(res, \u001b[39mtuple\u001b[39m) \u001b[39melse\u001b[39;00m res\n\u001b[1;32m--> 484\u001b[0m out \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39maggregate(out, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39maggr_kwargs)\n\u001b[0;32m    486\u001b[0m \u001b[39mfor\u001b[39;00m hook \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_aggregate_forward_hooks\u001b[39m.\u001b[39mvalues():\n\u001b[0;32m    487\u001b[0m     res \u001b[39m=\u001b[39m hook(\u001b[39mself\u001b[39m, (aggr_kwargs, ), out)\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch_geometric\\nn\\conv\\message_passing.py:608\u001b[0m, in \u001b[0;36mMessagePassing.aggregate\u001b[1;34m(self, inputs, index, ptr, dim_size)\u001b[0m\n\u001b[0;32m    595\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39maggregate\u001b[39m(\u001b[39mself\u001b[39m, inputs: Tensor, index: Tensor,\n\u001b[0;32m    596\u001b[0m               ptr: Optional[Tensor] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[0;32m    597\u001b[0m               dim_size: Optional[\u001b[39mint\u001b[39m] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Tensor:\n\u001b[0;32m    598\u001b[0m \u001b[39m    \u001b[39m\u001b[39mr\u001b[39m\u001b[39m\"\"\"Aggregates messages from neighbors as\u001b[39;00m\n\u001b[0;32m    599\u001b[0m \u001b[39m    :math:`\\bigoplus_{j \\in \\mathcal{N}(i)}`.\u001b[39;00m\n\u001b[0;32m    600\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    606\u001b[0m \u001b[39m    as specified in :meth:`__init__` by the :obj:`aggr` argument.\u001b[39;00m\n\u001b[0;32m    607\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 608\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49maggr_module(inputs, index, ptr\u001b[39m=\u001b[39;49mptr, dim_size\u001b[39m=\u001b[39;49mdim_size,\n\u001b[0;32m    609\u001b[0m                             dim\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mnode_dim)\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch_geometric\\nn\\aggr\\base.py:109\u001b[0m, in \u001b[0;36mAggregation.__call__\u001b[1;34m(self, x, index, ptr, dim_size, dim, **kwargs)\u001b[0m\n\u001b[0;32m    106\u001b[0m     dim_size \u001b[39m=\u001b[39m \u001b[39mint\u001b[39m(index\u001b[39m.\u001b[39mmax()) \u001b[39m+\u001b[39m \u001b[39m1\u001b[39m \u001b[39mif\u001b[39;00m index\u001b[39m.\u001b[39mnumel() \u001b[39m>\u001b[39m \u001b[39m0\u001b[39m \u001b[39melse\u001b[39;00m \u001b[39m0\u001b[39m\n\u001b[0;32m    108\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 109\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39msuper\u001b[39m()\u001b[39m.\u001b[39m\u001b[39m__call__\u001b[39m(x, index, ptr, dim_size, dim, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m    110\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIndexError\u001b[39;00m, \u001b[39mRuntimeError\u001b[39;00m) \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m    111\u001b[0m     \u001b[39mif\u001b[39;00m index \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch\\nn\\modules\\module.py:1501\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1496\u001b[0m \u001b[39m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[0;32m   1497\u001b[0m \u001b[39m# this function, and just call forward.\u001b[39;00m\n\u001b[0;32m   1498\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m (\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_backward_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_backward_pre_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_forward_hooks \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_forward_pre_hooks\n\u001b[0;32m   1499\u001b[0m         \u001b[39mor\u001b[39;00m _global_backward_pre_hooks \u001b[39mor\u001b[39;00m _global_backward_hooks\n\u001b[0;32m   1500\u001b[0m         \u001b[39mor\u001b[39;00m _global_forward_hooks \u001b[39mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[1;32m-> 1501\u001b[0m     \u001b[39mreturn\u001b[39;00m forward_call(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m   1502\u001b[0m \u001b[39m# Do not call functions when jit is used\u001b[39;00m\n\u001b[0;32m   1503\u001b[0m full_backward_hooks, non_full_backward_hooks \u001b[39m=\u001b[39m [], []\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch_geometric\\nn\\aggr\\basic.py:34\u001b[0m, in \u001b[0;36mMeanAggregation.forward\u001b[1;34m(self, x, index, ptr, dim_size, dim)\u001b[0m\n\u001b[0;32m     31\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mforward\u001b[39m(\u001b[39mself\u001b[39m, x: Tensor, index: Optional[Tensor] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[0;32m     32\u001b[0m             ptr: Optional[Tensor] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m, dim_size: Optional[\u001b[39mint\u001b[39m] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[0;32m     33\u001b[0m             dim: \u001b[39mint\u001b[39m \u001b[39m=\u001b[39m \u001b[39m-\u001b[39m\u001b[39m2\u001b[39m) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Tensor:\n\u001b[1;32m---> 34\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mreduce(x, index, ptr, dim_size, dim, reduce\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mmean\u001b[39;49m\u001b[39m'\u001b[39;49m)\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch_geometric\\nn\\aggr\\base.py:155\u001b[0m, in \u001b[0;36mAggregation.reduce\u001b[1;34m(self, x, index, ptr, dim_size, dim, reduce)\u001b[0m\n\u001b[0;32m    152\u001b[0m     \u001b[39mreturn\u001b[39;00m segment(x, ptr, reduce\u001b[39m=\u001b[39mreduce)\n\u001b[0;32m    154\u001b[0m \u001b[39massert\u001b[39;00m index \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m--> 155\u001b[0m \u001b[39mreturn\u001b[39;00m scatter(x, index, dim, dim_size, reduce)\n",
      "File \u001b[1;32mc:\\Users\\rosa-\\anaconda3\\envs\\xaifognn\\lib\\site-packages\\torch_geometric\\utils\\scatter.py:82\u001b[0m, in \u001b[0;36mscatter\u001b[1;34m(src, index, dim, dim_size, reduce)\u001b[0m\n\u001b[0;32m     79\u001b[0m     count \u001b[39m=\u001b[39m count\u001b[39m.\u001b[39mclamp(\u001b[39mmin\u001b[39m\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m)\n\u001b[0;32m     81\u001b[0m     index \u001b[39m=\u001b[39m broadcast(index, src, dim)\n\u001b[1;32m---> 82\u001b[0m     out \u001b[39m=\u001b[39m src\u001b[39m.\u001b[39;49mnew_zeros(size)\u001b[39m.\u001b[39;49mscatter_add_(dim, index, src)\n\u001b[0;32m     84\u001b[0m     \u001b[39mreturn\u001b[39;00m out \u001b[39m/\u001b[39m broadcast(count, out, dim)\n\u001b[0;32m     86\u001b[0m \u001b[39m# For \"min\" and \"max\" reduction, we prefer `scatter_reduce_` on CPU or\u001b[39;00m\n\u001b[0;32m     87\u001b[0m \u001b[39m# in case the input does not require gradients:\u001b[39;00m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "def get_node_idx(id, nodes): \n",
    "    return nodes[nodes['id'] == id].index.values.astype(int)[0]\n",
    "\n",
    "found_explanation = []\n",
    "\n",
    "for ind, edge in symptoms_drugs.iterrows(): \n",
    "    symptom_id = edge['Symptom']\n",
    "    \n",
    "    for drug_id in edge['Candidates']:\n",
    "        \n",
    "        print('Explain for edge between nodes', node_idx1, 'and', node_idx2)\n",
    "\n",
    "        node_idx1 = get_node_idx(symptom_id, nodes)\n",
    "        print('node1:')\n",
    "        print(nodes.loc[[node_idx1]])\n",
    "\n",
    "        node_idx2 = get_node_idx(str(drug_id), nodes)\n",
    "        print('node2:')\n",
    "        print(nodes.loc[[node_idx2]])\n",
    "        \n",
    "        explained = explain_edge(node_idx1, node_idx2)\n",
    "        if explained:\n",
    "            found_explanation.append({'symptom': symptom_id, 'drug': drug_id})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "found_explanation"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xaifognn",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
