{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook loads all edges and nodes of the relevant knowledge graph. New dataframes are created for the nodes and edges including index values for each node."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify which dataset is used.\n",
    "*   1 (Original knowledge graph)\n",
    "*   2 (Restructured knowledge graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_nr = 2\n",
    "assert dataset_nr == 1 or 2\n",
    "\n",
    "NO_GENE_PRODUCTS = 'no gene products'\n",
    "SINGLE_RELATION_TYPE = 'only one relation type' # TODO: implement this as well\n",
    "concept_changes = NO_GENE_PRODUCTS\n",
    "\n",
    "if dataset_nr == 1:\n",
    "    prefix = 'prev'\n",
    "else:\n",
    "    prefix = 'new'\n",
    "    \n",
    "if concept_changes == NO_GENE_PRODUCTS:\n",
    "    suffix = '_nogeneprods'\n",
    "elif concept_changes == SINGLE_RELATION_TYPE:\n",
    "    suffix = '_singlerel'\n",
    "else:\n",
    "    suffix = ''\n",
    "    \n",
    "FILE_EDGES = f'{prefix}_kg_edges.csv'\n",
    "FILE_NODES = f'{prefix}_kg_nodes.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Edges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load all edges of knowledge graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>subject</th>\n",
       "      <th>object</th>\n",
       "      <th>relation_id</th>\n",
       "      <th>relation_label</th>\n",
       "      <th>relation_iri</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15a20d1b-1d4c-4ef5-8960-0bbcee1b4261</td>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>RO:0002434</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0002434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>78c75c6d-806d-47b4-930f-4544eaa2c5a5</td>\n",
       "      <td>ZFIN:ZDB-GENE-020129-1</td>\n",
       "      <td>ZP:0004595</td>\n",
       "      <td>RO:0003303</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0003303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>34b96ade-f51a-4680-8c53-c00fbf079968</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000001848</td>\n",
       "      <td>RGD:1310277</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_HOM0000017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d32fd8ea-e219-45d9-be6d-7adcba8289c0</td>\n",
       "      <td>MGI:101864</td>\n",
       "      <td>MP:0002961</td>\n",
       "      <td>RO:0003303</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0003303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>58c47a12-8477-40a3-9009-a200854f84fe</td>\n",
       "      <td>ENSEMBL:ENSMODG00000018459</td>\n",
       "      <td>WormBase:WBGene00003911</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_HOM0000017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2ec9e759-e9ea-4e9d-ac75-57a643cfd845</td>\n",
       "      <td>MGI:1277151</td>\n",
       "      <td>HP:0012757</td>\n",
       "      <td>RO:0003303</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0003303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>18ad99b9-19c6-419a-bb5e-d49926575675</td>\n",
       "      <td>FlyBase:FBgn0261563</td>\n",
       "      <td>ENSEMBL:ENSMMUG00000004321</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_HOM0000017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5990c621-27da-4f75-bc39-55ad604f191f</td>\n",
       "      <td>ENSEMBL:ENSFCAG00000012569</td>\n",
       "      <td>HGNC:11873</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_HOM0000017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6bdb9894-3967-441e-aa68-f5a37e900c56</td>\n",
       "      <td>WormBase:WBGene00006062</td>\n",
       "      <td>WBPhenotype:0001316</td>\n",
       "      <td>RO:0003303</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_0003303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ed0f0944-a99c-4508-ba34-59675d615d4c</td>\n",
       "      <td>HGNC:8591</td>\n",
       "      <td>MGI:1339984</td>\n",
       "      <td>RO:HOM0000017</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>http://purl.obolibrary.org/obo/RO_HOM0000017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     id                     subject   \n",
       "0  15a20d1b-1d4c-4ef5-8960-0bbcee1b4261     WormBase:WBGene00006787  \\\n",
       "1  78c75c6d-806d-47b4-930f-4544eaa2c5a5      ZFIN:ZDB-GENE-020129-1   \n",
       "2  34b96ade-f51a-4680-8c53-c00fbf079968  ENSEMBL:ENSCAFG00000001848   \n",
       "3  d32fd8ea-e219-45d9-be6d-7adcba8289c0                  MGI:101864   \n",
       "4  58c47a12-8477-40a3-9009-a200854f84fe  ENSEMBL:ENSMODG00000018459   \n",
       "5  2ec9e759-e9ea-4e9d-ac75-57a643cfd845                 MGI:1277151   \n",
       "6  18ad99b9-19c6-419a-bb5e-d49926575675         FlyBase:FBgn0261563   \n",
       "7  5990c621-27da-4f75-bc39-55ad604f191f  ENSEMBL:ENSFCAG00000012569   \n",
       "8  6bdb9894-3967-441e-aa68-f5a37e900c56     WormBase:WBGene00006062   \n",
       "9  ed0f0944-a99c-4508-ba34-59675d615d4c                   HGNC:8591   \n",
       "\n",
       "                       object    relation_id                  relation_label   \n",
       "0     WormBase:WBGene00003929     RO:0002434                  interacts with  \\\n",
       "1                  ZP:0004595     RO:0003303                causes condition   \n",
       "2                 RGD:1310277  RO:HOM0000017  in orthology relationship with   \n",
       "3                  MP:0002961     RO:0003303                causes condition   \n",
       "4     WormBase:WBGene00003911  RO:HOM0000017  in orthology relationship with   \n",
       "5                  HP:0012757     RO:0003303                causes condition   \n",
       "6  ENSEMBL:ENSMMUG00000004321  RO:HOM0000017  in orthology relationship with   \n",
       "7                  HGNC:11873  RO:HOM0000017  in orthology relationship with   \n",
       "8         WBPhenotype:0001316     RO:0003303                causes condition   \n",
       "9                 MGI:1339984  RO:HOM0000017  in orthology relationship with   \n",
       "\n",
       "                                   relation_iri  \n",
       "0     http://purl.obolibrary.org/obo/RO_0002434  \n",
       "1     http://purl.obolibrary.org/obo/RO_0003303  \n",
       "2  http://purl.obolibrary.org/obo/RO_HOM0000017  \n",
       "3     http://purl.obolibrary.org/obo/RO_0003303  \n",
       "4  http://purl.obolibrary.org/obo/RO_HOM0000017  \n",
       "5     http://purl.obolibrary.org/obo/RO_0003303  \n",
       "6  http://purl.obolibrary.org/obo/RO_HOM0000017  \n",
       "7  http://purl.obolibrary.org/obo/RO_HOM0000017  \n",
       "8     http://purl.obolibrary.org/obo/RO_0003303  \n",
       "9  http://purl.obolibrary.org/obo/RO_HOM0000017  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = pd.read_csv(f'output/{FILE_EDGES}', header=0)\n",
    "edges.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract the relevant columns and rename them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZFIN:ZDB-GENE-020129-1</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>ZP:0004595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSEMBL:ENSCAFG00000001848</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>RGD:1310277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:101864</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>MP:0002961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSEMBL:ENSMODG00000018459</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>WormBase:WBGene00003911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MGI:1277151</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>HP:0012757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FlyBase:FBgn0261563</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSMMUG00000004321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSEMBL:ENSFCAG00000012569</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>HGNC:11873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WormBase:WBGene00006062</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HGNC:8591</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>MGI:1339984</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         head                        relation   \n",
       "0     WormBase:WBGene00006787                  interacts with  \\\n",
       "1      ZFIN:ZDB-GENE-020129-1                causes condition   \n",
       "2  ENSEMBL:ENSCAFG00000001848  in orthology relationship with   \n",
       "3                  MGI:101864                causes condition   \n",
       "4  ENSEMBL:ENSMODG00000018459  in orthology relationship with   \n",
       "5                 MGI:1277151                causes condition   \n",
       "6         FlyBase:FBgn0261563  in orthology relationship with   \n",
       "7  ENSEMBL:ENSFCAG00000012569  in orthology relationship with   \n",
       "8     WormBase:WBGene00006062                causes condition   \n",
       "9                   HGNC:8591  in orthology relationship with   \n",
       "\n",
       "                         tail  \n",
       "0     WormBase:WBGene00003929  \n",
       "1                  ZP:0004595  \n",
       "2                 RGD:1310277  \n",
       "3                  MP:0002961  \n",
       "4     WormBase:WBGene00003911  \n",
       "5                  HP:0012757  \n",
       "6  ENSEMBL:ENSMMUG00000004321  \n",
       "7                  HGNC:11873  \n",
       "8         WBPhenotype:0001316  \n",
       "9                 MGI:1339984  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = edges[['subject', 'relation_label', 'object']]\n",
    "edges.rename(columns={'subject': 'head', 'relation_label': 'relation', 'object': 'tail'}, inplace=True)\n",
    "edges.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_int_str(value):\n",
    "    try:\n",
    "        return str(int(float(value)))\n",
    "    except ValueError:\n",
    "        return value\n",
    "\n",
    "edges['head'] = edges['head'].apply(convert_to_int_str)\n",
    "edges['tail'] = edges['tail'].apply(convert_to_int_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 85992 edges\n"
     ]
    }
   ],
   "source": [
    "print(f'There are {edges.shape[0]} edges')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load all nodes of knowledge graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>label</th>\n",
       "      <th>iri</th>\n",
       "      <th>semantic</th>\n",
       "      <th>index_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MP:0004187</td>\n",
       "      <td>cardia bifida</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MP_0004187</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZP:0100138</td>\n",
       "      <td>muscle tendon junction myotome increased amoun...</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ZP_0100138</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1346525</td>\n",
       "      <td>Sgcd</td>\n",
       "      <td>http://www.informatics.jax.org/accession/MGI:1...</td>\n",
       "      <td>gene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OMIM:300377.0044</td>\n",
       "      <td>DMD, LYS770TER</td>\n",
       "      <td>http://omim.org/entry/300377.0044</td>\n",
       "      <td>variant</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZP:0002210</td>\n",
       "      <td>posterior lateral line neuromast primordium mi...</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ZP_0002210</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MP:0020335</td>\n",
       "      <td>abnormal dentate gyrus neuron dendrite morphology</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MP_0020335</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HP:0000316</td>\n",
       "      <td>Hypertelorism</td>\n",
       "      <td>http://purl.obolibrary.org/obo/HP_0000316</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ZP:0004032</td>\n",
       "      <td>margin constricted, abnormal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ZP_0004032</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ZP:0003164</td>\n",
       "      <td>atrium decreased size, abnormal</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ZP_0003164</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MP:0003331</td>\n",
       "      <td>increased hepatocellular carcinoma incidence</td>\n",
       "      <td>http://purl.obolibrary.org/obo/MP_0003331</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id                                              label   \n",
       "0        MP:0004187                                      cardia bifida  \\\n",
       "1        ZP:0100138  muscle tendon junction myotome increased amoun...   \n",
       "2       MGI:1346525                                               Sgcd   \n",
       "3  OMIM:300377.0044                                     DMD, LYS770TER   \n",
       "4        ZP:0002210  posterior lateral line neuromast primordium mi...   \n",
       "5        MP:0020335  abnormal dentate gyrus neuron dendrite morphology   \n",
       "6        HP:0000316                                      Hypertelorism   \n",
       "7        ZP:0004032                       margin constricted, abnormal   \n",
       "8        ZP:0003164                    atrium decreased size, abnormal   \n",
       "9        MP:0003331       increased hepatocellular carcinoma incidence   \n",
       "\n",
       "                                                 iri   semantic  index_id  \n",
       "0          http://purl.obolibrary.org/obo/MP_0004187  phenotype         0  \n",
       "1          http://purl.obolibrary.org/obo/ZP_0100138  phenotype         1  \n",
       "2  http://www.informatics.jax.org/accession/MGI:1...       gene         2  \n",
       "3                  http://omim.org/entry/300377.0044    variant         3  \n",
       "4          http://purl.obolibrary.org/obo/ZP_0002210  phenotype         4  \n",
       "5          http://purl.obolibrary.org/obo/MP_0020335  phenotype         5  \n",
       "6          http://purl.obolibrary.org/obo/HP_0000316  phenotype         6  \n",
       "7          http://purl.obolibrary.org/obo/ZP_0004032  phenotype         7  \n",
       "8          http://purl.obolibrary.org/obo/ZP_0003164  phenotype         8  \n",
       "9          http://purl.obolibrary.org/obo/MP_0003331  phenotype         9  "
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = pd.read_csv(f'output/{FILE_NODES}', header=0)\n",
    "nodes['index_id'] = nodes.index\n",
    "nodes.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract relevant columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>MP:0004187</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>cardia bifida</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ZP:0100138</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>muscle tendon junction myotome increased amoun...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1346525</td>\n",
       "      <td>gene</td>\n",
       "      <td>Sgcd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>OMIM:300377.0044</td>\n",
       "      <td>variant</td>\n",
       "      <td>DMD, LYS770TER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ZP:0002210</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>posterior lateral line neuromast primordium mi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>MP:0020335</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>abnormal dentate gyrus neuron dendrite morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>HP:0000316</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>Hypertelorism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>ZP:0004032</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>margin constricted, abnormal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>ZP:0003164</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>atrium decreased size, abnormal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>MP:0003331</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>increased hepatocellular carcinoma incidence</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_id                id   semantic   \n",
       "0         0        MP:0004187  phenotype  \\\n",
       "1         1        ZP:0100138  phenotype   \n",
       "2         2       MGI:1346525       gene   \n",
       "3         3  OMIM:300377.0044    variant   \n",
       "4         4        ZP:0002210  phenotype   \n",
       "5         5        MP:0020335  phenotype   \n",
       "6         6        HP:0000316  phenotype   \n",
       "7         7        ZP:0004032  phenotype   \n",
       "8         8        ZP:0003164  phenotype   \n",
       "9         9        MP:0003331  phenotype   \n",
       "\n",
       "                                               label  \n",
       "0                                      cardia bifida  \n",
       "1  muscle tendon junction myotome increased amoun...  \n",
       "2                                               Sgcd  \n",
       "3                                     DMD, LYS770TER  \n",
       "4  posterior lateral line neuromast primordium mi...  \n",
       "5  abnormal dentate gyrus neuron dendrite morphology  \n",
       "6                                      Hypertelorism  \n",
       "7                       margin constricted, abnormal  \n",
       "8                    atrium decreased size, abnormal  \n",
       "9       increased hepatocellular carcinoma incidence  "
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = nodes[['index_id', 'id', 'semantic', 'label']]\n",
    "nodes.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 10275 nodes\n"
     ]
    }
   ],
   "source": [
    "print(f'There are {nodes.shape[0]} nodes')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert semantic groups from categorical to numerical values (`semantic_id`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>MP:0004187</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>cardia bifida</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ZP:0100138</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>muscle tendon junction myotome increased amoun...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1346525</td>\n",
       "      <td>gene</td>\n",
       "      <td>Sgcd</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>OMIM:300377.0044</td>\n",
       "      <td>variant</td>\n",
       "      <td>DMD, LYS770TER</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ZP:0002210</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>posterior lateral line neuromast primordium mi...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>MP:0020335</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>abnormal dentate gyrus neuron dendrite morphology</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>HP:0000316</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>Hypertelorism</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>ZP:0004032</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>margin constricted, abnormal</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>ZP:0003164</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>atrium decreased size, abnormal</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>MP:0003331</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>increased hepatocellular carcinoma incidence</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index_id                id   semantic   \n",
       "0         0        MP:0004187  phenotype  \\\n",
       "1         1        ZP:0100138  phenotype   \n",
       "2         2       MGI:1346525       gene   \n",
       "3         3  OMIM:300377.0044    variant   \n",
       "4         4        ZP:0002210  phenotype   \n",
       "5         5        MP:0020335  phenotype   \n",
       "6         6        HP:0000316  phenotype   \n",
       "7         7        ZP:0004032  phenotype   \n",
       "8         8        ZP:0003164  phenotype   \n",
       "9         9        MP:0003331  phenotype   \n",
       "\n",
       "                                               label  semantic_id  \n",
       "0                                      cardia bifida            9  \n",
       "1  muscle tendon junction myotome increased amoun...            9  \n",
       "2                                               Sgcd            5  \n",
       "3                                     DMD, LYS770TER           11  \n",
       "4  posterior lateral line neuromast primordium mi...            9  \n",
       "5  abnormal dentate gyrus neuron dendrite morphology            9  \n",
       "6                                      Hypertelorism            9  \n",
       "7                       margin constricted, abnormal            9  \n",
       "8                    atrium decreased size, abnormal            9  \n",
       "9       increased hepatocellular carcinoma incidence            9  "
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes['semantic'] = nodes['semantic'].astype('category')\n",
    "nodes['semantic_id'] = nodes['semantic'].cat.codes\n",
    "nodes.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a dictionary that can be used as mapping between `semantic_id` and `semantic`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'biological artifact',\n",
       " 1: 'biological process',\n",
       " 2: 'cellular component',\n",
       " 3: 'disease',\n",
       " 4: 'drug',\n",
       " 5: 'gene',\n",
       " 6: 'gene product',\n",
       " 7: 'genotype',\n",
       " 8: 'molecular function',\n",
       " 9: 'phenotype',\n",
       " 10: 'taxon',\n",
       " 11: 'variant'}"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_semantic_dict = dict(enumerate(nodes['semantic'].cat.categories))\n",
    "node_semantic_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save new dataframe containing all nodes into csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes.to_csv(f'output/indexed_nodes_{dataset_nr}{suffix}.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Joining Dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge dataframes with nodes and edges in order to join semantic classes and indices of nodes with heads and tails of the edges."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00006789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000015555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ZFIN:ZDB-GENE-021226-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0000781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSPTRG00000021480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000015556</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      head label_head  class_head  index_head   \n",
       "0  WormBase:WBGene00006787     unc-52           5         304  \\\n",
       "1  WormBase:WBGene00006787     unc-52           5         304   \n",
       "2  WormBase:WBGene00006787     unc-52           5         304   \n",
       "3  WormBase:WBGene00006787     unc-52           5         304   \n",
       "4  WormBase:WBGene00006787     unc-52           5         304   \n",
       "5  WormBase:WBGene00006787     unc-52           5         304   \n",
       "6  WormBase:WBGene00006787     unc-52           5         304   \n",
       "7  WormBase:WBGene00006787     unc-52           5         304   \n",
       "8  WormBase:WBGene00006787     unc-52           5         304   \n",
       "9  WormBase:WBGene00006787     unc-52           5         304   \n",
       "\n",
       "                         relation                        tail  \n",
       "0                  interacts with     WormBase:WBGene00003929  \n",
       "1                  interacts with     WormBase:WBGene00006789  \n",
       "2  in orthology relationship with  ENSEMBL:ENSSSCG00000015555  \n",
       "3  in orthology relationship with      ZFIN:ZDB-GENE-021226-3  \n",
       "4  in orthology relationship with  ENSEMBL:ENSOANG00000001050  \n",
       "5                causes condition         WBPhenotype:0001171  \n",
       "6                causes condition         WBPhenotype:0001425  \n",
       "7                causes condition         WBPhenotype:0000781  \n",
       "8  in orthology relationship with  ENSEMBL:ENSPTRG00000021480  \n",
       "9  in orthology relationship with  ENSEMBL:ENSSSCG00000015556  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kg_df = pd.merge(edges, nodes,\n",
    "                 left_on='head', right_on='id',\n",
    "                 how='inner')\n",
    "kg_df = kg_df[['head', 'label', 'semantic_id', 'index_id', 'relation', 'tail']]\n",
    "kg_df.rename(columns={'label': 'label_head', 'semantic_id': 'class_head', 'index_id': 'index_head'}, inplace=True)\n",
    "kg_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>pat-2</td>\n",
       "      <td>5</td>\n",
       "      <td>1542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00006789</td>\n",
       "      <td>unc-54</td>\n",
       "      <td>5</td>\n",
       "      <td>6544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000015555</td>\n",
       "      <td>LAMC1</td>\n",
       "      <td>5</td>\n",
       "      <td>9268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ZFIN:ZDB-GENE-021226-3</td>\n",
       "      <td>lamc1</td>\n",
       "      <td>5</td>\n",
       "      <td>5387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "      <td>5</td>\n",
       "      <td>2204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001171</td>\n",
       "      <td>shortened life span</td>\n",
       "      <td>9</td>\n",
       "      <td>5475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001425</td>\n",
       "      <td>receptor mediated endocytosis defective</td>\n",
       "      <td>9</td>\n",
       "      <td>3057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0000781</td>\n",
       "      <td>body wall muscle thin filament variant</td>\n",
       "      <td>9</td>\n",
       "      <td>7821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSPTRG00000021480</td>\n",
       "      <td>ENSEMBL:ENSPTRG00000021480</td>\n",
       "      <td>5</td>\n",
       "      <td>6347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000015556</td>\n",
       "      <td>LAMC2</td>\n",
       "      <td>5</td>\n",
       "      <td>769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      head label_head  class_head  index_head   \n",
       "0  WormBase:WBGene00006787     unc-52           5         304  \\\n",
       "1  WormBase:WBGene00006787     unc-52           5         304   \n",
       "2  WormBase:WBGene00006787     unc-52           5         304   \n",
       "3  WormBase:WBGene00006787     unc-52           5         304   \n",
       "4  WormBase:WBGene00006787     unc-52           5         304   \n",
       "5  WormBase:WBGene00006787     unc-52           5         304   \n",
       "6  WormBase:WBGene00006787     unc-52           5         304   \n",
       "7  WormBase:WBGene00006787     unc-52           5         304   \n",
       "8  WormBase:WBGene00006787     unc-52           5         304   \n",
       "9  WormBase:WBGene00006787     unc-52           5         304   \n",
       "\n",
       "                         relation                        tail   \n",
       "0                  interacts with     WormBase:WBGene00003929  \\\n",
       "1                  interacts with     WormBase:WBGene00006789   \n",
       "2  in orthology relationship with  ENSEMBL:ENSSSCG00000015555   \n",
       "3  in orthology relationship with      ZFIN:ZDB-GENE-021226-3   \n",
       "4  in orthology relationship with  ENSEMBL:ENSOANG00000001050   \n",
       "5                causes condition         WBPhenotype:0001171   \n",
       "6                causes condition         WBPhenotype:0001425   \n",
       "7                causes condition         WBPhenotype:0000781   \n",
       "8  in orthology relationship with  ENSEMBL:ENSPTRG00000021480   \n",
       "9  in orthology relationship with  ENSEMBL:ENSSSCG00000015556   \n",
       "\n",
       "                                label_tail  class_tail  index_tail  \n",
       "0                                    pat-2           5        1542  \n",
       "1                                   unc-54           5        6544  \n",
       "2                                    LAMC1           5        9268  \n",
       "3                                    lamc1           5        5387  \n",
       "4               ENSEMBL:ENSOANG00000001050           5        2204  \n",
       "5                      shortened life span           9        5475  \n",
       "6  receptor mediated endocytosis defective           9        3057  \n",
       "7   body wall muscle thin filament variant           9        7821  \n",
       "8               ENSEMBL:ENSPTRG00000021480           5        6347  \n",
       "9                                    LAMC2           5         769  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kg_df = pd.merge(kg_df, nodes,\n",
    "                 left_on='tail', right_on='id',\n",
    "                 how='left')\n",
    "kg_df = kg_df[['head', 'label_head', 'class_head', 'index_head', 'relation', 'tail', 'label', 'semantic_id', 'index_id']]\n",
    "kg_df.rename(columns={'label': 'label_tail', 'semantic_id': 'class_tail', 'index_id': 'index_tail'}, inplace=True)\n",
    "kg_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Encode string values of relation labels as numeric values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>pat-2</td>\n",
       "      <td>5</td>\n",
       "      <td>1542</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00006789</td>\n",
       "      <td>unc-54</td>\n",
       "      <td>5</td>\n",
       "      <td>6544</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000015555</td>\n",
       "      <td>LAMC1</td>\n",
       "      <td>5</td>\n",
       "      <td>9268</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ZFIN:ZDB-GENE-021226-3</td>\n",
       "      <td>lamc1</td>\n",
       "      <td>5</td>\n",
       "      <td>5387</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "      <td>5</td>\n",
       "      <td>2204</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001171</td>\n",
       "      <td>shortened life span</td>\n",
       "      <td>9</td>\n",
       "      <td>5475</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0001425</td>\n",
       "      <td>receptor mediated endocytosis defective</td>\n",
       "      <td>9</td>\n",
       "      <td>3057</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>WBPhenotype:0000781</td>\n",
       "      <td>body wall muscle thin filament variant</td>\n",
       "      <td>9</td>\n",
       "      <td>7821</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSPTRG00000021480</td>\n",
       "      <td>ENSEMBL:ENSPTRG00000021480</td>\n",
       "      <td>5</td>\n",
       "      <td>6347</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000015556</td>\n",
       "      <td>LAMC2</td>\n",
       "      <td>5</td>\n",
       "      <td>769</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      head label_head  class_head  index_head   \n",
       "0  WormBase:WBGene00006787     unc-52           5         304  \\\n",
       "1  WormBase:WBGene00006787     unc-52           5         304   \n",
       "2  WormBase:WBGene00006787     unc-52           5         304   \n",
       "3  WormBase:WBGene00006787     unc-52           5         304   \n",
       "4  WormBase:WBGene00006787     unc-52           5         304   \n",
       "5  WormBase:WBGene00006787     unc-52           5         304   \n",
       "6  WormBase:WBGene00006787     unc-52           5         304   \n",
       "7  WormBase:WBGene00006787     unc-52           5         304   \n",
       "8  WormBase:WBGene00006787     unc-52           5         304   \n",
       "9  WormBase:WBGene00006787     unc-52           5         304   \n",
       "\n",
       "                         relation                        tail   \n",
       "0                  interacts with     WormBase:WBGene00003929  \\\n",
       "1                  interacts with     WormBase:WBGene00006789   \n",
       "2  in orthology relationship with  ENSEMBL:ENSSSCG00000015555   \n",
       "3  in orthology relationship with      ZFIN:ZDB-GENE-021226-3   \n",
       "4  in orthology relationship with  ENSEMBL:ENSOANG00000001050   \n",
       "5                causes condition         WBPhenotype:0001171   \n",
       "6                causes condition         WBPhenotype:0001425   \n",
       "7                causes condition         WBPhenotype:0000781   \n",
       "8  in orthology relationship with  ENSEMBL:ENSPTRG00000021480   \n",
       "9  in orthology relationship with  ENSEMBL:ENSSSCG00000015556   \n",
       "\n",
       "                                label_tail  class_tail  index_tail  type  \n",
       "0                                    pat-2           5        1542     0  \n",
       "1                                   unc-54           5        6544     0  \n",
       "2                                    LAMC1           5        9268     1  \n",
       "3                                    lamc1           5        5387     1  \n",
       "4               ENSEMBL:ENSOANG00000001050           5        2204     1  \n",
       "5                      shortened life span           9        5475     2  \n",
       "6  receptor mediated endocytosis defective           9        3057     2  \n",
       "7   body wall muscle thin filament variant           9        7821     2  \n",
       "8               ENSEMBL:ENSPTRG00000021480           5        6347     1  \n",
       "9                                    LAMC2           5         769     1  "
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kg_df['type'], relation_labels = pd.factorize(kg_df['relation'])\n",
    "kg_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['interacts with', 'in orthology relationship with', 'causes condition',\n",
       "       'involved in', 'found in', 'is part of', 'enables', 'colocalizes with',\n",
       "       'has role in modeling', 'contributes to condition', 'expresses gene',\n",
       "       'has affected feature', 'is of', 'likely causes condition',\n",
       "       'is variant in', 'is allele of', 'is substance that treats', 'targets',\n",
       "       'associated with phenotype', 'is product of', 'has genotype'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "relation_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save new dataframe containing all edges into csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 10275 nodes in total before change\n",
      "There are 10237 nodes after change\n",
      "There are 38 gene product nodes removed after change\n",
      "There are 85992 edges in total before change\n",
      "There are 85954 edges after change\n",
      "There are 38 gene product related rows removed after change\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\rosa-\\AppData\\Local\\Temp\\ipykernel_20476\\679677061.py:21: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  no_gene_products_nodes['index_id'] = no_gene_products_nodes.index\n",
      "C:\\Users\\rosa-\\AppData\\Local\\Temp\\ipykernel_20476\\679677061.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  no_gene_products_nodes.drop(['index'], axis=1, inplace=True)\n",
      "C:\\Users\\rosa-\\AppData\\Local\\Temp\\ipykernel_20476\\679677061.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  no_gene_products_nodes['semantic_id'] = no_gene_products_nodes['semantic'].cat.codes\n"
     ]
    }
   ],
   "source": [
    "if concept_changes == NO_GENE_PRODUCTS and dataset_nr == 2:\n",
    "    all_edges = kg_df.copy()\n",
    "    \n",
    "    all_gene_products = all_edges.loc[all_edges['type'] == list(relation_labels).index('is product of')]\n",
    "    all_targeting_drugs = all_edges.loc[all_edges['type'] == list(relation_labels).index('targets')]\n",
    "    \n",
    "    merged_df = pd.merge(all_targeting_drugs, all_gene_products, how='left', left_on='tail', right_on='head', indicator=True)\n",
    "    result_df = merged_df[merged_df['_merge'] == 'both']\n",
    "    \n",
    "    new_edges = result_df[['head_x', 'label_head_x', 'class_head_x', 'index_head_x', 'relation_x', 'tail_y', 'label_tail_y', 'class_tail_y', 'index_tail_y', 'type_y']]\n",
    "    drug_gene_edges = new_edges.rename(columns={'head_x': 'head', 'label_head_x': 'label_head', 'class_head_x': 'class_head', 'index_head_x': 'index_head', \n",
    "                                                'relation_x': 'relation', 'tail_y': 'tail', 'label_tail_y': 'label_tail', 'class_tail_y': 'class_tail', 'index_tail_y': 'index_tail', 'type_y': 'type'})\n",
    "    \n",
    "    \n",
    "    all_non_gene_products = all_edges.loc[(all_edges['type'] != list(relation_labels).index('is product of')) & (all_edges['type'] != list(relation_labels).index('targets'))]\n",
    "    all_new_edges = pd.concat([all_non_gene_products, drug_gene_edges], ignore_index=True)\n",
    "\n",
    "    gene_products_nodes = nodes.loc[nodes['semantic'] == 'gene product']\n",
    "    no_gene_products_nodes = nodes.loc[nodes['semantic'] != 'gene product']\n",
    "    no_gene_products_nodes.reset_index(inplace=True)\n",
    "    no_gene_products_nodes['index_id'] = no_gene_products_nodes.index\n",
    "    \n",
    "    # Remap indices of head and tail of edges with node indices\n",
    "    reindexed_left = pd.merge(all_new_edges, no_gene_products_nodes, how='inner', left_on='index_head', right_on='index')[['head', 'label_head', 'class_head', 'index_id', \n",
    "                                                                                                                           'relation', \n",
    "                                                                                                                           'tail', 'label_tail', 'class_tail', 'index_tail', 'type']]\n",
    "    reindexed_left.rename(columns={'index_id': 'index_head'}, inplace=True)\n",
    "    reindexed_left_right = pd.merge(reindexed_left, no_gene_products_nodes, how='inner', left_on='index_tail', right_on='index')[['head', 'label_head', 'class_head', 'index_head', \n",
    "                                                                                                                                  'relation', \n",
    "                                                                                                                                  'tail', 'label_tail', 'class_tail', 'index_id', 'type']]\n",
    "    reindexed_left_right.rename(columns={'index_id': 'index_tail'}, inplace=True)\n",
    "    \n",
    "    no_gene_products_nodes.drop(['index'], axis=1, inplace=True)\n",
    "    \n",
    "    # Remap relation type indices\n",
    "    reindexed_left_right['type'], relation_labels_new = pd.factorize(reindexed_left_right['relation'])\n",
    "    \n",
    "    # Remap node type indices\n",
    "    no_gene_products_nodes['semantic_id'] = no_gene_products_nodes['semantic'].cat.codes\n",
    "    \n",
    "    print(f'There are {nodes.shape[0]} nodes in total before change')\n",
    "    print(f'There are {no_gene_products_nodes.shape[0]} nodes after change')\n",
    "    print(f'There are {gene_products_nodes.shape[0]} gene product nodes removed after change')\n",
    "\n",
    "    print(f'There are {kg_df.shape[0]} edges in total before change')\n",
    "    print(f'There are {reindexed_left_right.shape[0]} edges after change')\n",
    "    print(f'There are {all_gene_products.shape[0]} gene product related rows removed after change')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>301</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>pat-2</td>\n",
       "      <td>5</td>\n",
       "      <td>1537</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WormBase:WBGene00000051</td>\n",
       "      <td>acr-12</td>\n",
       "      <td>5</td>\n",
       "      <td>8381</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>pat-2</td>\n",
       "      <td>5</td>\n",
       "      <td>1537</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WormBase:WBGene00001328</td>\n",
       "      <td>epi-1</td>\n",
       "      <td>5</td>\n",
       "      <td>3940</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>pat-2</td>\n",
       "      <td>5</td>\n",
       "      <td>1537</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WormBase:WBGene00000042</td>\n",
       "      <td>acr-2</td>\n",
       "      <td>5</td>\n",
       "      <td>657</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>pat-2</td>\n",
       "      <td>5</td>\n",
       "      <td>1537</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FlyBase:FBgn0001250</td>\n",
       "      <td>if (fruit fly)</td>\n",
       "      <td>5</td>\n",
       "      <td>458</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>pat-2</td>\n",
       "      <td>5</td>\n",
       "      <td>1537</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85949</th>\n",
       "      <td>dbSNP:rs56979833</td>\n",
       "      <td>rs56979833-?</td>\n",
       "      <td>11</td>\n",
       "      <td>5590</td>\n",
       "      <td>has affected feature</td>\n",
       "      <td>ENSEMBL:ENSG00000259647</td>\n",
       "      <td>ENSEMBL:ENSG00000259647</td>\n",
       "      <td>5</td>\n",
       "      <td>911</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85950</th>\n",
       "      <td>dbSNP:rs35396788</td>\n",
       "      <td>rs35396788-?</td>\n",
       "      <td>11</td>\n",
       "      <td>4731</td>\n",
       "      <td>has affected feature</td>\n",
       "      <td>HGNC:26832</td>\n",
       "      <td>TAPT1-AS1</td>\n",
       "      <td>5</td>\n",
       "      <td>3519</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85951</th>\n",
       "      <td>dbSNP:rs9975725</td>\n",
       "      <td>rs9975725-?</td>\n",
       "      <td>11</td>\n",
       "      <td>7460</td>\n",
       "      <td>has affected feature</td>\n",
       "      <td>ENSEMBL:ENSG00000231236</td>\n",
       "      <td>ENSEMBL:ENSG00000231236</td>\n",
       "      <td>5</td>\n",
       "      <td>1462</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85952</th>\n",
       "      <td>dbSNP:rs35693284</td>\n",
       "      <td>rs35693284-?</td>\n",
       "      <td>11</td>\n",
       "      <td>9142</td>\n",
       "      <td>has affected feature</td>\n",
       "      <td>ENSEMBL:ENSG00000260569</td>\n",
       "      <td>ENSEMBL:ENSG00000260569</td>\n",
       "      <td>5</td>\n",
       "      <td>4077</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85953</th>\n",
       "      <td>dbSNP:rs2725797</td>\n",
       "      <td>rs2725797-?</td>\n",
       "      <td>11</td>\n",
       "      <td>2410</td>\n",
       "      <td>has affected feature</td>\n",
       "      <td>HGNC:33796</td>\n",
       "      <td>LINC02694</td>\n",
       "      <td>5</td>\n",
       "      <td>214</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85954 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          head      label_head  class_head  index_head   \n",
       "0      WormBase:WBGene00006787          unc-52           5         301  \\\n",
       "1      WormBase:WBGene00000051          acr-12           5        8381   \n",
       "2      WormBase:WBGene00001328           epi-1           5        3940   \n",
       "3      WormBase:WBGene00000042           acr-2           5         657   \n",
       "4          FlyBase:FBgn0001250  if (fruit fly)           5         458   \n",
       "...                        ...             ...         ...         ...   \n",
       "85949         dbSNP:rs56979833    rs56979833-?          11        5590   \n",
       "85950         dbSNP:rs35396788    rs35396788-?          11        4731   \n",
       "85951          dbSNP:rs9975725     rs9975725-?          11        7460   \n",
       "85952         dbSNP:rs35693284    rs35693284-?          11        9142   \n",
       "85953          dbSNP:rs2725797     rs2725797-?          11        2410   \n",
       "\n",
       "                             relation                     tail   \n",
       "0                      interacts with  WormBase:WBGene00003929  \\\n",
       "1                      interacts with  WormBase:WBGene00003929   \n",
       "2                      interacts with  WormBase:WBGene00003929   \n",
       "3                      interacts with  WormBase:WBGene00003929   \n",
       "4      in orthology relationship with  WormBase:WBGene00003929   \n",
       "...                               ...                      ...   \n",
       "85949            has affected feature  ENSEMBL:ENSG00000259647   \n",
       "85950            has affected feature               HGNC:26832   \n",
       "85951            has affected feature  ENSEMBL:ENSG00000231236   \n",
       "85952            has affected feature  ENSEMBL:ENSG00000260569   \n",
       "85953            has affected feature               HGNC:33796   \n",
       "\n",
       "                    label_tail  class_tail  index_tail  type  \n",
       "0                        pat-2           5        1537     0  \n",
       "1                        pat-2           5        1537     0  \n",
       "2                        pat-2           5        1537     0  \n",
       "3                        pat-2           5        1537     0  \n",
       "4                        pat-2           5        1537     1  \n",
       "...                        ...         ...         ...   ...  \n",
       "85949  ENSEMBL:ENSG00000259647           5         911    14  \n",
       "85950                TAPT1-AS1           5        3519    14  \n",
       "85951  ENSEMBL:ENSG00000231236           5        1462    14  \n",
       "85952  ENSEMBL:ENSG00000260569           5        4077    14  \n",
       "85953                LINC02694           5         214    14  \n",
       "\n",
       "[85954 rows x 10 columns]"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reindexed_left_right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index_id</th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9142</th>\n",
       "      <td>9142</td>\n",
       "      <td>dbSNP:rs35693284</td>\n",
       "      <td>variant</td>\n",
       "      <td>rs35693284-?</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      index_id                id semantic         label  semantic_id\n",
       "9142      9142  dbSNP:rs35693284  variant  rs35693284-?           11"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_gene_products_nodes.loc[no_gene_products_nodes['index_id'] == 9142]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Changed from 21 types of relations to 20 types of relations.\n",
      "Changed from 12 types of nodes to 11 nodes.\n"
     ]
    }
   ],
   "source": [
    "print(f'Changed from {len(relation_labels)} types of relations to {len(relation_labels_new)} types of relations.')\n",
    "print(f\"Changed from {len(node_semantic_dict)} types of nodes to {no_gene_products_nodes['semantic_id'].nunique()} nodes.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "if concept_changes == NO_GENE_PRODUCTS and dataset_nr == 2:\n",
    "    reindexed_left_right.to_csv(f'output/indexed_edges_{dataset_nr}{suffix}.csv', index=False)\n",
    "    no_gene_products_nodes.to_csv(f'output/indexed_nodes_{dataset_nr}{suffix}.csv', index=False)\n",
    "    \n",
    "else:\n",
    "    kg_df.to_csv(f'output/indexed_edges_{dataset_nr}{suffix}.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xaifognn",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
