{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "from sklearn.metrics import classification_report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_nr = 2\n",
    "assert dataset_nr == 1 or 2\n",
    "\n",
    "embedding_method = 'e2v'\n",
    "assert embedding_method == 'e2v' or 'm2v'\n",
    "\n",
    "# Additional experiments\n",
    "NO_GENE_PRODUCTS = 'no gene products'\n",
    "SINGLE_RELATION_TYPE = 'only one relation type'\n",
    "RANDOM_FEATURES = 'random node embedding values'\n",
    "concept_changes = None  # set to None when no conceptual changes are wanted\n",
    "\n",
    "if concept_changes:\n",
    "    concept_changes_str = concept_changes\n",
    "else:\n",
    "    concept_changes_str = ''\n",
    "    \n",
    "if concept_changes == NO_GENE_PRODUCTS:\n",
    "    suffix = '_nogeneprods'\n",
    "elif concept_changes == SINGLE_RELATION_TYPE:\n",
    "    suffix = '_singlerel'\n",
    "elif concept_changes == RANDOM_FEATURES:\n",
    "    suffix = '_randomfeat'\n",
    "else:\n",
    "    suffix = ''\n",
    "\n",
    "seeded_emb = False\n",
    "\n",
    "if seeded_emb:\n",
    "    fixed_emb = '_seeded'\n",
    "else:\n",
    "    fixed_emb = ''\n",
    "    \n",
    "if seeded_emb:\n",
    "    title_seeded = ' with fixed node embeddings'\n",
    "else:\n",
    "    title_seeded = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get Result Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\n",
      "A total of 10 runs will be included in the analysis.\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\\run_001\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\\run_002\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\\run_003\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\\run_004\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\\run_005\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\\run_006\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\\run_007\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\\run_008\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\\run_009\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\\run_010\\pred\n"
     ]
    }
   ],
   "source": [
    "curr_working_dir = os.getcwd()\n",
    "curr_output_dir = os.path.join(curr_working_dir, 'output')\n",
    "dataset_output_dir = os.path.join(curr_output_dir, f'g{dataset_nr}_{embedding_method}{fixed_emb}{suffix}')\n",
    "\n",
    "if not os.path.exists(dataset_output_dir):\n",
    "    print('First, run the edge2vec embedding script. Then, run this script.')\n",
    "else:\n",
    "    print(f'Output folder for dataset {dataset_nr} exists and will be loaded: {dataset_output_dir}')\n",
    "    \n",
    "run_folders_list = []\n",
    "for item in os.listdir(dataset_output_dir):\n",
    "    curr_path = os.path.join(dataset_output_dir, item)\n",
    "    if os.path.isdir(curr_path) and 'run' in item:\n",
    "        run_folders_list.append(item)\n",
    "\n",
    "print(f'A total of {len(run_folders_list)} runs will be included in the analysis.')\n",
    "\n",
    "run_folders_paths = []\n",
    "pred_folders_paths = []\n",
    "for run_folder in run_folders_list:\n",
    "    run_path = os.path.join(dataset_output_dir, run_folder)\n",
    "    run_folders_paths.append(run_path)\n",
    "    pred_run_path = os.path.join(run_path, 'pred')\n",
    "    pred_folders_paths.append(pred_run_path)\n",
    "    print(pred_run_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_scores_all_runs = []\n",
    "auc_loss_scores_all_runs = []\n",
    "\n",
    "for run_name, pred_folder in zip(run_folders_list, pred_folders_paths):\n",
    "    with open(f'{pred_folder}/performance_scores_{dataset_nr}_{embedding_method}{suffix}.pkl', 'rb') as f:\n",
    "        loaded_info = pickle.load(f)\n",
    "        \n",
    "    keys = ['AUC Train', 'AUC Validation', 'AUC Test']\n",
    "    for key in keys:\n",
    "        auc_scores = loaded_info[key]\n",
    "        for index, auc_score in enumerate(auc_scores):\n",
    "            auc_scores_per_run = {'run': run_name, 'name': key, 'iteration': index, 'score': auc_score}\n",
    "            auc_scores_all_runs.append(auc_scores_per_run)\n",
    "            auc_loss_scores_all_runs.append(auc_scores_per_run)\n",
    "\n",
    "    loss_scores = loaded_info['Loss']\n",
    "    for index, loss_score in enumerate(loss_scores):\n",
    "        formatted_loss_score = float(np.log10(loss_score))\n",
    "        loss_scores_per_run = {'run': run_name, 'name': 'Cross-Entropy Loss', 'iteration': index, 'score': formatted_loss_score}\n",
    "        auc_loss_scores_all_runs.append(loss_scores_per_run)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot ROC Curves, AUC-ROC Scores and F1 Scores for Each Model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ROC Curves"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g1_e2v\n",
      "A total of 10 runs will be included in the analysis.\n",
      "F1-Score in the test set of dataset 1 and method e2v: 0.9239595390267223\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.89      0.97      0.93     10412\n",
      "           1       0.97      0.88      0.92     10412\n",
      "\n",
      "    accuracy                           0.93     20824\n",
      "   macro avg       0.93      0.93      0.93     20824\n",
      "weighted avg       0.93      0.93      0.93     20824\n",
      "\n",
      "F1-Score in the test set of dataset 1 and method e2v: 0.9186950813049186\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.88      0.97      0.93     10412\n",
      "           1       0.97      0.87      0.92     10412\n",
      "\n",
      "    accuracy                           0.92     20824\n",
      "   macro avg       0.93      0.92      0.92     20824\n",
      "weighted avg       0.93      0.92      0.92     20824\n",
      "\n",
      "F1-Score in the test set of dataset 1 and method e2v: 0.9215429549803607\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.89      0.97      0.93     10412\n",
      "           1       0.97      0.88      0.92     10412\n",
      "\n",
      "    accuracy                           0.93     20824\n",
      "   macro avg       0.93      0.93      0.93     20824\n",
      "weighted avg       0.93      0.93      0.93     20824\n",
      "\n",
      "F1-Score in the test set of dataset 1 and method e2v: 0.9262587219517093\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.90      0.97      0.93     10412\n",
      "           1       0.97      0.89      0.93     10412\n",
      "\n",
      "    accuracy                           0.93     20824\n",
      "   macro avg       0.93      0.93      0.93     20824\n",
      "weighted avg       0.93      0.93      0.93     20824\n",
      "\n",
      "F1-Score in the test set of dataset 1 and method e2v: 0.9242995825159701\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.89      0.97      0.93     10412\n",
      "           1       0.97      0.88      0.92     10412\n",
      "\n",
      "    accuracy                           0.93     20824\n",
      "   macro avg       0.93      0.93      0.93     20824\n",
      "weighted avg       0.93      0.93      0.93     20824\n",
      "\n",
      "F1-Score in the test set of dataset 1 and method e2v: 0.9250213150107829\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.89      0.97      0.93     10412\n",
      "           1       0.97      0.89      0.93     10412\n",
      "\n",
      "    accuracy                           0.93     20824\n",
      "   macro avg       0.93      0.93      0.93     20824\n",
      "weighted avg       0.93      0.93      0.93     20824\n",
      "\n",
      "F1-Score in the test set of dataset 1 and method e2v: 0.9233877076915307\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.89      0.98      0.93     10412\n",
      "           1       0.97      0.88      0.92     10412\n",
      "\n",
      "    accuracy                           0.93     20824\n",
      "   macro avg       0.93      0.93      0.93     20824\n",
      "weighted avg       0.93      0.93      0.93     20824\n",
      "\n",
      "F1-Score in the test set of dataset 1 and method e2v: 0.9269445422092614\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.89      0.98      0.93     10412\n",
      "           1       0.97      0.89      0.93     10412\n",
      "\n",
      "    accuracy                           0.93     20824\n",
      "   macro avg       0.93      0.93      0.93     20824\n",
      "weighted avg       0.93      0.93      0.93     20824\n",
      "\n",
      "F1-Score in the test set of dataset 1 and method e2v: 0.9205981571924877\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.89      0.97      0.93     10412\n",
      "           1       0.97      0.88      0.92     10412\n",
      "\n",
      "    accuracy                           0.92     20824\n",
      "   macro avg       0.93      0.92      0.92     20824\n",
      "weighted avg       0.93      0.92      0.92     20824\n",
      "\n",
      "F1-Score in the test set of dataset 1 and method e2v: 0.9265843125439389\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.90      0.97      0.93     10412\n",
      "           1       0.97      0.89      0.93     10412\n",
      "\n",
      "    accuracy                           0.93     20824\n",
      "   macro avg       0.93      0.93      0.93     20824\n",
      "weighted avg       0.93      0.93      0.93     20824\n",
      "\n",
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\n",
      "A total of 10 runs will be included in the analysis.\n",
      "F1-Score in the test set of dataset 2 and method e2v: 0.908176210682882\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.94      0.87      0.90     11008\n",
      "           1       0.88      0.94      0.91     11008\n",
      "\n",
      "    accuracy                           0.90     22016\n",
      "   macro avg       0.91      0.90      0.90     22016\n",
      "weighted avg       0.91      0.90      0.90     22016\n",
      "\n",
      "F1-Score in the test set of dataset 2 and method e2v: 0.9150641025641024\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.93      0.89      0.91     11008\n",
      "           1       0.90      0.93      0.92     11008\n",
      "\n",
      "    accuracy                           0.91     22016\n",
      "   macro avg       0.91      0.91      0.91     22016\n",
      "weighted avg       0.91      0.91      0.91     22016\n",
      "\n",
      "F1-Score in the test set of dataset 2 and method e2v: 0.910081257255112\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.92      0.89      0.91     11008\n",
      "           1       0.89      0.93      0.91     11008\n",
      "\n",
      "    accuracy                           0.91     22016\n",
      "   macro avg       0.91      0.91      0.91     22016\n",
      "weighted avg       0.91      0.91      0.91     22016\n",
      "\n",
      "F1-Score in the test set of dataset 2 and method e2v: 0.9158878504672897\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.92      0.90      0.91     11008\n",
      "           1       0.91      0.93      0.92     11008\n",
      "\n",
      "    accuracy                           0.91     22016\n",
      "   macro avg       0.92      0.91      0.91     22016\n",
      "weighted avg       0.92      0.91      0.91     22016\n",
      "\n",
      "F1-Score in the test set of dataset 2 and method e2v: 0.9225507402907159\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.93      0.91      0.92     11008\n",
      "           1       0.91      0.93      0.92     11008\n",
      "\n",
      "    accuracy                           0.92     22016\n",
      "   macro avg       0.92      0.92      0.92     22016\n",
      "weighted avg       0.92      0.92      0.92     22016\n",
      "\n",
      "F1-Score in the test set of dataset 2 and method e2v: 0.919266875279392\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.93      0.90      0.92     11008\n",
      "           1       0.90      0.93      0.92     11008\n",
      "\n",
      "    accuracy                           0.92     22016\n",
      "   macro avg       0.92      0.92      0.92     22016\n",
      "weighted avg       0.92      0.92      0.92     22016\n",
      "\n",
      "F1-Score in the test set of dataset 2 and method e2v: 0.9191289418993404\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.92      0.91      0.92     11008\n",
      "           1       0.91      0.92      0.92     11008\n",
      "\n",
      "    accuracy                           0.92     22016\n",
      "   macro avg       0.92      0.92      0.92     22016\n",
      "weighted avg       0.92      0.92      0.92     22016\n",
      "\n",
      "F1-Score in the test set of dataset 2 and method e2v: 0.9183802880695716\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.92      0.92      0.92     11008\n",
      "           1       0.92      0.92      0.92     11008\n",
      "\n",
      "    accuracy                           0.92     22016\n",
      "   macro avg       0.92      0.92      0.92     22016\n",
      "weighted avg       0.92      0.92      0.92     22016\n",
      "\n",
      "F1-Score in the test set of dataset 2 and method e2v: 0.9149114673025125\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.94      0.88      0.91     11008\n",
      "           1       0.89      0.94      0.91     11008\n",
      "\n",
      "    accuracy                           0.91     22016\n",
      "   macro avg       0.91      0.91      0.91     22016\n",
      "weighted avg       0.91      0.91      0.91     22016\n",
      "\n",
      "F1-Score in the test set of dataset 2 and method e2v: 0.917631284916201\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.93      0.90      0.91     11008\n",
      "           1       0.90      0.93      0.92     11008\n",
      "\n",
      "    accuracy                           0.92     22016\n",
      "   macro avg       0.92      0.92      0.92     22016\n",
      "weighted avg       0.92      0.92      0.92     22016\n",
      "\n"
     ]
    }
   ],
   "source": [
    "all_settings = [\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '',\n",
    "     'dataset_nr': 1},\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '',\n",
    "     'dataset_nr': 2}\n",
    "]\n",
    "\n",
    "curr_working_dir = os.getcwd()\n",
    "curr_output_dir = os.path.join(curr_working_dir, 'output')\n",
    "\n",
    "final_test_auc_roc_scores_all_runs_all_models = []\n",
    "roc_curve_all_runs_all_models = []\n",
    "f1_scores_all_runs_all_models = []\n",
    "\n",
    "for setting in all_settings:\n",
    "    curr_dataset_output_dir = os.path.join(curr_output_dir, f'g{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}')\n",
    "\n",
    "    if not os.path.exists(curr_dataset_output_dir):\n",
    "        print('First, run the edge2vec embedding script. Then, run this script.')\n",
    "    else:\n",
    "        print(f'Output folder for dataset {dataset_nr} exists and will be loaded: {curr_dataset_output_dir}')\n",
    "            \n",
    "        run_folders_list = []\n",
    "        for item in os.listdir(curr_dataset_output_dir):\n",
    "            curr_path = os.path.join(curr_dataset_output_dir, item)\n",
    "            if os.path.isdir(curr_path) and 'run' in item:\n",
    "                run_folders_list.append(item)\n",
    "\n",
    "        print(f'A total of {len(run_folders_list)} runs will be included in the analysis.')\n",
    "\n",
    "        run_folders_paths = []\n",
    "        curr_pred_folders_paths = []\n",
    "        for run_folder in run_folders_list:\n",
    "            run_path = os.path.join(curr_dataset_output_dir, run_folder)\n",
    "            run_folders_paths.append(run_path)\n",
    "            pred_run_path = os.path.join(run_path, 'pred')\n",
    "            curr_pred_folders_paths.append(pred_run_path)\n",
    "            \n",
    "        for run_name, pred_folder in zip(run_folders_list, curr_pred_folders_paths):\n",
    "            with open(f'{pred_folder}/performance_scores_{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}.pkl', 'rb') as f:\n",
    "                loaded_info = pickle.load(f)\n",
    "\n",
    "            auc_roc_score = loaded_info['ROC AUC Score']\n",
    "            formatted_auc_roc_score = float(auc_roc_score)\n",
    "            auc_roc_score_per_run = {'Model': f'g{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}', \n",
    "                                        'ROC AUC Score': formatted_auc_roc_score}\n",
    "            final_test_auc_roc_scores_all_runs_all_models.append(auc_roc_score_per_run)\n",
    "\n",
    "            roc_fpr_scores = loaded_info['ROC FPR']\n",
    "            roc_tpr_scores = loaded_info['ROC TPR']\n",
    "            \n",
    "            for fpr, tpr in zip(roc_fpr_scores, roc_tpr_scores):\n",
    "                auc_per_threshold_per_run = {'Model': f'g{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}',\n",
    "                                             'ROC FPR': fpr, 'ROC TPR': tpr}\n",
    "                roc_curve_all_runs_all_models.append(auc_per_threshold_per_run)\n",
    "\n",
    "            f1_score = loaded_info['F1 Score']\n",
    "            formatted_f1_score = float(f1_score)\n",
    "            f1_score_per_run = {'Model': f'g{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}', \n",
    "                                'F1 Score': formatted_f1_score}\n",
    "            f1_scores_all_runs_all_models.append(f1_score_per_run)\n",
    "\n",
    "            print(f'F1-Score in the test set of dataset {setting[\"dataset_nr\"]} and method {setting[\"embedding_method\"]}:', f1_score)\n",
    "            print(classification_report(loaded_info['True Labels'], loaded_info['Predicted Labels']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>ROC FPR</th>\n",
       "      <th>ROC TPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36427</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.875091</td>\n",
       "      <td>0.999637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36428</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.875091</td>\n",
       "      <td>0.999818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36429</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.934502</td>\n",
       "      <td>0.999818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36430</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.934502</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36431</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36432 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Model   ROC FPR   ROC TPR\n",
       "0      g1_e2v  0.000000  0.000000\n",
       "1      g1_e2v  0.000000  0.000192\n",
       "2      g1_e2v  0.000000  0.006531\n",
       "3      g1_e2v  0.000000  0.006915\n",
       "4      g1_e2v  0.000000  0.010949\n",
       "...       ...       ...       ...\n",
       "36427  g2_e2v  0.875091  0.999637\n",
       "36428  g2_e2v  0.875091  0.999818\n",
       "36429  g2_e2v  0.934502  0.999818\n",
       "36430  g2_e2v  0.934502  1.000000\n",
       "36431  g2_e2v  1.000000  1.000000\n",
       "\n",
       "[36432 rows x 3 columns]"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roc_curve_all_runs_all_models = pd.DataFrame(roc_curve_all_runs_all_models)\n",
    "roc_curve_all_runs_all_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "ax.set_title(f'ROC TPR/ROC FPR Distribution over Each Run for Each Model')\n",
    "sns.scatterplot(data=roc_curve_all_runs_all_models, x=\"ROC FPR\", y=\"ROC TPR\", hue=\"Model\", s=1)\n",
    "ax.set_xlabel('False Positive Rate')\n",
    "ax.set_ylabel('True Positive Rate')\n",
    "ax.legend(markerscale=10)\n",
    "\n",
    "fig.savefig(f'{curr_output_dir}/roc_curves.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## F1 Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>F1 Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.923960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.918695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.921543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.926259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.924300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.925021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.923388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.926945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.920598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.926584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.908176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.915064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.910081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.915888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.922551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.919267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.919129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.918380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.914911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.917631</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Model  F1 Score\n",
       "0   g1_e2v  0.923960\n",
       "1   g1_e2v  0.918695\n",
       "2   g1_e2v  0.921543\n",
       "3   g1_e2v  0.926259\n",
       "4   g1_e2v  0.924300\n",
       "5   g1_e2v  0.925021\n",
       "6   g1_e2v  0.923388\n",
       "7   g1_e2v  0.926945\n",
       "8   g1_e2v  0.920598\n",
       "9   g1_e2v  0.926584\n",
       "10  g2_e2v  0.908176\n",
       "11  g2_e2v  0.915064\n",
       "12  g2_e2v  0.910081\n",
       "13  g2_e2v  0.915888\n",
       "14  g2_e2v  0.922551\n",
       "15  g2_e2v  0.919267\n",
       "16  g2_e2v  0.919129\n",
       "17  g2_e2v  0.918380\n",
       "18  g2_e2v  0.914911\n",
       "19  g2_e2v  0.917631"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f1_scores_all_runs_all_models = pd.DataFrame(f1_scores_all_runs_all_models)\n",
    "f1_scores_all_runs_all_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "ax.set_title(f'F1 Scores Overview for Each Model')\n",
    "sns.barplot(f1_scores_all_runs_all_models, x=\"Model\", y=\"F1 Score\", errorbar=\"sd\", color='cornflowerblue')\n",
    "ax.bar_label(ax.containers[0], fontsize=10, padding=5)\n",
    "ax.set_xlabel('Model Variant')\n",
    "ax.set_ylabel('F1 Score')\n",
    "\n",
    "fig.savefig(f'{curr_output_dir}/f1_scores.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final AUC-ROC Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>ROC AUC Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.981466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.978083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.977793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.979621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.977190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.978637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.976998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.979838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.976336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.978401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.966562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.972659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.970360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.972339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.971157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.973195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.968916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.972354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.970316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.969869</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Model  ROC AUC Score\n",
       "0   g1_e2v       0.981466\n",
       "1   g1_e2v       0.978083\n",
       "2   g1_e2v       0.977793\n",
       "3   g1_e2v       0.979621\n",
       "4   g1_e2v       0.977190\n",
       "5   g1_e2v       0.978637\n",
       "6   g1_e2v       0.976998\n",
       "7   g1_e2v       0.979838\n",
       "8   g1_e2v       0.976336\n",
       "9   g1_e2v       0.978401\n",
       "10  g2_e2v       0.966562\n",
       "11  g2_e2v       0.972659\n",
       "12  g2_e2v       0.970360\n",
       "13  g2_e2v       0.972339\n",
       "14  g2_e2v       0.971157\n",
       "15  g2_e2v       0.973195\n",
       "16  g2_e2v       0.968916\n",
       "17  g2_e2v       0.972354\n",
       "18  g2_e2v       0.970316\n",
       "19  g2_e2v       0.969869"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_test_auc_roc_scores_all_runs_all_models = pd.DataFrame(final_test_auc_roc_scores_all_runs_all_models)\n",
    "final_test_auc_roc_scores_all_runs_all_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "ax.set_title(f'AUC-ROC Scores Overview for Each Model')\n",
    "sns.barplot(final_test_auc_roc_scores_all_runs_all_models, x=\"Model\", y=\"ROC AUC Score\", errorbar=\"sd\")\n",
    "ax.bar_label(ax.containers[0], fontsize=10, padding=10)\n",
    "ax.set_ylim(0.85,1) # 0.85, 1\n",
    "ax.set_xlabel('Model Variant')\n",
    "ax.set_ylabel('AUC-ROC Score')\n",
    "\n",
    "fig.savefig(f'{curr_output_dir}/auc_roc_scores.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Training Curve"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run</th>\n",
       "      <th>name</th>\n",
       "      <th>iteration</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>0</td>\n",
       "      <td>0.784185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>1</td>\n",
       "      <td>0.796646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>2</td>\n",
       "      <td>0.813906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>3</td>\n",
       "      <td>0.831519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>4</td>\n",
       "      <td>0.847915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2965</th>\n",
       "      <td>run_010</td>\n",
       "      <td>AUC Test</td>\n",
       "      <td>94</td>\n",
       "      <td>0.745730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2966</th>\n",
       "      <td>run_010</td>\n",
       "      <td>AUC Test</td>\n",
       "      <td>95</td>\n",
       "      <td>0.747070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2967</th>\n",
       "      <td>run_010</td>\n",
       "      <td>AUC Test</td>\n",
       "      <td>96</td>\n",
       "      <td>0.748329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2968</th>\n",
       "      <td>run_010</td>\n",
       "      <td>AUC Test</td>\n",
       "      <td>97</td>\n",
       "      <td>0.749532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2969</th>\n",
       "      <td>run_010</td>\n",
       "      <td>AUC Test</td>\n",
       "      <td>98</td>\n",
       "      <td>0.750787</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2970 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          run       name  iteration     score\n",
       "0     run_001  AUC Train          0  0.784185\n",
       "1     run_001  AUC Train          1  0.796646\n",
       "2     run_001  AUC Train          2  0.813906\n",
       "3     run_001  AUC Train          3  0.831519\n",
       "4     run_001  AUC Train          4  0.847915\n",
       "...       ...        ...        ...       ...\n",
       "2965  run_010   AUC Test         94  0.745730\n",
       "2966  run_010   AUC Test         95  0.747070\n",
       "2967  run_010   AUC Test         96  0.748329\n",
       "2968  run_010   AUC Test         97  0.749532\n",
       "2969  run_010   AUC Test         98  0.750787\n",
       "\n",
       "[2970 rows x 4 columns]"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "auc_scores_all_runs = pd.DataFrame(auc_scores_all_runs)\n",
    "auc_scores_all_runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run</th>\n",
       "      <th>name</th>\n",
       "      <th>iteration</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>0</td>\n",
       "      <td>0.784185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>1</td>\n",
       "      <td>0.796646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>2</td>\n",
       "      <td>0.813906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>3</td>\n",
       "      <td>0.831519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>4</td>\n",
       "      <td>0.847915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3955</th>\n",
       "      <td>run_010</td>\n",
       "      <td>Cross-Entropy Loss</td>\n",
       "      <td>94</td>\n",
       "      <td>-0.158000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>run_010</td>\n",
       "      <td>Cross-Entropy Loss</td>\n",
       "      <td>95</td>\n",
       "      <td>-0.155441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>run_010</td>\n",
       "      <td>Cross-Entropy Loss</td>\n",
       "      <td>96</td>\n",
       "      <td>-0.159594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>run_010</td>\n",
       "      <td>Cross-Entropy Loss</td>\n",
       "      <td>97</td>\n",
       "      <td>-0.161458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>run_010</td>\n",
       "      <td>Cross-Entropy Loss</td>\n",
       "      <td>98</td>\n",
       "      <td>-0.165230</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3960 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          run                name  iteration     score\n",
       "0     run_001           AUC Train          0  0.784185\n",
       "1     run_001           AUC Train          1  0.796646\n",
       "2     run_001           AUC Train          2  0.813906\n",
       "3     run_001           AUC Train          3  0.831519\n",
       "4     run_001           AUC Train          4  0.847915\n",
       "...       ...                 ...        ...       ...\n",
       "3955  run_010  Cross-Entropy Loss         94 -0.158000\n",
       "3956  run_010  Cross-Entropy Loss         95 -0.155441\n",
       "3957  run_010  Cross-Entropy Loss         96 -0.159594\n",
       "3958  run_010  Cross-Entropy Loss         97 -0.161458\n",
       "3959  run_010  Cross-Entropy Loss         98 -0.165230\n",
       "\n",
       "[3960 rows x 4 columns]"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "auc_loss_scores_all_runs = pd.DataFrame(auc_loss_scores_all_runs)\n",
    "auc_loss_scores_all_runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_randomfeat\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x800 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "\n",
    "ax.set_title(f'Training curve on dataset {dataset_nr} {concept_changes_str}')\n",
    "sns.lineplot(data=auc_scores_all_runs, x='iteration', y='score', hue='name')\n",
    "ax.set_ylim(0.85,1) # 0.85, 1\n",
    "ax.set_xlabel('Iteration')\n",
    "ax.set_ylabel('AUC-ROC Score')\n",
    "\n",
    "print(dataset_output_dir)\n",
    "fig.savefig(f'{dataset_output_dir}/training_curve.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x800 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.set_title(f'Training curve on dataset {dataset_nr} {concept_changes_str}')\n",
    "sns.lineplot(data=auc_loss_scores_all_runs, x='iteration', y='score', hue='name')\n",
    "ax.set_ylim(top=1)\n",
    "ax.set_xlabel('Iteration')\n",
    "ax.set_ylabel('AUC-ROC/log10(Loss)')\n",
    "\n",
    "fig.savefig(f'{dataset_output_dir}/training_curve_with_loss.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Similarity between top scoring symptom-drug pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_symptom_pairs_per_run = []\n",
    "\n",
    "for index, pred_path in enumerate(pred_folders_paths):\n",
    "    with open(f'{pred_path}/candidates_per_symptom_{dataset_nr}_{embedding_method}{suffix}.pkl', 'rb') as f:\n",
    "        loaded_list = pickle.load(f)\n",
    "        \n",
    "        drug_symptom_pairs = []\n",
    "        \n",
    "        for _, row in loaded_list.iterrows():\n",
    "            symptom_id = row['Symptom']\n",
    "            candidates = row['Candidates']\n",
    "            \n",
    "            for candidate in candidates:\n",
    "                drug_symptom_pairs.append(tuple([symptom_id, candidate]))\n",
    "                \n",
    "        total_drug_symptom_pairs = len(drug_symptom_pairs)\n",
    "    \n",
    "    drug_symptom_pairs_per_run.append(drug_symptom_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run 1</th>\n",
       "      <th>run 2</th>\n",
       "      <th>run 3</th>\n",
       "      <th>run 4</th>\n",
       "      <th>run 5</th>\n",
       "      <th>run 6</th>\n",
       "      <th>run 7</th>\n",
       "      <th>run 8</th>\n",
       "      <th>run 9</th>\n",
       "      <th>run 10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>run 1</th>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>4.938272</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>7.407407</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 2</th>\n",
       "      <td>3.703704</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>7.407407</td>\n",
       "      <td>4.938272</td>\n",
       "      <td>8.641975</td>\n",
       "      <td>6.172840</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>3.703704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 3</th>\n",
       "      <td>4.938272</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>7.407407</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>4.938272</td>\n",
       "      <td>2.469136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 4</th>\n",
       "      <td>2.469136</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>7.407407</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>6.172840</td>\n",
       "      <td>4.938272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 5</th>\n",
       "      <td>2.469136</td>\n",
       "      <td>7.407407</td>\n",
       "      <td>7.407407</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>7.407407</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>6.172840</td>\n",
       "      <td>6.172840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 6</th>\n",
       "      <td>1.234568</td>\n",
       "      <td>4.938272</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>7.407407</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.938272</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>2.469136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 7</th>\n",
       "      <td>7.407407</td>\n",
       "      <td>8.641975</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>7.407407</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>6.172840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 8</th>\n",
       "      <td>1.234568</td>\n",
       "      <td>6.172840</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>4.938272</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>2.469136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 9</th>\n",
       "      <td>1.234568</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>4.938272</td>\n",
       "      <td>6.172840</td>\n",
       "      <td>6.172840</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>1.234568</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2.469136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 10</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.703704</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>4.938272</td>\n",
       "      <td>6.172840</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>6.172840</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>2.469136</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             run 1       run 2       run 3       run 4       run 5   \n",
       "run 1   100.000000    3.703704    4.938272    2.469136    2.469136  \\\n",
       "run 2     3.703704  100.000000    2.469136    1.234568    7.407407   \n",
       "run 3     4.938272    2.469136  100.000000    2.469136    7.407407   \n",
       "run 4     2.469136    1.234568    2.469136  100.000000    3.703704   \n",
       "run 5     2.469136    7.407407    7.407407    3.703704  100.000000   \n",
       "run 6     1.234568    4.938272    2.469136    2.469136    7.407407   \n",
       "run 7     7.407407    8.641975    3.703704    7.407407    3.703704   \n",
       "run 8     1.234568    6.172840    2.469136    1.234568    1.234568   \n",
       "run 9     1.234568    3.703704    4.938272    6.172840    6.172840   \n",
       "run 10    0.000000    3.703704    2.469136    4.938272    6.172840   \n",
       "\n",
       "             run 6       run 7       run 8       run 9      run 10  \n",
       "run 1     1.234568    7.407407    1.234568    1.234568    0.000000  \n",
       "run 2     4.938272    8.641975    6.172840    3.703704    3.703704  \n",
       "run 3     2.469136    3.703704    2.469136    4.938272    2.469136  \n",
       "run 4     2.469136    7.407407    1.234568    6.172840    4.938272  \n",
       "run 5     7.407407    3.703704    1.234568    6.172840    6.172840  \n",
       "run 6   100.000000    0.000000    4.938272    1.234568    2.469136  \n",
       "run 7     0.000000  100.000000    3.703704    1.234568    6.172840  \n",
       "run 8     4.938272    3.703704  100.000000    1.234568    2.469136  \n",
       "run 9     1.234568    1.234568    1.234568  100.000000    2.469136  \n",
       "run 10    2.469136    6.172840    2.469136    2.469136  100.000000  "
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similarity_matrix = {}\n",
    "ratios_non_diagonals = []\n",
    "\n",
    "for index1, pairs1 in enumerate(drug_symptom_pairs_per_run):\n",
    "    \n",
    "    similarities = {}\n",
    "    \n",
    "    for index2, pairs2 in enumerate(drug_symptom_pairs_per_run):\n",
    "        overlap = set([tuple(sorted(ele)) for ele in pairs1]) & set([tuple(sorted(ele)) for ele in pairs2])\n",
    "        ratio_overlap = len(overlap) / total_drug_symptom_pairs * 100\n",
    "        \n",
    "        similarities[f'run {index2+1}'] = ratio_overlap\n",
    "        \n",
    "        if index1 != index2:\n",
    "            ratios_non_diagonals.append(ratio_overlap)\n",
    "        \n",
    "    similarity_matrix[f'run {index1+1}'] = similarities\n",
    "    \n",
    "similarity_matrix_df = pd.DataFrame(similarity_matrix)\n",
    "similarity_matrix_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean: 3.676268861454046\n",
      "Median: 3.7037037037037033\n"
     ]
    }
   ],
   "source": [
    "mean_overlap_ratio = np.mean(ratios_non_diagonals)\n",
    "median_overlap_ratio = np.median(ratios_non_diagonals)\n",
    "print('Mean:', mean_overlap_ratio)\n",
    "print('Median:', median_overlap_ratio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x800 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.set_title(f'Overlap ratio (Mean: {round(mean_overlap_ratio, 2)}, Median: {round(median_overlap_ratio, 2)}) between list of predicted symptom-drug pairs per run on dataset {dataset_nr} {concept_changes_str}')\n",
    "sns.heatmap(similarity_matrix_df, annot=True, fmt='.1f', linewidths=0.5, ax=ax, cmap='RdYlGn')\n",
    "ax.collections[0].set_clim(0,100)\n",
    "\n",
    "fig.savefig(f'{dataset_output_dir}/overlap_between_runs.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 0 symptom-drug pairs that are found in the top list of drug candidates in 10 runs: \n",
      " set()\n"
     ]
    }
   ],
   "source": [
    "for i in range(0, len(drug_symptom_pairs_per_run)):\n",
    "    if i == 0:\n",
    "        overlapping_pairs_all_runs = set(drug_symptom_pairs_per_run[i])\n",
    "    else:\n",
    "        overlapping_pairs_all_runs = overlapping_pairs_all_runs & set(drug_symptom_pairs_per_run[i])\n",
    "            \n",
    "print(f'There are {len(overlapping_pairs_all_runs)} symptom-drug pairs that are found in the top list of drug candidates in {len(drug_symptom_pairs_per_run)} runs: \\n {overlapping_pairs_all_runs}')\n",
    "\n",
    "with open(f'{dataset_output_dir}/symptom_drug_pair_overlapping_all_runs_{dataset_nr}_{embedding_method}{fixed_emb}.pkl', 'wb') as f:\n",
    "    pickle.dump(overlapping_pairs_all_runs, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 0 symptom-drug pairs that are found in the top list of drug candidates in at least 7 of the 10 runs: \n",
      " set()\n"
     ]
    }
   ],
   "source": [
    "same_drug_symptom_pairs_thresholded = set()\n",
    "\n",
    "threshold = 0.7\n",
    "total_runs = len(drug_symptom_pairs_per_run)\n",
    "min_nr_runs = int(threshold * total_runs)\n",
    "\n",
    "for i in range(0, len(drug_symptom_pairs_per_run)):\n",
    "    for pair in drug_symptom_pairs_per_run[i]:\n",
    "        same_pairs = 0\n",
    "        for j in range(0, len(drug_symptom_pairs_per_run)):\n",
    "            for pair_to_compare in drug_symptom_pairs_per_run[j]:\n",
    "                if pair == pair_to_compare:\n",
    "                    same_pairs += 1\n",
    "                        \n",
    "        #print(f'For pair {pair} from run {i}, there are {same_pairs} same pairs found in list of all runs.')\n",
    "        if same_pairs >= min_nr_runs:\n",
    "            same_drug_symptom_pairs_thresholded.add(pair)\n",
    "            \n",
    "print(f'There are {len(same_drug_symptom_pairs_thresholded)} symptom-drug pairs that are found in the top list of drug candidates in at least {min_nr_runs} of the {total_runs} runs: \\n {same_drug_symptom_pairs_thresholded}')\n",
    "\n",
    "with open(f'{dataset_output_dir}/symptom_drug_pair_overlapping_{min_nr_runs}_runs_{dataset_nr}_{embedding_method}{fixed_emb}{suffix}.pkl', 'wb') as f:\n",
    "    pickle.dump(same_drug_symptom_pairs_thresholded, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Symptoms related to DMD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00003929</td>\n",
       "      <td>pat-2</td>\n",
       "      <td>5</td>\n",
       "      <td>1542</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>WormBase:WBGene00006789</td>\n",
       "      <td>unc-54</td>\n",
       "      <td>5</td>\n",
       "      <td>6544</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSSSCG00000015555</td>\n",
       "      <td>LAMC1</td>\n",
       "      <td>5</td>\n",
       "      <td>9268</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ZFIN:ZDB-GENE-021226-3</td>\n",
       "      <td>lamc1</td>\n",
       "      <td>5</td>\n",
       "      <td>5387</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WormBase:WBGene00006787</td>\n",
       "      <td>unc-52</td>\n",
       "      <td>5</td>\n",
       "      <td>304</td>\n",
       "      <td>in orthology relationship with</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "      <td>ENSEMBL:ENSOANG00000001050</td>\n",
       "      <td>5</td>\n",
       "      <td>2204</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85987</th>\n",
       "      <td>458</td>\n",
       "      <td>scopolamine butylbromide</td>\n",
       "      <td>4</td>\n",
       "      <td>5945</td>\n",
       "      <td>targets</td>\n",
       "      <td>P11229</td>\n",
       "      <td>Muscarinic acetylcholine receptor M1</td>\n",
       "      <td>6</td>\n",
       "      <td>5919</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85988</th>\n",
       "      <td>OMIM:300377.0080</td>\n",
       "      <td>DMD, IVS62, A-G, -285</td>\n",
       "      <td>11</td>\n",
       "      <td>1578</td>\n",
       "      <td>is allele of</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>5</td>\n",
       "      <td>3310</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85989</th>\n",
       "      <td>5297</td>\n",
       "      <td>dacomitinib</td>\n",
       "      <td>4</td>\n",
       "      <td>8798</td>\n",
       "      <td>targets</td>\n",
       "      <td>P12931</td>\n",
       "      <td>Proto-oncogene tyrosine-protein kinase Src</td>\n",
       "      <td>6</td>\n",
       "      <td>2379</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85990</th>\n",
       "      <td>ClinVarVariant:981988</td>\n",
       "      <td>NC_000023.11:g.(31875374_31929595)_(31968515_3...</td>\n",
       "      <td>11</td>\n",
       "      <td>8189</td>\n",
       "      <td>has affected feature</td>\n",
       "      <td>HGNC:2928</td>\n",
       "      <td>DMD</td>\n",
       "      <td>5</td>\n",
       "      <td>3310</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85991</th>\n",
       "      <td>ClinVarVariant:584105</td>\n",
       "      <td>NC_000023.11:g.(?_32595737)_(32699313_?)del</td>\n",
       "      <td>11</td>\n",
       "      <td>9866</td>\n",
       "      <td>causes condition</td>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85992 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          head   \n",
       "0      WormBase:WBGene00006787  \\\n",
       "1      WormBase:WBGene00006787   \n",
       "2      WormBase:WBGene00006787   \n",
       "3      WormBase:WBGene00006787   \n",
       "4      WormBase:WBGene00006787   \n",
       "...                        ...   \n",
       "85987                      458   \n",
       "85988         OMIM:300377.0080   \n",
       "85989                     5297   \n",
       "85990    ClinVarVariant:981988   \n",
       "85991    ClinVarVariant:584105   \n",
       "\n",
       "                                              label_head  class_head   \n",
       "0                                                 unc-52           5  \\\n",
       "1                                                 unc-52           5   \n",
       "2                                                 unc-52           5   \n",
       "3                                                 unc-52           5   \n",
       "4                                                 unc-52           5   \n",
       "...                                                  ...         ...   \n",
       "85987                           scopolamine butylbromide           4   \n",
       "85988                              DMD, IVS62, A-G, -285          11   \n",
       "85989                                        dacomitinib           4   \n",
       "85990  NC_000023.11:g.(31875374_31929595)_(31968515_3...          11   \n",
       "85991        NC_000023.11:g.(?_32595737)_(32699313_?)del          11   \n",
       "\n",
       "       index_head                        relation                        tail   \n",
       "0             304                  interacts with     WormBase:WBGene00003929  \\\n",
       "1             304                  interacts with     WormBase:WBGene00006789   \n",
       "2             304  in orthology relationship with  ENSEMBL:ENSSSCG00000015555   \n",
       "3             304  in orthology relationship with      ZFIN:ZDB-GENE-021226-3   \n",
       "4             304  in orthology relationship with  ENSEMBL:ENSOANG00000001050   \n",
       "...           ...                             ...                         ...   \n",
       "85987        5945                         targets                      P11229   \n",
       "85988        1578                    is allele of                   HGNC:2928   \n",
       "85989        8798                         targets                      P12931   \n",
       "85990        8189            has affected feature                   HGNC:2928   \n",
       "85991        9866                causes condition               MONDO:0010679   \n",
       "\n",
       "                                       label_tail  class_tail  index_tail   \n",
       "0                                           pat-2           5        1542  \\\n",
       "1                                          unc-54           5        6544   \n",
       "2                                           LAMC1           5        9268   \n",
       "3                                           lamc1           5        5387   \n",
       "4                      ENSEMBL:ENSOANG00000001050           5        2204   \n",
       "...                                           ...         ...         ...   \n",
       "85987        Muscarinic acetylcholine receptor M1           6        5919   \n",
       "85988                                         DMD           5        3310   \n",
       "85989  Proto-oncogene tyrosine-protein kinase Src           6        2379   \n",
       "85990                                         DMD           5        3310   \n",
       "85991                 Duchenne muscular dystrophy           3        1913   \n",
       "\n",
       "       type  \n",
       "0         0  \n",
       "1         0  \n",
       "2         1  \n",
       "3         1  \n",
       "4         1  \n",
       "...     ...  \n",
       "85987    17  \n",
       "85988    15  \n",
       "85989    17  \n",
       "85990    11  \n",
       "85991     2  \n",
       "\n",
       "[85992 rows x 10 columns]"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges = pd.read_csv(f'output/indexed_edges_{dataset_nr}{suffix}.csv')\n",
    "edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A total of 27 symptoms found that are associated with MONDO:0010679\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>head</th>\n",
       "      <th>label_head</th>\n",
       "      <th>class_head</th>\n",
       "      <th>index_head</th>\n",
       "      <th>relation</th>\n",
       "      <th>tail</th>\n",
       "      <th>label_tail</th>\n",
       "      <th>class_tail</th>\n",
       "      <th>index_tail</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41373</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0003707</td>\n",
       "      <td>Calf muscle pseudohypertrophy</td>\n",
       "      <td>9</td>\n",
       "      <td>5727</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41374</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0003236</td>\n",
       "      <td>Elevated serum creatine kinase</td>\n",
       "      <td>9</td>\n",
       "      <td>1351</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41375</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0001256</td>\n",
       "      <td>Intellectual disability, mild</td>\n",
       "      <td>9</td>\n",
       "      <td>4533</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41376</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0001265</td>\n",
       "      <td>Hyporeflexia</td>\n",
       "      <td>9</td>\n",
       "      <td>6515</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41377</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0001290</td>\n",
       "      <td>Generalized hypotonia</td>\n",
       "      <td>9</td>\n",
       "      <td>5372</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41378</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0003307</td>\n",
       "      <td>Hyperlordosis</td>\n",
       "      <td>9</td>\n",
       "      <td>2737</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41379</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0003202</td>\n",
       "      <td>Skeletal muscle atrophy</td>\n",
       "      <td>9</td>\n",
       "      <td>788</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41380</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0001263</td>\n",
       "      <td>Global developmental delay</td>\n",
       "      <td>9</td>\n",
       "      <td>7259</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41381</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0002791</td>\n",
       "      <td>Hypoventilation</td>\n",
       "      <td>9</td>\n",
       "      <td>1682</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41382</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0001371</td>\n",
       "      <td>Flexion contracture</td>\n",
       "      <td>9</td>\n",
       "      <td>5052</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41383</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0003701</td>\n",
       "      <td>Proximal muscle weakness</td>\n",
       "      <td>9</td>\n",
       "      <td>5762</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41384</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0001270</td>\n",
       "      <td>Motor delay</td>\n",
       "      <td>9</td>\n",
       "      <td>1429</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41386</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0003560</td>\n",
       "      <td>Muscular dystrophy</td>\n",
       "      <td>9</td>\n",
       "      <td>898</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41387</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0008981</td>\n",
       "      <td>Calf muscle hypertrophy</td>\n",
       "      <td>9</td>\n",
       "      <td>4673</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41388</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0002650</td>\n",
       "      <td>Scoliosis</td>\n",
       "      <td>9</td>\n",
       "      <td>4540</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41389</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0001638</td>\n",
       "      <td>Cardiomyopathy</td>\n",
       "      <td>9</td>\n",
       "      <td>3222</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41390</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0003323</td>\n",
       "      <td>Progressive muscle weakness</td>\n",
       "      <td>9</td>\n",
       "      <td>10187</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41391</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0001635</td>\n",
       "      <td>Congestive heart failure</td>\n",
       "      <td>9</td>\n",
       "      <td>1660</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41392</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0000750</td>\n",
       "      <td>Delayed speech and language development</td>\n",
       "      <td>9</td>\n",
       "      <td>7722</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41393</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0003115</td>\n",
       "      <td>Abnormal EKG</td>\n",
       "      <td>9</td>\n",
       "      <td>3941</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41394</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0011675</td>\n",
       "      <td>Arrhythmia</td>\n",
       "      <td>9</td>\n",
       "      <td>4336</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41395</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0003391</td>\n",
       "      <td>Gowers sign</td>\n",
       "      <td>9</td>\n",
       "      <td>8594</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41397</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0002093</td>\n",
       "      <td>Respiratory insufficiency</td>\n",
       "      <td>9</td>\n",
       "      <td>1688</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41398</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0100543</td>\n",
       "      <td>Cognitive impairment</td>\n",
       "      <td>9</td>\n",
       "      <td>9574</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41399</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0001644</td>\n",
       "      <td>Dilated cardiomyopathy</td>\n",
       "      <td>9</td>\n",
       "      <td>556</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41400</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0002515</td>\n",
       "      <td>Waddling gait</td>\n",
       "      <td>9</td>\n",
       "      <td>1474</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41401</th>\n",
       "      <td>MONDO:0010679</td>\n",
       "      <td>Duchenne muscular dystrophy</td>\n",
       "      <td>3</td>\n",
       "      <td>1913</td>\n",
       "      <td>associated with phenotype</td>\n",
       "      <td>HP:0001328</td>\n",
       "      <td>Specific learning disability</td>\n",
       "      <td>9</td>\n",
       "      <td>1174</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                head                   label_head  class_head  index_head   \n",
       "41373  MONDO:0010679  Duchenne muscular dystrophy           3        1913  \\\n",
       "41374  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41375  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41376  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41377  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41378  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41379  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41380  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41381  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41382  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41383  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41384  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41386  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41387  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41388  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41389  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41390  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41391  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41392  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41393  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41394  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41395  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41397  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41398  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41399  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41400  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "41401  MONDO:0010679  Duchenne muscular dystrophy           3        1913   \n",
       "\n",
       "                        relation        tail   \n",
       "41373  associated with phenotype  HP:0003707  \\\n",
       "41374  associated with phenotype  HP:0003236   \n",
       "41375  associated with phenotype  HP:0001256   \n",
       "41376  associated with phenotype  HP:0001265   \n",
       "41377  associated with phenotype  HP:0001290   \n",
       "41378  associated with phenotype  HP:0003307   \n",
       "41379  associated with phenotype  HP:0003202   \n",
       "41380  associated with phenotype  HP:0001263   \n",
       "41381  associated with phenotype  HP:0002791   \n",
       "41382  associated with phenotype  HP:0001371   \n",
       "41383  associated with phenotype  HP:0003701   \n",
       "41384  associated with phenotype  HP:0001270   \n",
       "41386  associated with phenotype  HP:0003560   \n",
       "41387  associated with phenotype  HP:0008981   \n",
       "41388  associated with phenotype  HP:0002650   \n",
       "41389  associated with phenotype  HP:0001638   \n",
       "41390  associated with phenotype  HP:0003323   \n",
       "41391  associated with phenotype  HP:0001635   \n",
       "41392  associated with phenotype  HP:0000750   \n",
       "41393  associated with phenotype  HP:0003115   \n",
       "41394  associated with phenotype  HP:0011675   \n",
       "41395  associated with phenotype  HP:0003391   \n",
       "41397  associated with phenotype  HP:0002093   \n",
       "41398  associated with phenotype  HP:0100543   \n",
       "41399  associated with phenotype  HP:0001644   \n",
       "41400  associated with phenotype  HP:0002515   \n",
       "41401  associated with phenotype  HP:0001328   \n",
       "\n",
       "                                    label_tail  class_tail  index_tail  type  \n",
       "41373            Calf muscle pseudohypertrophy           9        5727    18  \n",
       "41374           Elevated serum creatine kinase           9        1351    18  \n",
       "41375            Intellectual disability, mild           9        4533    18  \n",
       "41376                             Hyporeflexia           9        6515    18  \n",
       "41377                    Generalized hypotonia           9        5372    18  \n",
       "41378                            Hyperlordosis           9        2737    18  \n",
       "41379                  Skeletal muscle atrophy           9         788    18  \n",
       "41380               Global developmental delay           9        7259    18  \n",
       "41381                          Hypoventilation           9        1682    18  \n",
       "41382                      Flexion contracture           9        5052    18  \n",
       "41383                 Proximal muscle weakness           9        5762    18  \n",
       "41384                              Motor delay           9        1429    18  \n",
       "41386                       Muscular dystrophy           9         898    18  \n",
       "41387                  Calf muscle hypertrophy           9        4673    18  \n",
       "41388                                Scoliosis           9        4540    18  \n",
       "41389                           Cardiomyopathy           9        3222    18  \n",
       "41390              Progressive muscle weakness           9       10187    18  \n",
       "41391                 Congestive heart failure           9        1660    18  \n",
       "41392  Delayed speech and language development           9        7722    18  \n",
       "41393                             Abnormal EKG           9        3941    18  \n",
       "41394                               Arrhythmia           9        4336    18  \n",
       "41395                              Gowers sign           9        8594    18  \n",
       "41397                Respiratory insufficiency           9        1688    18  \n",
       "41398                     Cognitive impairment           9        9574    18  \n",
       "41399                   Dilated cardiomyopathy           9         556    18  \n",
       "41400                            Waddling gait           9        1474    18  \n",
       "41401             Specific learning disability           9        1174    18  "
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if dataset_nr == 1:\n",
    "    pheno_rel = 'has phenotype'\n",
    "else:\n",
    "    pheno_rel = 'associated with phenotype'\n",
    "    \n",
    "_, relation_labels = pd.factorize(edges['relation'])\n",
    "\n",
    "disease_ID = 'MONDO:0010679'\n",
    "relation_index = list(relation_labels).index(pheno_rel)\n",
    "\n",
    "symptoms = edges[(edges['head'] == 'MONDO:0010679') & (edges['type'] == relation_index)]\n",
    "\n",
    "print(f'A total of {symptoms.shape[0]} symptoms found that are associated with {disease_ID}')\n",
    "symptoms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [],
   "source": [
    "symptoms_df = symptoms[['tail', 'label_tail']]\n",
    "symptoms_df = symptoms_df.rename(columns={'tail': 'Symptom ID', 'label_tail': 'Symptom'})\n",
    "symptoms_df.to_csv(f'{curr_output_dir}/dmdrelatedsymptoms.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check overlap between each model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MP:0004187</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>cardia bifida</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZP:0100138</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>muscle tendon junction myotome increased amoun...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1346525</td>\n",
       "      <td>gene</td>\n",
       "      <td>Sgcd</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OMIM:300377.0044</td>\n",
       "      <td>variant</td>\n",
       "      <td>DMD, LYS770TER</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZP:0002210</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>posterior lateral line neuromast primordium mi...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10270</th>\n",
       "      <td>ZP:0014934</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>atrioventricular valve development process qua...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10271</th>\n",
       "      <td>ENSEMBL:ENSCAFG00000011207</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000011207</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10272</th>\n",
       "      <td>ENSEMBL:ENSXETG00000039922</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSXETG00000039922</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10273</th>\n",
       "      <td>ENSEMBL:ENSACAG00000010058</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSACAG00000010058</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10274</th>\n",
       "      <td>WBPhenotype:0000019</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>pharyngeal pumping reduced</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10275 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               id   semantic   \n",
       "0                      MP:0004187  phenotype  \\\n",
       "1                      ZP:0100138  phenotype   \n",
       "2                     MGI:1346525       gene   \n",
       "3                OMIM:300377.0044    variant   \n",
       "4                      ZP:0002210  phenotype   \n",
       "...                           ...        ...   \n",
       "10270                  ZP:0014934  phenotype   \n",
       "10271  ENSEMBL:ENSCAFG00000011207       gene   \n",
       "10272  ENSEMBL:ENSXETG00000039922       gene   \n",
       "10273  ENSEMBL:ENSACAG00000010058       gene   \n",
       "10274         WBPhenotype:0000019  phenotype   \n",
       "\n",
       "                                                   label  semantic_id  \n",
       "0                                          cardia bifida            9  \n",
       "1      muscle tendon junction myotome increased amoun...            9  \n",
       "2                                                   Sgcd            5  \n",
       "3                                         DMD, LYS770TER           11  \n",
       "4      posterior lateral line neuromast primordium mi...            9  \n",
       "...                                                  ...          ...  \n",
       "10270  atrioventricular valve development process qua...            9  \n",
       "10271                         ENSEMBL:ENSCAFG00000011207            5  \n",
       "10272                         ENSEMBL:ENSXETG00000039922            5  \n",
       "10273                         ENSEMBL:ENSACAG00000010058            5  \n",
       "10274                         pharyngeal pumping reduced            9  \n",
       "\n",
       "[10275 rows x 4 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = pd.read_csv(f'output/indexed_nodes_{dataset_nr}.csv')\n",
    "nodes.drop('index_id', axis=1, inplace=True)\n",
    "nodes['semantic'] = nodes['semantic'].astype('category')\n",
    "nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>606</td>\n",
       "      <td>drug</td>\n",
       "      <td>chloropyramine</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>3294</td>\n",
       "      <td>drug</td>\n",
       "      <td>iloperidone</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>1982</td>\n",
       "      <td>drug</td>\n",
       "      <td>olanzapine</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>1214</td>\n",
       "      <td>drug</td>\n",
       "      <td>fluphenazine enanthate</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>5303</td>\n",
       "      <td>drug</td>\n",
       "      <td>revefenacin</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10140</th>\n",
       "      <td>3659</td>\n",
       "      <td>drug</td>\n",
       "      <td>yohimbine</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10150</th>\n",
       "      <td>1836</td>\n",
       "      <td>drug</td>\n",
       "      <td>montelukast</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10174</th>\n",
       "      <td>1402</td>\n",
       "      <td>drug</td>\n",
       "      <td>hyoscyamine</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10201</th>\n",
       "      <td>MESH:D000077185</td>\n",
       "      <td>drug</td>\n",
       "      <td>Resveratrol</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10250</th>\n",
       "      <td>987</td>\n",
       "      <td>drug</td>\n",
       "      <td>edetic acid</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>291 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id semantic                   label  semantic_id\n",
       "89                 606     drug          chloropyramine            4\n",
       "98                3294     drug             iloperidone            4\n",
       "103               1982     drug              olanzapine            4\n",
       "129               1214     drug  fluphenazine enanthate            4\n",
       "246               5303     drug             revefenacin            4\n",
       "...                ...      ...                     ...          ...\n",
       "10140             3659     drug               yohimbine            4\n",
       "10150             1836     drug             montelukast            4\n",
       "10174             1402     drug             hyoscyamine            4\n",
       "10201  MESH:D000077185     drug             Resveratrol            4\n",
       "10250              987     drug             edetic acid            4\n",
       "\n",
       "[291 rows x 4 columns]"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if dataset_nr == 1:\n",
    "    drug_semantic = 'DRUG'\n",
    "else:\n",
    "    drug_semantic = 'drug'\n",
    "\n",
    "nodes[nodes['semantic'] == drug_semantic]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g1_e2v\n",
      "{('HP:0011675', '1576'), ('HP:0001635', '231'), ('HP:0002650', '5345'), ('HP:0003115', '231')}\n",
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\n",
      "{('HP:0001644', '1576'), ('HP:0001638', '1576'), ('HP:0001290', '269'), ('HP:0003115', '1576'), ('HP:0003236', '1576'), ('HP:0011675', '1576')}\n"
     ]
    }
   ],
   "source": [
    "all_settings = [\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '',\n",
    "     'dataset_nr': 1},\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '',\n",
    "     'dataset_nr': 2}\n",
    "]\n",
    "\n",
    "curr_working_dir = os.getcwd()\n",
    "curr_output_dir = os.path.join(curr_working_dir, 'output')\n",
    "\n",
    "overlapping_pairs = []\n",
    "\n",
    "for setting in all_settings:\n",
    "    dataset_output_dir = os.path.join(curr_output_dir, f'g{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}')\n",
    "\n",
    "    if not os.path.exists(dataset_output_dir):\n",
    "        print('First, run the edge2vec embedding script. Then, run this script.')\n",
    "    else:\n",
    "        print(f'Output folder for dataset {dataset_nr} exists and will be loaded: {dataset_output_dir}')\n",
    "        \n",
    "        with open(f'{dataset_output_dir}/symptom_drug_pair_overlapping_all_runs_{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}.pkl', 'rb') as f:\n",
    "            loaded_list = pickle.load(f)\n",
    "            print(loaded_list)\n",
    "            overlapping_pairs.append(loaded_list)\n",
    "            \n",
    "            pair_dict_list = []\n",
    "            for pair in loaded_list:\n",
    "                symptom_id, drug_id = pair\n",
    "                symptom_name = nodes.loc[nodes['id'] == symptom_id]['label'].iloc[0]\n",
    "                drug_name = nodes.loc[nodes['id'] == drug_id]['label'].iloc[0]\n",
    "                pair_dict_list.append({'Drug': drug_name, 'Symptom ID': symptom_id, 'Symptom': symptom_name})\n",
    "                \n",
    "            overlapping_all_runs_df = pd.DataFrame(pair_dict_list)\n",
    "            overlapping_all_runs_df.to_csv(f'{dataset_output_dir}/symptom_drug_pairs_all_runs_{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "levosimendan treats Arrhythmia\n"
     ]
    }
   ],
   "source": [
    "dataset1_emb_overlap = overlapping_pairs[0].intersection(overlapping_pairs[1])\n",
    "for pair in dataset1_emb_overlap:\n",
    "    symptom_id, drug_id = pair\n",
    "    symptom_name = nodes.loc[nodes['id'] == symptom_id]['label'].iloc[0]\n",
    "    drug_name = nodes.loc[nodes['id'] == drug_id]['label'].iloc[0]\n",
    "    print(drug_name, 'treats', symptom_name)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xaifognn",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
