{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "from sklearn.metrics import roc_auc_score, roc_curve, f1_score, classification_report, confusion_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_nr = 2\n",
    "assert dataset_nr == 1 or 2\n",
    "\n",
    "embedding_method = 'e2v'\n",
    "assert embedding_method == 'e2v' or 'm2v'\n",
    "\n",
    "seeded_emb = False\n",
    "\n",
    "if seeded_emb:\n",
    "    fixed_emb = '_seeded'\n",
    "else:\n",
    "    fixed_emb = ''\n",
    "    \n",
    "if seeded_emb:\n",
    "    title_seeded = ' with fixed node embeddings'\n",
    "else:\n",
    "    title_seeded = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get Result Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\n",
      "A total of 10 runs will be included in the analysis.\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\run_001\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\run_002\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\run_003\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\run_004\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\run_005\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\run_006\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\run_007\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\run_008\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\run_009\\pred\n",
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\\run_010\\pred\n"
     ]
    }
   ],
   "source": [
    "curr_working_dir = os.getcwd()\n",
    "curr_output_dir = os.path.join(curr_working_dir, 'output')\n",
    "dataset_output_dir = os.path.join(curr_output_dir, f'g{dataset_nr}_{embedding_method}{fixed_emb}')\n",
    "\n",
    "if not os.path.exists(dataset_output_dir):\n",
    "    print('First, run the edge2vec embedding script. Then, run this script.')\n",
    "else:\n",
    "    print(f'Output folder for dataset {dataset_nr} exists and will be loaded: {dataset_output_dir}')\n",
    "    \n",
    "run_folders_list = []\n",
    "for item in os.listdir(dataset_output_dir):\n",
    "    curr_path = os.path.join(dataset_output_dir, item)\n",
    "    if os.path.isdir(curr_path) and 'run' in item:\n",
    "        run_folders_list.append(item)\n",
    "\n",
    "print(f'A total of {len(run_folders_list)} runs will be included in the analysis.')\n",
    "\n",
    "run_folders_paths = []\n",
    "pred_folders_paths = []\n",
    "for run_folder in run_folders_list:\n",
    "    run_path = os.path.join(dataset_output_dir, run_folder)\n",
    "    run_folders_paths.append(run_path)\n",
    "    pred_run_path = os.path.join(run_path, 'pred')\n",
    "    pred_folders_paths.append(pred_run_path)\n",
    "    print(pred_run_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_scores_all_runs = []\n",
    "auc_loss_scores_all_runs = []\n",
    "\n",
    "for run_name, pred_folder in zip(run_folders_list, pred_folders_paths):\n",
    "    with open(f'{pred_folder}/performance_scores_{dataset_nr}_{embedding_method}.pkl', 'rb') as f:\n",
    "        loaded_info = pickle.load(f)\n",
    "        \n",
    "    keys = ['AUC Train', 'AUC Validation', 'AUC Test']\n",
    "    for key in keys:\n",
    "        auc_scores = loaded_info[key]\n",
    "        for index, auc_score in enumerate(auc_scores):\n",
    "            auc_scores_per_run = {'run': run_name, 'name': key, 'iteration': index, 'score': auc_score}\n",
    "            auc_scores_all_runs.append(auc_scores_per_run)\n",
    "            auc_loss_scores_all_runs.append(auc_scores_per_run)\n",
    "\n",
    "    loss_scores = loaded_info['Loss']\n",
    "    for index, loss_score in enumerate(loss_scores):\n",
    "        formatted_loss_score = float(np.log10(loss_score))\n",
    "        loss_scores_per_run = {'run': run_name, 'name': 'Cross-Entropy Loss', 'iteration': index, 'score': formatted_loss_score}\n",
    "        auc_loss_scores_all_runs.append(loss_scores_per_run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 386,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g1_e2v\n",
      "A total of 10 runs will be included in the analysis.\n",
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g1_e2v_seeded\n",
      "A total of 10 runs will be included in the analysis.\n",
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\n",
      "A total of 10 runs will be included in the analysis.\n",
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_seeded\n",
      "A total of 10 runs will be included in the analysis.\n"
     ]
    }
   ],
   "source": [
    "all_settings = [\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '',\n",
    "     'dataset_nr': 1},\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '_seeded',\n",
    "     'dataset_nr': 1},\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '',\n",
    "     'dataset_nr': 2},\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '_seeded',\n",
    "     'dataset_nr': 2}\n",
    "]\n",
    "\n",
    "curr_working_dir = os.getcwd()\n",
    "curr_output_dir = os.path.join(curr_working_dir, 'output')\n",
    "\n",
    "final_test_auc_roc_scores_all_runs_all_models = []\n",
    "roc_curve_all_runs_all_models = []\n",
    "f1_scores_all_runs_all_models = []\n",
    "\n",
    "for setting in all_settings:\n",
    "    curr_dataset_output_dir = os.path.join(curr_output_dir, f'g{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}')\n",
    "\n",
    "    if not os.path.exists(curr_dataset_output_dir):\n",
    "        print('First, run the edge2vec embedding script. Then, run this script.')\n",
    "    else:\n",
    "        print(f'Output folder for dataset {dataset_nr} exists and will be loaded: {curr_dataset_output_dir}')\n",
    "            \n",
    "        run_folders_list = []\n",
    "        for item in os.listdir(curr_dataset_output_dir):\n",
    "            curr_path = os.path.join(curr_dataset_output_dir, item)\n",
    "            if os.path.isdir(curr_path) and 'run' in item:\n",
    "                run_folders_list.append(item)\n",
    "\n",
    "        print(f'A total of {len(run_folders_list)} runs will be included in the analysis.')\n",
    "\n",
    "        run_folders_paths = []\n",
    "        curr_pred_folders_paths = []\n",
    "        for run_folder in run_folders_list:\n",
    "            run_path = os.path.join(curr_dataset_output_dir, run_folder)\n",
    "            run_folders_paths.append(run_path)\n",
    "            pred_run_path = os.path.join(run_path, 'pred')\n",
    "            curr_pred_folders_paths.append(pred_run_path)\n",
    "            \n",
    "        for run_name, pred_folder in zip(run_folders_list, curr_pred_folders_paths):\n",
    "            with open(f'{pred_folder}/performance_scores_{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}.pkl', 'rb') as f:\n",
    "                loaded_info = pickle.load(f)\n",
    "\n",
    "            auc_roc_score = loaded_info['ROC AUC Score']\n",
    "            formatted_auc_roc_score = float(auc_roc_score)\n",
    "            auc_roc_score_per_run = {'Model': f'g{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}', \n",
    "                                        'ROC AUC Score': formatted_auc_roc_score}\n",
    "            final_test_auc_roc_scores_all_runs_all_models.append(auc_roc_score_per_run)\n",
    "\n",
    "            roc_fpr_scores = loaded_info['ROC FPR']\n",
    "            roc_tpr_scores = loaded_info['ROC TPR']\n",
    "            \n",
    "            for fpr, tpr in zip(roc_fpr_scores, roc_tpr_scores):\n",
    "                auc_per_threshold_per_run = {'Model': f'g{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}',\n",
    "                                             'ROC FPR': fpr, 'ROC TPR': tpr}\n",
    "                roc_curve_all_runs_all_models.append(auc_per_threshold_per_run)\n",
    "\n",
    "            f1_score = loaded_info['F1 Score']\n",
    "            formatted_f1_score = float(f1_score)\n",
    "            f1_score_per_run = {'Model': f'g{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}', \n",
    "                                'F1 Score': formatted_f1_score}\n",
    "            f1_scores_all_runs_all_models.append(f1_score_per_run)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot ROC Curves for Each Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 387,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>ROC FPR</th>\n",
       "      <th>ROC TPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.026715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.027100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.050932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80491</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.99682</td>\n",
       "      <td>0.999637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80492</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.99682</td>\n",
       "      <td>0.999818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80493</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.99891</td>\n",
       "      <td>0.999818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80494</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.99891</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80495</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80496 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Model  ROC FPR   ROC TPR\n",
       "0             g1_e2v  0.00000  0.000000\n",
       "1             g1_e2v  0.00000  0.000192\n",
       "2             g1_e2v  0.00000  0.026715\n",
       "3             g1_e2v  0.00000  0.027100\n",
       "4             g1_e2v  0.00000  0.050932\n",
       "...              ...      ...       ...\n",
       "80491  g2_e2v_seeded  0.99682  0.999637\n",
       "80492  g2_e2v_seeded  0.99682  0.999818\n",
       "80493  g2_e2v_seeded  0.99891  0.999818\n",
       "80494  g2_e2v_seeded  0.99891  1.000000\n",
       "80495  g2_e2v_seeded  1.00000  1.000000\n",
       "\n",
       "[80496 rows x 3 columns]"
      ]
     },
     "execution_count": 387,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roc_curve_all_runs_all_models = pd.DataFrame(roc_curve_all_runs_all_models)\n",
    "roc_curve_all_runs_all_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 388,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "ax.set_title(f'ROC TPR/ROC FPR Distribution over Each Run for Each Model')\n",
    "sns.scatterplot(data=roc_curve_all_runs_all_models, x=\"ROC FPR\", y=\"ROC TPR\", hue=\"Model\", s=1)\n",
    "ax.set_xlabel('False Positive Rate')\n",
    "ax.set_ylabel('True Positive Rate')\n",
    "ax.legend(markerscale=10)\n",
    "\n",
    "fig.savefig(f'{curr_output_dir}/roc_curves.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot F1 Score for Each Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 389,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>F1 Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.921807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.923526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.925676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.922628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.924252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.924839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.920293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.926195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.925833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.931166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.884833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.889119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.899810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.888614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.898992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.890450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.891830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.889626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.891420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.896495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.910082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.910298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.919804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.908163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.911308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.911107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.920201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.910957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.916434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.912847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.896840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.894679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.906000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.894834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.895855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.903404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.892316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.903643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.900217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.892098</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Model  F1 Score\n",
       "0          g1_e2v  0.921807\n",
       "1          g1_e2v  0.923526\n",
       "2          g1_e2v  0.925676\n",
       "3          g1_e2v  0.922628\n",
       "4          g1_e2v  0.924252\n",
       "5          g1_e2v  0.924839\n",
       "6          g1_e2v  0.920293\n",
       "7          g1_e2v  0.926195\n",
       "8          g1_e2v  0.925833\n",
       "9          g1_e2v  0.931166\n",
       "10  g1_e2v_seeded  0.884833\n",
       "11  g1_e2v_seeded  0.889119\n",
       "12  g1_e2v_seeded  0.899810\n",
       "13  g1_e2v_seeded  0.888614\n",
       "14  g1_e2v_seeded  0.898992\n",
       "15  g1_e2v_seeded  0.890450\n",
       "16  g1_e2v_seeded  0.891830\n",
       "17  g1_e2v_seeded  0.889626\n",
       "18  g1_e2v_seeded  0.891420\n",
       "19  g1_e2v_seeded  0.896495\n",
       "20         g2_e2v  0.910082\n",
       "21         g2_e2v  0.910298\n",
       "22         g2_e2v  0.919804\n",
       "23         g2_e2v  0.908163\n",
       "24         g2_e2v  0.911308\n",
       "25         g2_e2v  0.911107\n",
       "26         g2_e2v  0.920201\n",
       "27         g2_e2v  0.910957\n",
       "28         g2_e2v  0.916434\n",
       "29         g2_e2v  0.912847\n",
       "30  g2_e2v_seeded  0.896840\n",
       "31  g2_e2v_seeded  0.894679\n",
       "32  g2_e2v_seeded  0.906000\n",
       "33  g2_e2v_seeded  0.894834\n",
       "34  g2_e2v_seeded  0.895855\n",
       "35  g2_e2v_seeded  0.903404\n",
       "36  g2_e2v_seeded  0.892316\n",
       "37  g2_e2v_seeded  0.903643\n",
       "38  g2_e2v_seeded  0.900217\n",
       "39  g2_e2v_seeded  0.892098"
      ]
     },
     "execution_count": 389,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f1_scores_all_runs_all_models = pd.DataFrame(f1_scores_all_runs_all_models)\n",
    "f1_scores_all_runs_all_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 390,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "ax.set_title(f'F1 Scores Overview for Each Model')\n",
    "sns.barplot(f1_scores_all_runs_all_models, x=\"Model\", y=\"F1 Score\", errorbar=\"sd\", color='cornflowerblue')\n",
    "ax.bar_label(ax.containers[0], fontsize=10)\n",
    "ax.set_xlabel('Model Variant')\n",
    "ax.set_ylabel('F1 Score')\n",
    "\n",
    "fig.savefig(f'{curr_output_dir}/f1_scores.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Final AUC-ROC Score for Each Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 391,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>ROC AUC Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.976740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.975250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.978672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.978382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.976009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.979404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.975915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.978415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.974870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>g1_e2v</td>\n",
       "      <td>0.979653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.958313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.960103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.965422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.960483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.963255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.961446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.961444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.954658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.961569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>g1_e2v_seeded</td>\n",
       "      <td>0.965426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.969946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.971045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.969295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.971127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.971025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.969182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.970812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.967586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.971979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>g2_e2v</td>\n",
       "      <td>0.970242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.955379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.955509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.956912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.953097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.954805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.958293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.954299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.958360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.957697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>g2_e2v_seeded</td>\n",
       "      <td>0.954220</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Model  ROC AUC Score\n",
       "0          g1_e2v       0.976740\n",
       "1          g1_e2v       0.975250\n",
       "2          g1_e2v       0.978672\n",
       "3          g1_e2v       0.978382\n",
       "4          g1_e2v       0.976009\n",
       "5          g1_e2v       0.979404\n",
       "6          g1_e2v       0.975915\n",
       "7          g1_e2v       0.978415\n",
       "8          g1_e2v       0.974870\n",
       "9          g1_e2v       0.979653\n",
       "10  g1_e2v_seeded       0.958313\n",
       "11  g1_e2v_seeded       0.960103\n",
       "12  g1_e2v_seeded       0.965422\n",
       "13  g1_e2v_seeded       0.960483\n",
       "14  g1_e2v_seeded       0.963255\n",
       "15  g1_e2v_seeded       0.961446\n",
       "16  g1_e2v_seeded       0.961444\n",
       "17  g1_e2v_seeded       0.954658\n",
       "18  g1_e2v_seeded       0.961569\n",
       "19  g1_e2v_seeded       0.965426\n",
       "20         g2_e2v       0.969946\n",
       "21         g2_e2v       0.971045\n",
       "22         g2_e2v       0.969295\n",
       "23         g2_e2v       0.971127\n",
       "24         g2_e2v       0.971025\n",
       "25         g2_e2v       0.969182\n",
       "26         g2_e2v       0.970812\n",
       "27         g2_e2v       0.967586\n",
       "28         g2_e2v       0.971979\n",
       "29         g2_e2v       0.970242\n",
       "30  g2_e2v_seeded       0.955379\n",
       "31  g2_e2v_seeded       0.955509\n",
       "32  g2_e2v_seeded       0.956912\n",
       "33  g2_e2v_seeded       0.953097\n",
       "34  g2_e2v_seeded       0.954805\n",
       "35  g2_e2v_seeded       0.958293\n",
       "36  g2_e2v_seeded       0.954299\n",
       "37  g2_e2v_seeded       0.958360\n",
       "38  g2_e2v_seeded       0.957697\n",
       "39  g2_e2v_seeded       0.954220"
      ]
     },
     "execution_count": 391,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_test_auc_roc_scores_all_runs_all_models = pd.DataFrame(final_test_auc_roc_scores_all_runs_all_models)\n",
    "final_test_auc_roc_scores_all_runs_all_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 392,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "ax.set_title(f'AUC-ROC Scores Overview for Each Model')\n",
    "sns.barplot(final_test_auc_roc_scores_all_runs_all_models, x=\"Model\", y=\"ROC AUC Score\", errorbar=\"sd\")\n",
    "ax.bar_label(ax.containers[0], fontsize=10)\n",
    "ax.set_xlabel('Model Variant')\n",
    "ax.set_ylabel('AUC-ROC Score')\n",
    "\n",
    "fig.savefig(f'{curr_output_dir}/auc_roc_scores.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Training Curve"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 393,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run</th>\n",
       "      <th>name</th>\n",
       "      <th>iteration</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>0</td>\n",
       "      <td>0.967075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>1</td>\n",
       "      <td>0.964843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>2</td>\n",
       "      <td>0.963116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>3</td>\n",
       "      <td>0.959631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>4</td>\n",
       "      <td>0.958112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2965</th>\n",
       "      <td>run_010</td>\n",
       "      <td>AUC Test</td>\n",
       "      <td>94</td>\n",
       "      <td>0.969745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2966</th>\n",
       "      <td>run_010</td>\n",
       "      <td>AUC Test</td>\n",
       "      <td>95</td>\n",
       "      <td>0.969877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2967</th>\n",
       "      <td>run_010</td>\n",
       "      <td>AUC Test</td>\n",
       "      <td>96</td>\n",
       "      <td>0.969975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2968</th>\n",
       "      <td>run_010</td>\n",
       "      <td>AUC Test</td>\n",
       "      <td>97</td>\n",
       "      <td>0.970116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2969</th>\n",
       "      <td>run_010</td>\n",
       "      <td>AUC Test</td>\n",
       "      <td>98</td>\n",
       "      <td>0.970242</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2970 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          run       name  iteration     score\n",
       "0     run_001  AUC Train          0  0.967075\n",
       "1     run_001  AUC Train          1  0.964843\n",
       "2     run_001  AUC Train          2  0.963116\n",
       "3     run_001  AUC Train          3  0.959631\n",
       "4     run_001  AUC Train          4  0.958112\n",
       "...       ...        ...        ...       ...\n",
       "2965  run_010   AUC Test         94  0.969745\n",
       "2966  run_010   AUC Test         95  0.969877\n",
       "2967  run_010   AUC Test         96  0.969975\n",
       "2968  run_010   AUC Test         97  0.970116\n",
       "2969  run_010   AUC Test         98  0.970242\n",
       "\n",
       "[2970 rows x 4 columns]"
      ]
     },
     "execution_count": 393,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "auc_scores_all_runs = pd.DataFrame(auc_scores_all_runs)\n",
    "auc_scores_all_runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 394,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run</th>\n",
       "      <th>name</th>\n",
       "      <th>iteration</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>0</td>\n",
       "      <td>0.967075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>1</td>\n",
       "      <td>0.964843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>2</td>\n",
       "      <td>0.963116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>3</td>\n",
       "      <td>0.959631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>run_001</td>\n",
       "      <td>AUC Train</td>\n",
       "      <td>4</td>\n",
       "      <td>0.958112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3955</th>\n",
       "      <td>run_010</td>\n",
       "      <td>Cross-Entropy Loss</td>\n",
       "      <td>94</td>\n",
       "      <td>-0.221635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3956</th>\n",
       "      <td>run_010</td>\n",
       "      <td>Cross-Entropy Loss</td>\n",
       "      <td>95</td>\n",
       "      <td>-0.232123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3957</th>\n",
       "      <td>run_010</td>\n",
       "      <td>Cross-Entropy Loss</td>\n",
       "      <td>96</td>\n",
       "      <td>-0.225450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3958</th>\n",
       "      <td>run_010</td>\n",
       "      <td>Cross-Entropy Loss</td>\n",
       "      <td>97</td>\n",
       "      <td>-0.232333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3959</th>\n",
       "      <td>run_010</td>\n",
       "      <td>Cross-Entropy Loss</td>\n",
       "      <td>98</td>\n",
       "      <td>-0.232543</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3960 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          run                name  iteration     score\n",
       "0     run_001           AUC Train          0  0.967075\n",
       "1     run_001           AUC Train          1  0.964843\n",
       "2     run_001           AUC Train          2  0.963116\n",
       "3     run_001           AUC Train          3  0.959631\n",
       "4     run_001           AUC Train          4  0.958112\n",
       "...       ...                 ...        ...       ...\n",
       "3955  run_010  Cross-Entropy Loss         94 -0.221635\n",
       "3956  run_010  Cross-Entropy Loss         95 -0.232123\n",
       "3957  run_010  Cross-Entropy Loss         96 -0.225450\n",
       "3958  run_010  Cross-Entropy Loss         97 -0.232333\n",
       "3959  run_010  Cross-Entropy Loss         98 -0.232543\n",
       "\n",
       "[3960 rows x 4 columns]"
      ]
     },
     "execution_count": 394,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "auc_loss_scores_all_runs = pd.DataFrame(auc_loss_scores_all_runs)\n",
    "auc_loss_scores_all_runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x800 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.set_title(f'Training curve on dataset {dataset_nr} with method {embedding_method}{title_seeded}')\n",
    "sns.lineplot(data=auc_scores_all_runs, x='iteration', y='score', hue='name')\n",
    "ax.set_ylim(0.85,1)\n",
    "ax.set_xlabel('Iteration')\n",
    "ax.set_ylabel('AUC-ROC Score')\n",
    "\n",
    "print(dataset_output_dir)\n",
    "fig.savefig(f'{dataset_output_dir}/training_curve.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x800 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.set_title(f'Training curve on dataset {dataset_nr} with method {embedding_method}{title_seeded}')\n",
    "sns.lineplot(data=auc_loss_scores_all_runs, x='iteration', y='score', hue='name')\n",
    "ax.set_ylim(top=1)\n",
    "ax.set_xlabel('Iteration')\n",
    "ax.set_ylabel('AUC-ROC/log10(Loss)')\n",
    "\n",
    "fig.savefig(f'{dataset_output_dir}/training_curve_with_loss.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Similarity between top scoring symptom-drug pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_symptom_pairs_per_run = []\n",
    "\n",
    "for index, pred_path in enumerate(pred_folders_paths):\n",
    "    with open(f'{pred_path}/candidates_per_symptom_{dataset_nr}_{embedding_method}.pkl', 'rb') as f:\n",
    "        loaded_list = pickle.load(f)\n",
    "        \n",
    "        drug_symptom_pairs = []\n",
    "        \n",
    "        for _, row in loaded_list.iterrows():\n",
    "            symptom_id = row['Symptom']\n",
    "            candidates = row['Candidates']\n",
    "            \n",
    "            for candidate in candidates:\n",
    "                drug_symptom_pairs.append(tuple([symptom_id, candidate]))\n",
    "                \n",
    "        total_drug_symptom_pairs = len(drug_symptom_pairs)\n",
    "    \n",
    "    drug_symptom_pairs_per_run.append(drug_symptom_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run 1</th>\n",
       "      <th>run 2</th>\n",
       "      <th>run 3</th>\n",
       "      <th>run 4</th>\n",
       "      <th>run 5</th>\n",
       "      <th>run 6</th>\n",
       "      <th>run 7</th>\n",
       "      <th>run 8</th>\n",
       "      <th>run 9</th>\n",
       "      <th>run 10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>run 1</th>\n",
       "      <td>100.000000</td>\n",
       "      <td>58.024691</td>\n",
       "      <td>53.086420</td>\n",
       "      <td>48.148148</td>\n",
       "      <td>45.679012</td>\n",
       "      <td>45.679012</td>\n",
       "      <td>59.259259</td>\n",
       "      <td>59.259259</td>\n",
       "      <td>44.444444</td>\n",
       "      <td>54.320988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 2</th>\n",
       "      <td>58.024691</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>45.679012</td>\n",
       "      <td>44.444444</td>\n",
       "      <td>53.086420</td>\n",
       "      <td>48.148148</td>\n",
       "      <td>48.148148</td>\n",
       "      <td>41.975309</td>\n",
       "      <td>43.209877</td>\n",
       "      <td>46.913580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 3</th>\n",
       "      <td>53.086420</td>\n",
       "      <td>45.679012</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>40.740741</td>\n",
       "      <td>34.567901</td>\n",
       "      <td>28.395062</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>56.790123</td>\n",
       "      <td>44.444444</td>\n",
       "      <td>59.259259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 4</th>\n",
       "      <td>48.148148</td>\n",
       "      <td>44.444444</td>\n",
       "      <td>40.740741</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>34.567901</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>48.148148</td>\n",
       "      <td>45.679012</td>\n",
       "      <td>40.740741</td>\n",
       "      <td>41.975309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 5</th>\n",
       "      <td>45.679012</td>\n",
       "      <td>53.086420</td>\n",
       "      <td>34.567901</td>\n",
       "      <td>34.567901</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>49.382716</td>\n",
       "      <td>49.382716</td>\n",
       "      <td>37.037037</td>\n",
       "      <td>43.209877</td>\n",
       "      <td>43.209877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 6</th>\n",
       "      <td>45.679012</td>\n",
       "      <td>48.148148</td>\n",
       "      <td>28.395062</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>49.382716</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>39.506173</td>\n",
       "      <td>46.913580</td>\n",
       "      <td>25.925926</td>\n",
       "      <td>41.975309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 7</th>\n",
       "      <td>59.259259</td>\n",
       "      <td>48.148148</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>48.148148</td>\n",
       "      <td>49.382716</td>\n",
       "      <td>39.506173</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>46.913580</td>\n",
       "      <td>37.037037</td>\n",
       "      <td>38.271605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 8</th>\n",
       "      <td>59.259259</td>\n",
       "      <td>41.975309</td>\n",
       "      <td>56.790123</td>\n",
       "      <td>45.679012</td>\n",
       "      <td>37.037037</td>\n",
       "      <td>46.913580</td>\n",
       "      <td>46.913580</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>37.037037</td>\n",
       "      <td>53.086420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 9</th>\n",
       "      <td>44.444444</td>\n",
       "      <td>43.209877</td>\n",
       "      <td>44.444444</td>\n",
       "      <td>40.740741</td>\n",
       "      <td>43.209877</td>\n",
       "      <td>25.925926</td>\n",
       "      <td>37.037037</td>\n",
       "      <td>37.037037</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>39.506173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run 10</th>\n",
       "      <td>54.320988</td>\n",
       "      <td>46.913580</td>\n",
       "      <td>59.259259</td>\n",
       "      <td>41.975309</td>\n",
       "      <td>43.209877</td>\n",
       "      <td>41.975309</td>\n",
       "      <td>38.271605</td>\n",
       "      <td>53.086420</td>\n",
       "      <td>39.506173</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             run 1       run 2       run 3       run 4       run 5   \n",
       "run 1   100.000000   58.024691   53.086420   48.148148   45.679012  \\\n",
       "run 2    58.024691  100.000000   45.679012   44.444444   53.086420   \n",
       "run 3    53.086420   45.679012  100.000000   40.740741   34.567901   \n",
       "run 4    48.148148   44.444444   40.740741  100.000000   34.567901   \n",
       "run 5    45.679012   53.086420   34.567901   34.567901  100.000000   \n",
       "run 6    45.679012   48.148148   28.395062   33.333333   49.382716   \n",
       "run 7    59.259259   48.148148   33.333333   48.148148   49.382716   \n",
       "run 8    59.259259   41.975309   56.790123   45.679012   37.037037   \n",
       "run 9    44.444444   43.209877   44.444444   40.740741   43.209877   \n",
       "run 10   54.320988   46.913580   59.259259   41.975309   43.209877   \n",
       "\n",
       "             run 6       run 7       run 8       run 9      run 10  \n",
       "run 1    45.679012   59.259259   59.259259   44.444444   54.320988  \n",
       "run 2    48.148148   48.148148   41.975309   43.209877   46.913580  \n",
       "run 3    28.395062   33.333333   56.790123   44.444444   59.259259  \n",
       "run 4    33.333333   48.148148   45.679012   40.740741   41.975309  \n",
       "run 5    49.382716   49.382716   37.037037   43.209877   43.209877  \n",
       "run 6   100.000000   39.506173   46.913580   25.925926   41.975309  \n",
       "run 7    39.506173  100.000000   46.913580   37.037037   38.271605  \n",
       "run 8    46.913580   46.913580  100.000000   37.037037   53.086420  \n",
       "run 9    25.925926   37.037037   37.037037  100.000000   39.506173  \n",
       "run 10   41.975309   38.271605   53.086420   39.506173  100.000000  "
      ]
     },
     "execution_count": 398,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similarity_matrix = {}\n",
    "ratios_non_diagonals = []\n",
    "\n",
    "for index1, pairs1 in enumerate(drug_symptom_pairs_per_run):\n",
    "    \n",
    "    similarities = {}\n",
    "    \n",
    "    for index2, pairs2 in enumerate(drug_symptom_pairs_per_run):\n",
    "        overlap = set([tuple(sorted(ele)) for ele in pairs1]) & set([tuple(sorted(ele)) for ele in pairs2])\n",
    "        ratio_overlap = len(overlap) / total_drug_symptom_pairs * 100\n",
    "        \n",
    "        similarities[f'run {index2+1}'] = ratio_overlap\n",
    "        \n",
    "        if index1 != index2:\n",
    "            ratios_non_diagonals.append(ratio_overlap)\n",
    "        \n",
    "    similarity_matrix[f'run {index1+1}'] = similarities\n",
    "    \n",
    "similarity_matrix_df = pd.DataFrame(similarity_matrix)\n",
    "similarity_matrix_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean: 44.6639231824417\n",
      "Median: 44.44444444444444\n"
     ]
    }
   ],
   "source": [
    "mean_overlap_ratio = np.mean(ratios_non_diagonals)\n",
    "median_overlap_ratio = np.median(ratios_non_diagonals)\n",
    "print('Mean:', mean_overlap_ratio)\n",
    "print('Median:', median_overlap_ratio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x800 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "ax.set_title(f'Overlap ratio (Mean: {mean_overlap_ratio}, Median: {median_overlap_ratio}) between list of predicted symptom-drug pairs per run on dataset {dataset_nr} with method {embedding_method}{title_seeded}')\n",
    "sns.heatmap(similarity_matrix_df, annot=True, fmt='.1f', linewidths=0.5, ax=ax, cmap='RdYlGn')\n",
    "ax.collections[0].set_clim(0,100)\n",
    "\n",
    "fig.savefig(f'{dataset_output_dir}/overlap_between_runs.png', bbox_inches='tight')\n",
    "fig.clear()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 8 symptom-drug pairs that are found in the top list of drug candidates in 10 runs: \n",
      " {('HP:0003236', '1576'), ('HP:0008981', '1576'), ('HP:0001290', '269'), ('HP:0001635', '1576'), ('HP:0003701', '1576'), ('HP:0011675', '231'), ('HP:0011675', '1576'), ('HP:0001644', '1576')}\n"
     ]
    }
   ],
   "source": [
    "for i in range(0, len(drug_symptom_pairs_per_run)):\n",
    "    if i == 0:\n",
    "        overlapping_pairs_all_runs = set(drug_symptom_pairs_per_run[i])\n",
    "    else:\n",
    "        overlapping_pairs_all_runs = overlapping_pairs_all_runs & set(drug_symptom_pairs_per_run[i])\n",
    "            \n",
    "print(f'There are {len(overlapping_pairs_all_runs)} symptom-drug pairs that are found in the top list of drug candidates in {len(drug_symptom_pairs_per_run)} runs: \\n {overlapping_pairs_all_runs}')\n",
    "\n",
    "with open(f'{dataset_output_dir}/symptom_drug_pair_overlapping_all_runs_{dataset_nr}_{embedding_method}{fixed_emb}.pkl', 'wb') as f:\n",
    "    pickle.dump(overlapping_pairs_all_runs, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check overlap between each setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>semantic</th>\n",
       "      <th>label</th>\n",
       "      <th>semantic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MP:0004187</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>cardia bifida</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZP:0100138</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>muscle tendon junction myotome increased amoun...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1346525</td>\n",
       "      <td>gene</td>\n",
       "      <td>Sgcd</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OMIM:300377.0044</td>\n",
       "      <td>variant</td>\n",
       "      <td>DMD, LYS770TER</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZP:0002210</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>posterior lateral line neuromast primordium mi...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10270</th>\n",
       "      <td>ZP:0014934</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>atrioventricular valve development process qua...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10271</th>\n",
       "      <td>ENSEMBL:ENSCAFG00000011207</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSCAFG00000011207</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10272</th>\n",
       "      <td>ENSEMBL:ENSXETG00000039922</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSXETG00000039922</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10273</th>\n",
       "      <td>ENSEMBL:ENSACAG00000010058</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSEMBL:ENSACAG00000010058</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10274</th>\n",
       "      <td>WBPhenotype:0000019</td>\n",
       "      <td>phenotype</td>\n",
       "      <td>pharyngeal pumping reduced</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10275 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               id   semantic   \n",
       "0                      MP:0004187  phenotype  \\\n",
       "1                      ZP:0100138  phenotype   \n",
       "2                     MGI:1346525       gene   \n",
       "3                OMIM:300377.0044    variant   \n",
       "4                      ZP:0002210  phenotype   \n",
       "...                           ...        ...   \n",
       "10270                  ZP:0014934  phenotype   \n",
       "10271  ENSEMBL:ENSCAFG00000011207       gene   \n",
       "10272  ENSEMBL:ENSXETG00000039922       gene   \n",
       "10273  ENSEMBL:ENSACAG00000010058       gene   \n",
       "10274         WBPhenotype:0000019  phenotype   \n",
       "\n",
       "                                                   label  semantic_id  \n",
       "0                                          cardia bifida            9  \n",
       "1      muscle tendon junction myotome increased amoun...            9  \n",
       "2                                                   Sgcd            5  \n",
       "3                                         DMD, LYS770TER           11  \n",
       "4      posterior lateral line neuromast primordium mi...            9  \n",
       "...                                                  ...          ...  \n",
       "10270  atrioventricular valve development process qua...            9  \n",
       "10271                         ENSEMBL:ENSCAFG00000011207            5  \n",
       "10272                         ENSEMBL:ENSXETG00000039922            5  \n",
       "10273                         ENSEMBL:ENSACAG00000010058            5  \n",
       "10274                         pharyngeal pumping reduced            9  \n",
       "\n",
       "[10275 rows x 4 columns]"
      ]
     },
     "execution_count": 402,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = pd.read_csv(f'output/indexed_nodes_{dataset_nr}.csv')\n",
    "nodes.drop('index_id', axis=1, inplace=True)\n",
    "nodes['semantic'] = nodes['semantic'].astype('category')\n",
    "nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 403,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g1_e2v\n",
      "{('HP:0001644', '1576'), ('HP:0001638', '1576'), ('HP:0001635', '1576'), ('HP:0011675', '1576')}\n",
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g1_e2v_seeded\n",
      "{('HP:0001635', '231'), ('HP:0011675', '231'), ('HP:0001638', '231'), ('HP:0000750', '4225'), ('HP:0003560', '624')}\n",
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v\n",
      "{('HP:0003236', '1576'), ('HP:0008981', '1576'), ('HP:0001290', '269'), ('HP:0001635', '1576'), ('HP:0003701', '1576'), ('HP:0011675', '231'), ('HP:0011675', '1576'), ('HP:0001644', '1576')}\n",
      "Output folder for dataset 2 exists and will be loaded: c:\\Users\\rosa-\\Google Drive\\Msc_Bioinformatics\\thesis\\XAIFO-ThesisProject\\output\\g2_e2v_seeded\n",
      "{('HP:0011675', '231'), ('HP:0003307', '269'), ('HP:0001635', '1576')}\n"
     ]
    }
   ],
   "source": [
    "all_settings = [\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '',\n",
    "     'dataset_nr': 1},\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '_seeded',\n",
    "     'dataset_nr': 1},\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '',\n",
    "     'dataset_nr': 2},\n",
    "    {'embedding_method': 'e2v',\n",
    "     'fixed_emb': '_seeded',\n",
    "     'dataset_nr': 2}\n",
    "]\n",
    "\n",
    "curr_working_dir = os.getcwd()\n",
    "curr_output_dir = os.path.join(curr_working_dir, 'output')\n",
    "\n",
    "overlapping_pairs = []\n",
    "\n",
    "for setting in all_settings:\n",
    "    dataset_output_dir = os.path.join(curr_output_dir, f'g{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}')\n",
    "\n",
    "    if not os.path.exists(dataset_output_dir):\n",
    "        print('First, run the edge2vec embedding script. Then, run this script.')\n",
    "    else:\n",
    "        print(f'Output folder for dataset {dataset_nr} exists and will be loaded: {dataset_output_dir}')\n",
    "        \n",
    "        with open(f'{dataset_output_dir}/symptom_drug_pair_overlapping_all_runs_{setting[\"dataset_nr\"]}_{setting[\"embedding_method\"]}{setting[\"fixed_emb\"]}.pkl', 'rb') as f:\n",
    "            loaded_list = pickle.load(f)\n",
    "            print(loaded_list)\n",
    "            overlapping_pairs.append(loaded_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 404,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "levosimendan treats Congestive heart failure\n",
      "levosimendan treats Dilated cardiomyopathy\n",
      "levosimendan treats Arrhythmia\n"
     ]
    }
   ],
   "source": [
    "non_fixed_emb_overlap = overlapping_pairs[0].intersection(overlapping_pairs[2])\n",
    "for pair in non_fixed_emb_overlap:\n",
    "    symptom_id, drug_id = pair\n",
    "    symptom_name = nodes.loc[nodes['id'] == symptom_id]['label'].iloc[0]\n",
    "    drug_name = nodes.loc[nodes['id'] == drug_id]['label'].iloc[0]\n",
    "    print(drug_name, 'treats', symptom_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 405,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aprindine treats Arrhythmia\n"
     ]
    }
   ],
   "source": [
    "fixed_emb_overlap = overlapping_pairs[1].intersection(overlapping_pairs[3])\n",
    "for pair in fixed_emb_overlap:\n",
    "    symptom_id, drug_id = pair\n",
    "    symptom_name = nodes.loc[nodes['id'] == symptom_id]['label'].iloc[0]\n",
    "    drug_name = nodes.loc[nodes['id'] == drug_id]['label'].iloc[0]\n",
    "    print(drug_name, 'treats', symptom_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset1_emb_overlap = overlapping_pairs[0].intersection(overlapping_pairs[1])\n",
    "for pair in dataset1_emb_overlap:\n",
    "    symptom_id, drug_id = pair\n",
    "    symptom_name = nodes.loc[nodes['id'] == symptom_id]['label'].iloc[0]\n",
    "    drug_name = nodes.loc[nodes['id'] == drug_id]['label'].iloc[0]\n",
    "    print(drug_name, 'treats', symptom_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aprindine treats Arrhythmia\n",
      "levosimendan treats Congestive heart failure\n"
     ]
    }
   ],
   "source": [
    "dataset2_emb_overlap = overlapping_pairs[2].intersection(overlapping_pairs[3])\n",
    "for pair in dataset2_emb_overlap:\n",
    "    symptom_id, drug_id = pair\n",
    "    symptom_name = nodes.loc[nodes['id'] == symptom_id]['label'].iloc[0]\n",
    "    drug_name = nodes.loc[nodes['id'] == drug_id]['label'].iloc[0]\n",
    "    print(drug_name, 'treats', symptom_name)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xaifognn",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
